<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.11"/>
<title>AgeStrucNe: agestrucne.pgdriveneestimator Namespace Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { init_search(); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">AgeStrucNe
   </div>
   <div id="projectbrief">GUI Front end with refactoring of population genetics programming provided by Tiago Antao&#39;s AgeStructureNe code, with enhanced subsampling, simulation parameterization, and plotting.  We have several outside dependancies (chiefly simuPop and LDNe2)</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.11 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li class="current"><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="namespaces.html"><span>Namespace&#160;List</span></a></li>
      <li><a href="namespacemembers.html"><span>Namespace&#160;Members</span></a></li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>agestrucne</b></li><li class="navelem"><a class="el" href="namespaceagestrucne_1_1pgdriveneestimator.html">pgdriveneestimator</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#func-members">Functions</a> &#124;
<a href="#var-members">Variables</a>  </div>
  <div class="headertitle">
<div class="title">agestrucne.pgdriveneestimator Namespace Reference</div>  </div>
</div><!--header-->
<div class="contents">
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classagestrucne_1_1pgdriveneestimator_1_1ArgSet.html">ArgSet</a></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classagestrucne_1_1pgdriveneestimator_1_1Counter.html">Counter</a></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classagestrucne_1_1pgdriveneestimator_1_1DebugMode.html">DebugMode</a></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:ac0353109e8454726f6ccf72bad125f1f"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac0353109e8454726f6ccf72bad125f1f"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>set_indices_ne_estimator_output_fields_to_skip</b> ()</td></tr>
<tr class="separator:ac0353109e8454726f6ccf72bad125f1f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aab5085ea0d96c3badf282a68ef214e8c"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aab5085ea0d96c3badf282a68ef214e8c"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>get_datetime_as_string</b> ()</td></tr>
<tr class="separator:aab5085ea0d96c3badf282a68ef214e8c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae1f2a4a7ade58c038c2027844b552149"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ae1f2a4a7ade58c038c2027844b552149"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>set_debug_mode</b> (s_arg)</td></tr>
<tr class="separator:ae1f2a4a7ade58c038c2027844b552149"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2f5f9109904c6af1b938aa800aba93df"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceagestrucne_1_1pgdriveneestimator.html#a2f5f9109904c6af1b938aa800aba93df">get_invalid_file_names</a> (ls_file_names)</td></tr>
<tr class="separator:a2f5f9109904c6af1b938aa800aba93df"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7984118f0813bc707165a4c33b02d11d"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a7984118f0813bc707165a4c33b02d11d"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>file_name_list_is_all_strings</b> (ls_files)</td></tr>
<tr class="separator:a7984118f0813bc707165a4c33b02d11d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aceaee36e8e89f8873e39642960bac265"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceagestrucne_1_1pgdriveneestimator.html#aceaee36e8e89f8873e39642960bac265">get_genepop_file_list</a> (s_genepop_files_arg)</td></tr>
<tr class="separator:aceaee36e8e89f8873e39642960bac265"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abb9562b0e172d40ac1b4d8be37cbc7d6"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="abb9562b0e172d40ac1b4d8be37cbc7d6"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>parse_args</b> (args)</td></tr>
<tr class="separator:abb9562b0e172d40ac1b4d8be37cbc7d6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abc3a175c2af2fd3749c6d7b9ce36cf2f"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="abc3a175c2af2fd3749c6d7b9ce36cf2f"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>validate_chromlocifile</b> (s_chromlocifile, ls_genepop_files, i_allele_pairing_scheme)</td></tr>
<tr class="separator:abc3a175c2af2fd3749c6d7b9ce36cf2f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6a5611213886279b5560f4e2029141bc"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a6a5611213886279b5560f4e2029141bc"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>write_parms_to_file</b> (lv_params)</td></tr>
<tr class="separator:a6a5611213886279b5560f4e2029141bc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7e685dd48b043c1e68edaf09199d4e3c"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a7e685dd48b043c1e68edaf09199d4e3c"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>get_sample_val_and_rep_number_from_sample_name</b> (s_sample_name, s_sample_scheme)</td></tr>
<tr class="separator:a7e685dd48b043c1e68edaf09199d4e3c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a447133ba0efcd50042521407336db81f"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceagestrucne_1_1pgdriveneestimator.html#a447133ba0efcd50042521407336db81f">raise_exception_if_non_single_pop_file</a> (o_genepopfile)</td></tr>
<tr class="separator:a447133ba0efcd50042521407336db81f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abd76f8aa32c542ee3e90799c0e7b166a"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceagestrucne_1_1pgdriveneestimator.html#abd76f8aa32c542ee3e90799c0e7b166a">get_pops_in_file_outside_valid_size_range</a> (o_genepopfile, i_min_pop_size, i_max_pop_size)</td></tr>
<tr class="separator:abd76f8aa32c542ee3e90799c0e7b166a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a78e7def8930f60b0f5021f16b356e1e3"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceagestrucne_1_1pgdriveneestimator.html#a78e7def8930f60b0f5021f16b356e1e3">get_string_values_ldne_ratio_and_bias_adjustment</a> (o_ne_estimator, lv_results_list, f_nbne_ratio)</td></tr>
<tr class="separator:a78e7def8930f60b0f5021f16b356e1e3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a96db81ad67e3433849cb2a83c721b28e"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a96db81ad67e3433849cb2a83c721b28e"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>get_nbne_ratio_from_genepop_file_header</b> (o_genepopfile)</td></tr>
<tr class="separator:a96db81ad67e3433849cb2a83c721b28e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2efaad3207ead8feb0abb42b921ab402"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceagestrucne_1_1pgdriveneestimator.html#a2efaad3207ead8feb0abb42b921ab402">get_mean_het_based_on_allele_freqs</a> (o_genepopfile, s_pop_subsample_tag=None, s_individual_subsample_tag=None, s_loci_subsample_tag=None)</td></tr>
<tr class="separator:a2efaad3207ead8feb0abb42b921ab402"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaf2e5f67e91fb3bff9ad32c7dd0ccf53"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceagestrucne_1_1pgdriveneestimator.html#aaf2e5f67e91fb3bff9ad32c7dd0ccf53">do_ldne_bias_adjustment</a> (f_ldne_estimate, f_nbne_ratio)</td></tr>
<tr class="separator:aaf2e5f67e91fb3bff9ad32c7dd0ccf53"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae574fb9ce112652e14c9b6e97942f634"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceagestrucne_1_1pgdriveneestimator.html#ae574fb9ce112652e14c9b6e97942f634">do_estimate</a> (xxx_todo_changeme)</td></tr>
<tr class="separator:ae574fb9ce112652e14c9b6e97942f634"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8bf1e35804b282f801e806e23eb86c85"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a8bf1e35804b282f801e806e23eb86c85"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>write_results</b> (ds_results, o_main_outfile, o_secondary_outfile)</td></tr>
<tr class="separator:a8bf1e35804b282f801e806e23eb86c85"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abc650a9da8c7f87ab247722309c431d3"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="abc650a9da8c7f87ab247722309c431d3"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>update_indiv_list</b> (dddli_indiv_list, dv_indiv_info_for_replicate, lv_sample_vals)</td></tr>
<tr class="separator:abc650a9da8c7f87ab247722309c431d3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac472c260af8b82aa3a06c620ae59bfb1"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac472c260af8b82aa3a06c620ae59bfb1"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>make_indiv_table_name</b> ()</td></tr>
<tr class="separator:ac472c260af8b82aa3a06c620ae59bfb1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad25aa00447baed0de61dc4c2fcecdd1a"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ad25aa00447baed0de61dc4c2fcecdd1a"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>get_indiv_table_header</b> (lf_proportions)</td></tr>
<tr class="separator:ad25aa00447baed0de61dc4c2fcecdd1a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2ddc3ab4bba61ac360abd9cf2f702e78"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a2ddc3ab4bba61ac360abd9cf2f702e78"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>write_indiv_table</b> (dddli_indiv_list, s_file_name, lf_proportions)</td></tr>
<tr class="separator:a2ddc3ab4bba61ac360abd9cf2f702e78"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a81af4b87f0baa9c662d8384c3bdd9bb1"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a81af4b87f0baa9c662d8384c3bdd9bb1"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>print_test_list_replicate_selection_indices</b> (o_genepopfile, s_sample_scheme, s_indiv_subsample_tag, s_loci_sample_value, s_loci_replicate_number, o_secondary_outfile, s_population_subsample_tag)</td></tr>
<tr class="separator:a81af4b87f0baa9c662d8384c3bdd9bb1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa1487bffaa5efe1e3ce66928866a6e8f"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceagestrucne_1_1pgdriveneestimator.html#aa1487bffaa5efe1e3ce66928866a6e8f">parse_sampling_list_that_uses_indiv_id_fields</a> (ls_sampling_list_items)</td></tr>
<tr class="separator:aa1487bffaa5efe1e3ce66928866a6e8f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adda0b57d249a1bdb8544447189a283c8"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="adda0b57d249a1bdb8544447189a283c8"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>get_field_names_from_criteria_test</b> (ls_field_names, s_test)</td></tr>
<tr class="separator:adda0b57d249a1bdb8544447189a283c8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a41b8feb7b06e0645952de8387f7880f3"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a41b8feb7b06e0645952de8387f7880f3"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>get_criteria_sampling_params</b> (lv_sample_values)</td></tr>
<tr class="separator:a41b8feb7b06e0645952de8387f7880f3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1007aa9bd91c0dce7c53fb4e4940d0b7"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a1007aa9bd91c0dce7c53fb4e4940d0b7"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>get_cohorts_sampling_params</b> (lv_sample_values)</td></tr>
<tr class="separator:a1007aa9bd91c0dce7c53fb4e4940d0b7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4a1e1a8bbb368db807cf857c1c27c784"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a4a1e1a8bbb368db807cf857c1c27c784"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>get_relateds_sampling_params</b> (lv_sample_values)</td></tr>
<tr class="separator:a4a1e1a8bbb368db807cf857c1c27c784"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1d38a43a7280fa4c608458abd861bacc"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceagestrucne_1_1pgdriveneestimator.html#a1d38a43a7280fa4c608458abd861bacc">do_sample</a> (o_genepopfile, i_min_pop_range, i_max_pop_range, i_min_pop_size, i_max_pop_size, s_sample_scheme, lv_sample_values, i_replicates, s_loci_sampling_scheme, v_loci_sampling_scheme_param, i_min_loci_position, i_max_loci_position, i_min_total_loci, i_max_total_loci, i_loci_replicates)</td></tr>
<tr class="separator:a1d38a43a7280fa4c608458abd861bacc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8d8109b98382d21b972e701323440f95"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceagestrucne_1_1pgdriveneestimator.html#a8d8109b98382d21b972e701323440f95">do_sample_individuals</a> (o_genepopfile, i_min_pop_range, i_max_pop_range, i_min_pop_size, i_max_pop_size, s_sample_scheme, lv_sample_values, i_replicates)</td></tr>
<tr class="separator:a8d8109b98382d21b972e701323440f95"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9544ec0a253ab2bcc1fcfbe1edb66120"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceagestrucne_1_1pgdriveneestimator.html#a9544ec0a253ab2bcc1fcfbe1edb66120">do_sample_loci</a> (o_genepopfile, s_loci_sampling_scheme, v_loci_sampling_scheme_param, i_min_loci_position, i_max_loci_position, i_min_total_loci, i_max_total_loci, li_population_list, i_loci_replicates=1)</td></tr>
<tr class="separator:a9544ec0a253ab2bcc1fcfbe1edb66120"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adebe964cadd26d7aca365586ef17e3b1"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceagestrucne_1_1pgdriveneestimator.html#adebe964cadd26d7aca365586ef17e3b1">get_subsample_genepop_file_name</a> (s_original_genepop_file_name, s_loci_subsample_tag, s_this_pop_sample_name, s_temporary_directory)</td></tr>
<tr class="separator:adebe964cadd26d7aca365586ef17e3b1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2ea5983385f4ef857bcf935d904c59b1"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceagestrucne_1_1pgdriveneestimator.html#a2ea5983385f4ef857bcf935d904c59b1">make_subsample_genepop_file_name</a> (s_original_genepop_file, s_temporary_directory, i_genepop_file_count, i_indiv_sample_count, i_loci_subsample_count)</td></tr>
<tr class="separator:a2ea5983385f4ef857bcf935d904c59b1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af69b7cf5fb237f79980c1d77c9134e6e"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceagestrucne_1_1pgdriveneestimator.html#af69b7cf5fb237f79980c1d77c9134e6e">add_to_set_of_calls_to_do_estimate</a> (o_genepopfile, s_sample_scheme, f_min_allele_freq, b_monogamy, s_chromlocifile, i_allele_pairing_scheme, i_min_pop_size, i_max_pop_size, i_min_loci_position, i_max_loci_position, o_debug_mode, llv_args_each_process, IDX_NE_ESTIMATOR_OUTPUT_FIELDS_TO_SKIP, o_secondary_outfile, f_nbne_ratio, s_temporary_directory, i_genepop_file_count)</td></tr>
<tr class="separator:af69b7cf5fb237f79980c1d77c9134e6e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6eea3cf1501fe02f36abfa30d6b4b7e3"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceagestrucne_1_1pgdriveneestimator.html#a6eea3cf1501fe02f36abfa30d6b4b7e3">get_loci_sample_val_and_rep_number_from_loci_sample_tag</a> (s_loci_subsample_tag)</td></tr>
<tr class="separator:a6eea3cf1501fe02f36abfa30d6b4b7e3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4a9e6dc9bc38b06c097411e444b40097"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a4a9e6dc9bc38b06c097411e444b40097"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>get_loci_subsample_tags_for_this_indiv_subsample</b> (o_genepopfile, s_indiv_sample_tag)</td></tr>
<tr class="separator:a4a9e6dc9bc38b06c097411e444b40097"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2e891426f08df5f6bb919ac0593b2775"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a2e891426f08df5f6bb919ac0593b2775"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>write_result_sets</b> (lds_results, lv_sample_values, o_debug_mode, o_main_outfile, o_secondary_outfile)</td></tr>
<tr class="separator:a2e891426f08df5f6bb919ac0593b2775"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac17c9812db925b33202fbe6d7d905602"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac17c9812db925b33202fbe6d7d905602"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>execute_ne_for_each_sample</b> (llv_args_each_process, o_process_pool, o_debug_mode, o_multiprocessing_event)</td></tr>
<tr class="separator:ac17c9812db925b33202fbe6d7d905602"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaba26add66aaee9119650828c2803cf7"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceagestrucne_1_1pgdriveneestimator.html#aaba26add66aaee9119650828c2803cf7">get_count_estimator_fields</a> ()</td></tr>
<tr class="separator:aaba26add66aaee9119650828c2803cf7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a36bb91fea51142d72513285ba491ec76"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a36bb91fea51142d72513285ba491ec76"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>write_header_main_table</b> (IDX_NE_ESTIMATOR_OUTPUT_FIELDS_TO_SKIP, o_main_outfile)</td></tr>
<tr class="separator:a36bb91fea51142d72513285ba491ec76"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae832b335a69b61e7b1d17466e1bf7e75"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ae832b335a69b61e7b1d17466e1bf7e75"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>get_sample_abbrevs_for_criteria</b> (lds_results)</td></tr>
<tr class="separator:ae832b335a69b61e7b1d17466e1bf7e75"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7206ae30b244244c34c6ce220c1794de"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a7206ae30b244244c34c6ce220c1794de"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>get_total_bytes_needed_for_genepop_object</b> (s_filename)</td></tr>
<tr class="separator:a7206ae30b244244c34c6ce220c1794de"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abdd18d9adf72d7095604907f27df6364"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="abdd18d9adf72d7095604907f27df6364"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>get_total_ram_usage_for_this_file</b> (s_filename, o_debug_mode, i_total_processes)</td></tr>
<tr class="separator:abdd18d9adf72d7095604907f27df6364"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae8b455ac8aaa3daf7462bed4ded22816"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ae8b455ac8aaa3daf7462bed4ded22816"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>file_can_be_added_to_current_set</b> (s_filename, o_total_bytes_currently_used_for_gp_objects, i_total_processes, o_debug_mode, i_bytes_available_virtual_memory)</td></tr>
<tr class="separator:ae8b455ac8aaa3daf7462bed4ded22816"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a78aff716936fb1c83dcaad808aca597e"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a78aff716936fb1c83dcaad808aca597e"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>add_file_to_current_set</b> (s_filename, i_min_pop_range, i_max_pop_range, i_min_pop_size, i_max_pop_size, s_sample_scheme, lv_sample_values, i_total_replicates, s_loci_sampling_scheme, v_loci_sampling_scheme_param, i_min_loci_position, i_max_loci_position, i_min_total_loci, i_max_total_loci, i_loci_replicates, llv_args_each_process, b_do_nb_bias_adjustment, f_nbne_ratio, f_min_allele_freq, b_monogamy, s_chromlocifile, i_allele_pairing_scheme, o_debug_mode, IDX_NE_ESTIMATOR_OUTPUT_FIELDS_TO_SKIP, i_genepop_file_count, s_temporary_directory, o_secondary_outfile)</td></tr>
<tr class="separator:a78aff716936fb1c83dcaad808aca597e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a715965f92746e915d823b24bda44c507"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a715965f92746e915d823b24bda44c507"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>raise_exception_file_size_limit</b> (s_filename, o_debug_mode, i_total_processes)</td></tr>
<tr class="separator:a715965f92746e915d823b24bda44c507"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0a714bfdcabf2fc3d9832c31767a4ab7"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a0a714bfdcabf2fc3d9832c31767a4ab7"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>process_current_set_of_gp_files</b> (llv_args_each_process, s_filename, o_debug_mode, i_total_processes, o_multiprocessing_event, s_sample_scheme, lv_sample_values, o_main_outfile, o_secondary_outfile, o_total_calls_to_do_estimate, IDX_NE_ESTIMATOR_OUTPUT_FIELDS_TO_SKIP, s_all_interrupt_messages)</td></tr>
<tr class="separator:a0a714bfdcabf2fc3d9832c31767a4ab7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3c8f6d0b93af69f39eda872b14a5aeda"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a3c8f6d0b93af69f39eda872b14a5aeda"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>drive_estimator</b> (args)</td></tr>
<tr class="separator:a3c8f6d0b93af69f39eda872b14a5aeda"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9d972a5914ff6739e6948aeb462bac75"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a9d972a5914ff6739e6948aeb462bac75"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>set_mod_level_var_to_ldne2_executable_path</b> ()</td></tr>
<tr class="separator:a9d972a5914ff6739e6948aeb462bac75"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af8fae2302e9c6cd92e92a8c76e267209"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="af8fae2302e9c6cd92e92a8c76e267209"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>did_find_ne_estimator_executable</b> ()</td></tr>
<tr class="separator:af8fae2302e9c6cd92e92a8c76e267209"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af82482f5cfed335c23bde8de169b44be"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceagestrucne_1_1pgdriveneestimator.html#af82482f5cfed335c23bde8de169b44be">mymainlongfilelist</a> (q_args)</td></tr>
<tr class="separator:af82482f5cfed335c23bde8de169b44be"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac504fdce8e8069dc8342ab9bcb40c49e"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceagestrucne_1_1pgdriveneestimator.html#ac504fdce8e8069dc8342ab9bcb40c49e">set_messaging_procedures</a> (s_gui_flag)</td></tr>
<tr class="separator:ac504fdce8e8069dc8342ab9bcb40c49e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8373a4ba256fa5cd5d0bc88da91815b4"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespaceagestrucne_1_1pgdriveneestimator.html#a8373a4ba256fa5cd5d0bc88da91815b4">mymain</a> (q_args)</td></tr>
<tr class="separator:a8373a4ba256fa5cd5d0bc88da91815b4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a02fc6e8ca3d338d48d7b1579aa9702ea"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a02fc6e8ca3d338d48d7b1579aa9702ea"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>get_ranges_for_list_slices</b> (i_length_of_list, i_slice_size=POOL_BATCH_SIZE)</td></tr>
<tr class="separator:a02fc6e8ca3d338d48d7b1579aa9702ea"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="var-members"></a>
Variables</h2></td></tr>
<tr class="memitem:ad064b0ee52e3f43776eae5998a1179a8"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ad064b0ee52e3f43776eae5998a1179a8"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><b>VERBOSE</b> = False</td></tr>
<tr class="separator:ad064b0ee52e3f43776eae5998a1179a8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae69b14d90d15d9f9f96f2ed8af7b5c67"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ae69b14d90d15d9f9f96f2ed8af7b5c67"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><b>VERY_VERBOSE</b> = False</td></tr>
<tr class="separator:ae69b14d90d15d9f9f96f2ed8af7b5c67"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a81888de46cbb73fd010d3e85be9bb557"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a81888de46cbb73fd010d3e85be9bb557"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>ACTIVATE_GUI_MESSAGING</b> = None</td></tr>
<tr class="separator:a81888de46cbb73fd010d3e85be9bb557"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac85fd42c8429a1209ec19416a326eeff"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac85fd42c8429a1209ec19416a326eeff"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>GUIERR</b> = None</td></tr>
<tr class="separator:ac85fd42c8429a1209ec19416a326eeff"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a51bf178db84b4a07667999442a1da2bd"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a51bf178db84b4a07667999442a1da2bd"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>GUIINFO</b> = None</td></tr>
<tr class="separator:a51bf178db84b4a07667999442a1da2bd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abfa46e39b3b19203cfd40f69a5ba9488"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="abfa46e39b3b19203cfd40f69a5ba9488"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>NEESTIMATOR</b> = pgne.NEESTIMATOR</td></tr>
<tr class="separator:abfa46e39b3b19203cfd40f69a5ba9488"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a23e1540cb588d15bf270ba18fe695b9d"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a23e1540cb588d15bf270ba18fe695b9d"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>LDNE2</b> = pgne.LDNE_ESTIMATION</td></tr>
<tr class="separator:a23e1540cb588d15bf270ba18fe695b9d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac0f5e21e0cf4e067383e3875721c5a6d"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac0f5e21e0cf4e067383e3875721c5a6d"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>ESTIMATOR_TO_USE</b> = LDNE2</td></tr>
<tr class="separator:ac0f5e21e0cf4e067383e3875721c5a6d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a738ae8bbdcb9f1d686ff9c5c958abd08"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a738ae8bbdcb9f1d686ff9c5c958abd08"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>PATH_TO_LDNE2</b> = None</td></tr>
<tr class="separator:a738ae8bbdcb9f1d686ff9c5c958abd08"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0c1ee4219ee4c76944f2146c1d30b136"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a0c1ee4219ee4c76944f2146c1d30b136"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>spacer</b> = &quot; &quot;</td></tr>
<tr class="separator:a0c1ee4219ee4c76944f2146c1d30b136"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa06336ee4a29461cbcdba4b3a671ce18"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aa06336ee4a29461cbcdba4b3a671ce18"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>spacer2</b> = &quot; &quot;</td></tr>
<tr class="separator:aa06336ee4a29461cbcdba4b3a671ce18"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af341928c41f42d5f39f8635d2b9fe2a6"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="af341928c41f42d5f39f8635d2b9fe2a6"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>spacer3</b> = &quot; &quot;</td></tr>
<tr class="separator:af341928c41f42d5f39f8635d2b9fe2a6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5a9b1911ac3eac594c92887c5856072b"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a5a9b1911ac3eac594c92887c5856072b"></a>
list&#160;</td><td class="memItemRight" valign="bottom"><b>LS_FLAGS_SHORT_REQUIRED</b> = [ &quot;-f&quot;, &quot;-s&quot;, &quot;-p&quot;, &quot;-m&quot;, &quot;-a&quot;, &quot;-r&quot;, &quot;-e&quot;, &quot;-M&quot;, &quot;-H&quot;, &quot;-P&quot;, &quot;-c&quot;, &quot;-l&quot;, &quot;-i&quot;, &quot;-n&quot;, &quot;-x&quot;, &quot;-g&quot;,&quot;-q&quot; ]</td></tr>
<tr class="separator:a5a9b1911ac3eac594c92887c5856072b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a14642564b63c70a5e681137460d5badc"><td class="memItemLeft" align="right" valign="top">list&#160;</td><td class="memItemRight" valign="bottom"><b>LS_FLAGS_LONG_REQUIRED</b></td></tr>
<tr class="separator:a14642564b63c70a5e681137460d5badc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0a7356eec35cf0a00376c0f86224de68"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a0a7356eec35cf0a00376c0f86224de68"></a>
list&#160;</td><td class="memItemRight" valign="bottom"><b>LS_ARGS_HELP_REQUIRED</b></td></tr>
<tr class="separator:a0a7356eec35cf0a00376c0f86224de68"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a539a82b86e32b85bc118dbcff1234775"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a539a82b86e32b85bc118dbcff1234775"></a>
list&#160;</td><td class="memItemRight" valign="bottom"><b>LS_FLAGS_SHORT_OPTIONAL</b> = [ &quot;-o&quot;, &quot;-d&quot;, &quot;-b&quot;, &quot;-j&quot; ]</td></tr>
<tr class="separator:a539a82b86e32b85bc118dbcff1234775"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae2160c0f729d746db7ff5bb9632fb341"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ae2160c0f729d746db7ff5bb9632fb341"></a>
list&#160;</td><td class="memItemRight" valign="bottom"><b>LS_FLAGS_LONG_OPTIONAL</b> = [ &quot;--processes&quot;, &quot;--mode&quot;, &quot;--nbneratio&quot;, &quot;--donbbiasadjust&quot; ]</td></tr>
<tr class="separator:ae2160c0f729d746db7ff5bb9632fb341"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a632aa31326a0a89aca1ad86c5be25e7e"><td class="memItemLeft" align="right" valign="top">list&#160;</td><td class="memItemRight" valign="bottom"><b>LS_ARGS_HELP_OPTIONAL</b></td></tr>
<tr class="separator:a632aa31326a0a89aca1ad86c5be25e7e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ababc6c288ef0a624a64b8985810e1b5f"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ababc6c288ef0a624a64b8985810e1b5f"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_GENEPOP_FILES</b> = 0</td></tr>
<tr class="separator:ababc6c288ef0a624a64b8985810e1b5f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a06a0d015e5eb0a2fec8d8e7eb7dc4bf9"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a06a0d015e5eb0a2fec8d8e7eb7dc4bf9"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_SAMPLE_SCHEME</b> = 1</td></tr>
<tr class="separator:a06a0d015e5eb0a2fec8d8e7eb7dc4bf9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a88ae07e2e3c54cc0f2688cdc154f83c1"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a88ae07e2e3c54cc0f2688cdc154f83c1"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_SAMPLE_VALUE_LIST</b> = 2</td></tr>
<tr class="separator:a88ae07e2e3c54cc0f2688cdc154f83c1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3517dce4472311d8528f667dab1860ff"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a3517dce4472311d8528f667dab1860ff"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_MIN_POP_SIZE</b> = 3</td></tr>
<tr class="separator:a3517dce4472311d8528f667dab1860ff"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8f8053d57b4d067177241b13514151ec"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a8f8053d57b4d067177241b13514151ec"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_MAX_POP_SIZE</b> = 4</td></tr>
<tr class="separator:a8f8053d57b4d067177241b13514151ec"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a64435079755b0280b7b1f75615299f36"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a64435079755b0280b7b1f75615299f36"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_POP_RANGE</b> = 5</td></tr>
<tr class="separator:a64435079755b0280b7b1f75615299f36"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2f828f14857a7d165360868bef410ab6"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a2f828f14857a7d165360868bef410ab6"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_MIN_ALLELE_FREQ</b> = 6</td></tr>
<tr class="separator:a2f828f14857a7d165360868bef410ab6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3ba4220bc12953d5cd890b34d2b180bc"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a3ba4220bc12953d5cd890b34d2b180bc"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_MONOGAMY</b> = 7</td></tr>
<tr class="separator:a3ba4220bc12953d5cd890b34d2b180bc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad3179624dfa236cf247b27257e7987c0"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ad3179624dfa236cf247b27257e7987c0"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_CHROMLOCI_FILE</b> = 8</td></tr>
<tr class="separator:ad3179624dfa236cf247b27257e7987c0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a16a71c0720d89bdd97ed607a6fc926e8"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a16a71c0720d89bdd97ed607a6fc926e8"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_ALLELE_PAIRING_SCHEME</b> = 9</td></tr>
<tr class="separator:a16a71c0720d89bdd97ed607a6fc926e8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3ad8615094dd5afecdf7178c9d751fbd"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a3ad8615094dd5afecdf7178c9d751fbd"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_REPLICATES</b> = 10</td></tr>
<tr class="separator:a3ad8615094dd5afecdf7178c9d751fbd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adf73af96ae717d362a2f698fab9c9f56"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="adf73af96ae717d362a2f698fab9c9f56"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_LOCI_SAMPLING_SCHEME</b> = 11</td></tr>
<tr class="separator:adf73af96ae717d362a2f698fab9c9f56"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab1b57f5b087a991524c5e39a19484c84"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ab1b57f5b087a991524c5e39a19484c84"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_LOCI_SCHEME_PARAM</b> = 12</td></tr>
<tr class="separator:ab1b57f5b087a991524c5e39a19484c84"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a70d8db7eadd4cdbee3917aabdf916f86"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a70d8db7eadd4cdbee3917aabdf916f86"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_LOCI_MIN_TOTAL</b> = 13</td></tr>
<tr class="separator:a70d8db7eadd4cdbee3917aabdf916f86"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a91522fc32292c83476e52be12d728c5e"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a91522fc32292c83476e52be12d728c5e"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_LOCI_MAX_TOTAL</b> = 14</td></tr>
<tr class="separator:a91522fc32292c83476e52be12d728c5e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5ef5232c3c6e5387a1ae241ed7e31325"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a5ef5232c3c6e5387a1ae241ed7e31325"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_LOCI_RANGE</b> = 15</td></tr>
<tr class="separator:a5ef5232c3c6e5387a1ae241ed7e31325"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adb1684d7e204e6ba686450a3ec24a5fa"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="adb1684d7e204e6ba686450a3ec24a5fa"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_LOCI_SAMPLE_REPLICATES</b> = 16</td></tr>
<tr class="separator:adb1684d7e204e6ba686450a3ec24a5fa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab46a408976ea7853ce3117837a23e977"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ab46a408976ea7853ce3117837a23e977"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_PROCESSES</b> = 17</td></tr>
<tr class="separator:ab46a408976ea7853ce3117837a23e977"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a909c695c3ada17903d970b1288c84302"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a909c695c3ada17903d970b1288c84302"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_DEBUG_MODE</b> = 18</td></tr>
<tr class="separator:a909c695c3ada17903d970b1288c84302"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a767eb1973ab4602c09395cc19cae5d03"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a767eb1973ab4602c09395cc19cae5d03"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_NBNE_RATIO</b> = 19</td></tr>
<tr class="separator:a767eb1973ab4602c09395cc19cae5d03"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abbdda8e36b0ddeb710eed554dc051058"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="abbdda8e36b0ddeb710eed554dc051058"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_DO_BIAS_ADJUST</b> = 20</td></tr>
<tr class="separator:abbdda8e36b0ddeb710eed554dc051058"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad82f566b349233fbf17d16521198a475"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ad82f566b349233fbf17d16521198a475"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_MAIN_OUTFILE</b> = 21</td></tr>
<tr class="separator:ad82f566b349233fbf17d16521198a475"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afb809c881e7ab5bd253e2019125dbef7"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="afb809c881e7ab5bd253e2019125dbef7"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_SECONDARY_OUTFILE</b> = 22</td></tr>
<tr class="separator:afb809c881e7ab5bd253e2019125dbef7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afbac2cd698e98608beebb84b196ed515"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="afbac2cd698e98608beebb84b196ed515"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_MULTIPROCESSING_EVENT</b> = 23</td></tr>
<tr class="separator:afbac2cd698e98608beebb84b196ed515"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a99b55a82d45508e951a1b5217a724c65"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a99b55a82d45508e951a1b5217a724c65"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_TEMPORARY_DIRECTORY</b> = 24</td></tr>
<tr class="separator:a99b55a82d45508e951a1b5217a724c65"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab5c12882dca27fee8a124b62c69b4331"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ab5c12882dca27fee8a124b62c69b4331"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_USE_GUI_MESSAGING</b> = 25</td></tr>
<tr class="separator:ab5c12882dca27fee8a124b62c69b4331"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af7ff167e4961fabfa7de810094a069c4"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="af7ff167e4961fabfa7de810094a069c4"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>IDX_LAST_CONSOLE_ARG</b> = IDX_DO_BIAS_ADJUST</td></tr>
<tr class="separator:af7ff167e4961fabfa7de810094a069c4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8cc4108d0e8f1f28708a86f2d97c788b"><td class="memItemLeft" align="right" valign="top">list&#160;</td><td class="memItemRight" valign="bottom"><b>HIDDEN_ARGS</b></td></tr>
<tr class="separator:a8cc4108d0e8f1f28708a86f2d97c788b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a152ba6e502315d9de22c1864713f044c"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a152ba6e502315d9de22c1864713f044c"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>DEFAULT_NUM_PROCESSES</b> = &quot;1&quot;</td></tr>
<tr class="separator:a152ba6e502315d9de22c1864713f044c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa4ee87ae3dd3f47bb0535b6658a2fcb7"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aa4ee87ae3dd3f47bb0535b6658a2fcb7"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>DEFAULT_DEBUG_MODE</b> = &quot;no_debug&quot;</td></tr>
<tr class="separator:aa4ee87ae3dd3f47bb0535b6658a2fcb7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac2972e35bf5aedc294a8384e846df4e2"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac2972e35bf5aedc294a8384e846df4e2"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>DEFAULT_NBNE_VAL</b> = &quot;None&quot;</td></tr>
<tr class="separator:ac2972e35bf5aedc294a8384e846df4e2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a24693d753431b98c4a3ce407f3f5bdd2"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a24693d753431b98c4a3ce407f3f5bdd2"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>MAX_GENEPOP_FILE_NAME_LENGTH</b> = 310</td></tr>
<tr class="separator:a24693d753431b98c4a3ce407f3f5bdd2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a17e9e437aef6bf7f4d74592ce2d7d9f7"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a17e9e437aef6bf7f4d74592ce2d7d9f7"></a>
float&#160;</td><td class="memItemRight" valign="bottom"><b>SECONDS_TO_SLEEP</b> = 1.00</td></tr>
<tr class="separator:a17e9e437aef6bf7f4d74592ce2d7d9f7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abd62126ec66523b4ade2773549c518f6"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="abd62126ec66523b4ade2773549c518f6"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>POOL_BATCH_SIZE</b> = 100</td></tr>
<tr class="separator:abd62126ec66523b4ade2773549c518f6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1f25e18b0f7554fef10298d8f72449ae"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a1f25e18b0f7554fef10298d8f72449ae"></a>
float&#160;</td><td class="memItemRight" valign="bottom"><b>PROPORTION_FILE_BYTES_USED_BY_OBJECT</b> = 0.12</td></tr>
<tr class="separator:a1f25e18b0f7554fef10298d8f72449ae"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2d869475f01666f05e5e66318e743ae7"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a2d869475f01666f05e5e66318e743ae7"></a>
float&#160;</td><td class="memItemRight" valign="bottom"><b>PROPORTION_GP_OBJECTS_PER_ADDITIONAL_PROCESS</b> = 0.3</td></tr>
<tr class="separator:a2d869475f01666f05e5e66318e743ae7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7cd39c355830e35a28cc3ccf25b1177a"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a7cd39c355830e35a28cc3ccf25b1177a"></a>
float&#160;</td><td class="memItemRight" valign="bottom"><b>PERC_AVAIL_RAM_TO_ACCESS</b> = 0.7</td></tr>
<tr class="separator:a7cd39c355830e35a28cc3ccf25b1177a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae48f3e45a71ea328667ece9f01d9fdfd"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ae48f3e45a71ea328667ece9f01d9fdfd"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>SAMPLE_BY_NONE</b> = &quot;none&quot;</td></tr>
<tr class="separator:ae48f3e45a71ea328667ece9f01d9fdfd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0c5458ee3e5678d22da1266132d484cc"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a0c5458ee3e5678d22da1266132d484cc"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>SAMPLE_BY_PERCENTAGES</b> = &quot;percent&quot;</td></tr>
<tr class="separator:a0c5458ee3e5678d22da1266132d484cc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1eea8a89b76613a384e16c8b62f1b471"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a1eea8a89b76613a384e16c8b62f1b471"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>SAMPLE_BY_REMOVAL</b> = &quot;remove&quot;</td></tr>
<tr class="separator:a1eea8a89b76613a384e16c8b62f1b471"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1bb68cd223590c754bb16e36e364fb5b"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a1bb68cd223590c754bb16e36e364fb5b"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>SAMPLE_BY_CRITERIA</b> = &quot;criteria&quot;</td></tr>
<tr class="separator:a1bb68cd223590c754bb16e36e364fb5b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a907b407a67eff111392d2641a818a113"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a907b407a67eff111392d2641a818a113"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>SAMPLE_BY_COHORT</b> = &quot;cohorts&quot;</td></tr>
<tr class="separator:a907b407a67eff111392d2641a818a113"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1c034d5a8aad5b74ba29944bb138cf97"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a1c034d5a8aad5b74ba29944bb138cf97"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>SAMPLE_BY_COHORT_PERCENTAGE</b> = &quot;cohortsperc&quot;</td></tr>
<tr class="separator:a1c034d5a8aad5b74ba29944bb138cf97"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac3dc6aee57275ddc87415ccd86b8eec1"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac3dc6aee57275ddc87415ccd86b8eec1"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>SAMPLE_BY_COHORT_COUNT</b> = &quot;cohortscount&quot;</td></tr>
<tr class="separator:ac3dc6aee57275ddc87415ccd86b8eec1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac5671dce1fa02116e8e640e43b971c07"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac5671dce1fa02116e8e640e43b971c07"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>SAMPLE_BY_COHORT_COUNT_MAX</b> = &quot;cohortsmax&quot;</td></tr>
<tr class="separator:ac5671dce1fa02116e8e640e43b971c07"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a039bcd35f71198ed1538209f7560bf6d"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a039bcd35f71198ed1538209f7560bf6d"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>SAMPLE_BY_RELATEDS</b> = &quot;relateds&quot;</td></tr>
<tr class="separator:a039bcd35f71198ed1538209f7560bf6d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a12890ea18077e7bdd6904a5907a4df8f"><td class="memItemLeft" align="right" valign="top">list&#160;</td><td class="memItemRight" valign="bottom"><b>SAMPLE_SCHEMES_NON_CRITERIA</b></td></tr>
<tr class="separator:a12890ea18077e7bdd6904a5907a4df8f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a87d495cee8e8a7d5337ab74972992bd7"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a87d495cee8e8a7d5337ab74972992bd7"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>SAMPLE_LOCI_SCHEME_NONE</b> = &quot;none&quot;</td></tr>
<tr class="separator:a87d495cee8e8a7d5337ab74972992bd7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a024063acb8c1389880757fd9fefbfd46"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a024063acb8c1389880757fd9fefbfd46"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>SAMPLE_LOCI_SCHEME_PERCENT</b> = &quot;percent&quot;</td></tr>
<tr class="separator:a024063acb8c1389880757fd9fefbfd46"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac8fcb8d0436a12aad84d95bc2a0668bb"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac8fcb8d0436a12aad84d95bc2a0668bb"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>SAMPLE_LOCI_CONSTANT_TOTAL</b> = &quot;total&quot;</td></tr>
<tr class="separator:ac8fcb8d0436a12aad84d95bc2a0668bb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa51cbc675141c881bda27289e9b0cc95"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aa51cbc675141c881bda27289e9b0cc95"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>OUTPUT_DELIMITER</b> = &quot;\t&quot;</td></tr>
<tr class="separator:aa51cbc675141c881bda27289e9b0cc95"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a29c2efad048604bbc94ce9574c01f3f7"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a29c2efad048604bbc94ce9574c01f3f7"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>ENDLINE_SEQ</b> = &quot;\n&quot;</td></tr>
<tr class="separator:a29c2efad048604bbc94ce9574c01f3f7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa0cdd94749afb2226db59ddea8a97a99"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aa0cdd94749afb2226db59ddea8a97a99"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>OUTPUT_ENDLINE</b> = &quot;\n&quot;</td></tr>
<tr class="separator:aa0cdd94749afb2226db59ddea8a97a99"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aec90386f8570d01014afe99b36c68e1d"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aec90386f8570d01014afe99b36c68e1d"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>REPLICATE_DELIMITER</b> = &quot;,&quot;</td></tr>
<tr class="separator:aec90386f8570d01014afe99b36c68e1d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa92ffa82812eab9e5e9c8121a412d3f7"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aa92ffa82812eab9e5e9c8121a412d3f7"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>SAMPLE_SCHEME_PARAM_DELIMITER</b> = &quot;,&quot;</td></tr>
<tr class="separator:aa92ffa82812eab9e5e9c8121a412d3f7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a55d46c674cb89e3ef4b0e8b6377e367b"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a55d46c674cb89e3ef4b0e8b6377e367b"></a>
list&#160;</td><td class="memItemRight" valign="bottom"><b>NE_ESTIMATOR_OUTPUT_FIELDS_TO_SKIP</b> = [ &quot;case_number&quot; ]</td></tr>
<tr class="separator:a55d46c674cb89e3ef4b0e8b6377e367b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab6551e3440f270ab0b8606075ae16774"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ab6551e3440f270ab0b8606075ae16774"></a>
list&#160;</td><td class="memItemRight" valign="bottom"><b>NE_ESTIMATOR_AND_LDNE_OUTPUT_FILE_TAGS</b> = [ &quot;NoDat.txt&quot;, &quot;xTp.txt&quot;, &quot;ne_run.txt&quot;, &quot;x.txt&quot; ]</td></tr>
<tr class="separator:ab6551e3440f270ab0b8606075ae16774"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aee36b33d32c1d5d3884d118fbfe876ee"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aee36b33d32c1d5d3884d118fbfe876ee"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>NE_ESTIMATOR_AND_LDNE_OUTPUT_FILES_GENERAL_TAG</b> = &quot;_g_[0-9]&quot;</td></tr>
<tr class="separator:aee36b33d32c1d5d3884d118fbfe876ee"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae781ea197c7e50e38e479ca334f8bc01"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ae781ea197c7e50e38e479ca334f8bc01"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>GLOB_INTERMEDIATE_GENEPOP_FILE_TAGS</b> = &quot;*_r_*_g_*&quot;</td></tr>
<tr class="separator:ae781ea197c7e50e38e479ca334f8bc01"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afc0811efe747ee446662813b09446329"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="afc0811efe747ee446662813b09446329"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>INPUT_FILE_DOT_CHAR_REPLACEMENT</b> = &quot;_&quot;</td></tr>
<tr class="separator:afc0811efe747ee446662813b09446329"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7c2814699a352b30869020a5235e308c"><td class="memItemLeft" align="right" valign="top">list&#160;</td><td class="memItemRight" valign="bottom"><b>MAIN_TABLE_RUN_INFO_COLS</b></td></tr>
<tr class="separator:a7c2814699a352b30869020a5235e308c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3867f2bdc93cf0f48d72c0961c0e3896"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a3867f2bdc93cf0f48d72c0961c0e3896"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>COL_NAME_NBNE_RATIO_FOR_BIAS_ADJUST</b> = &quot;nbne&quot;</td></tr>
<tr class="separator:a3867f2bdc93cf0f48d72c0961c0e3896"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a178b1b49a774e2e57f69dc0d7eb9fe55"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a178b1b49a774e2e57f69dc0d7eb9fe55"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>COL_NAME_BIAS_ADJUSTED_LDNE</b> = &quot;ne_est_adj&quot;</td></tr>
<tr class="separator:a178b1b49a774e2e57f69dc0d7eb9fe55"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a477ecffaa1689cc5adc0bc0e70417283"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a477ecffaa1689cc5adc0bc0e70417283"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>COL_NAME_NEESTIMATOR_NE_ESTIMATE</b> = &quot;est_ne&quot;</td></tr>
<tr class="separator:a477ecffaa1689cc5adc0bc0e70417283"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a562289b3a5ebf93d267830be30d6ad4a"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a562289b3a5ebf93d267830be30d6ad4a"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>COL_NAME_HET_VALUE</b> = &quot;mean_het&quot;</td></tr>
<tr class="separator:a562289b3a5ebf93d267830be30d6ad4a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8268e96ee501e6d9c3978400bf4ed93b"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a8268e96ee501e6d9c3978400bf4ed93b"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>COL_NAME_NEESTIMATOR_CI_HI</b> = &quot;95ci_high&quot;</td></tr>
<tr class="separator:a8268e96ee501e6d9c3978400bf4ed93b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5ad07082ce84a3992cdacd71fb12bbc3"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a5ad07082ce84a3992cdacd71fb12bbc3"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>COL_NAME_NEESTIMATOR_CI_LOW</b> = &quot;95ci_low&quot;</td></tr>
<tr class="separator:a5ad07082ce84a3992cdacd71fb12bbc3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a13262d5d7d1b51fb37748ae1a50212e1"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a13262d5d7d1b51fb37748ae1a50212e1"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>COL_NAME_BIAS_ADJUSTED_CI_HI</b> = &quot;95ci_high_adj&quot;</td></tr>
<tr class="separator:a13262d5d7d1b51fb37748ae1a50212e1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afb6fa3831af72ccc6d8f6f595dfcef9b"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="afb6fa3831af72ccc6d8f6f595dfcef9b"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>COL_NAME_BIAS_ADJUSTED_CI_LOW</b> = &quot;95ci_low_adj&quot;</td></tr>
<tr class="separator:afb6fa3831af72ccc6d8f6f595dfcef9b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae59293166cd73712f57ba570bbbf5181"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ae59293166cd73712f57ba570bbbf5181"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>RELTOL</b> = 1</td></tr>
<tr class="separator:ae59293166cd73712f57ba570bbbf5181"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a231cae7991dfbb05d74be7f39ab9a886"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a231cae7991dfbb05d74be7f39ab9a886"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>o_parser</b> = ap.ArgumentParser()</td></tr>
<tr class="separator:a231cae7991dfbb05d74be7f39ab9a886"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad7fb540484e35ce01a543be0e5a9873a"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ad7fb540484e35ce01a543be0e5a9873a"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>o_arglist</b> = o_parser.add_argument_group( &quot;args&quot; )</td></tr>
<tr class="separator:ad7fb540484e35ce01a543be0e5a9873a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab5ed34f7d30e99c312f34f9067b46270"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ab5ed34f7d30e99c312f34f9067b46270"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>i_total_nonopt</b> = len( LS_FLAGS_SHORT_REQUIRED )</td></tr>
<tr class="separator:ab5ed34f7d30e99c312f34f9067b46270"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a62895425b527b3551eb642025e8eb53f"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a62895425b527b3551eb642025e8eb53f"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>help</b></td></tr>
<tr class="separator:a62895425b527b3551eb642025e8eb53f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae5255e4a45aa0e2ca6d62f44274a6e5e"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ae5255e4a45aa0e2ca6d62f44274a6e5e"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>required</b></td></tr>
<tr class="separator:ae5255e4a45aa0e2ca6d62f44274a6e5e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad1acfacb657d121329b5b18df9bc7da1"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ad1acfacb657d121329b5b18df9bc7da1"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>i_total_opt</b> = len( LS_FLAGS_SHORT_OPTIONAL )</td></tr>
<tr class="separator:ad1acfacb657d121329b5b18df9bc7da1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0810336c5efd1ad9fd974f7836aeddee"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a0810336c5efd1ad9fd974f7836aeddee"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>o_args</b> = o_parser.parse_args()</td></tr>
<tr class="separator:a0810336c5efd1ad9fd974f7836aeddee"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a452df53c06de20036949d3d06108b4a3"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a452df53c06de20036949d3d06108b4a3"></a>
list&#160;</td><td class="memItemRight" valign="bottom"><b>ls_args_passed</b> = []</td></tr>
<tr class="separator:a452df53c06de20036949d3d06108b4a3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adb21e66371146f97cd21178485aea00d"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="adb21e66371146f97cd21178485aea00d"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>s_argname</b> = s_flag.replace( &quot;--&quot;, &quot;&quot; )</td></tr>
<tr class="separator:adb21e66371146f97cd21178485aea00d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a92cfb3a7527096737225d5116a9c24db"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a92cfb3a7527096737225d5116a9c24db"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>s_val</b> = getattr( o_args, s_argname )</td></tr>
<tr class="separator:a92cfb3a7527096737225d5116a9c24db"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aec8c1c15e13a62e8c6acd40ee73718db"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aec8c1c15e13a62e8c6acd40ee73718db"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>s_temp_dir</b> = pgut.get_temporary_directory()</td></tr>
<tr class="separator:aec8c1c15e13a62e8c6acd40ee73718db"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3bbf416c7a13027cd68661641872a3d3"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a3bbf416c7a13027cd68661641872a3d3"></a>
string&#160;</td><td class="memItemRight" valign="bottom"><b>s_msg</b> = &quot;In pgdriveneestimator main section, &quot;</td></tr>
<tr class="separator:a3bbf416c7a13027cd68661641872a3d3"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><pre class="fragment">Description

given 
    --&gt;a list of genepop files, 
    --&gt;a list of percentages of the total indiv
       to subsample
    --&gt; a number giving a minimum pop size 
       after subsampling
    --&gt;a minimum allele frequency (single float)
    --&gt; a number giving the total replicate NeEstimator runs 
to be executed on each file at each subsample percentage 

    subsamples and then runs the Ne estimator for each combination of 
    genepop_file,percentage,replicate_number. 
    
    As of Fri May 13 19:01:06 MDT 2016, still not sure
    how best to get the results to file or stream.

Mon Aug 15 20:04:00 MDT 2016
    revised so that now args list includes
    --&gt;a list of genepop files, 
    --&gt;one of [percent|remove] (names a subsampling scheme, to be enlarged)
    --&gt;a list of percentages, for percent samplin, or N-values for removing N
from the individuals    
    --&gt; a number giving a minimum pop size 
       after subsampling (fewer than the min count of indiv, and the pop is skipped)
    --&gt;a minimum allele frequency (single float)
    --&gt; a number giving the total replicate NeEstimator runs 
to be executed on each file at each subsample percentage 
    --&gt;optional the number of processes to create in performing the
estimates in parallele, on process for each population (as divided up inside a file,
or as given by a signle genepop file with only one population listed)
    
    Also revised to allow the improtation of this mod, so that the same operations
    can be intitiated by a call to "mymain".  The importation method allows for 
    passing in two file objects, open for writing.  The console method, that satisfies
    "if __name__== "__main__", does not allow the file object args and prints to stdout and stderr.

Wed Sep  7 18:12:26 MDT 2016
    revised so that now args list includes
    --&gt;a list of genepop files, 
    --&gt;one of [percent|remove|criteria] (names a subsampling scheme, to be enlarged)
    --&gt;a list of comma-delimited percentages, for percent samplin, or N-values for removing N
from the individuals, or, for criteria, a comma-delimited list consisting of:
        1. list of semicolong-delimited genepop indiv id fields (example "numberage;sex;parent")
        2. list of python types corresponding to each field (example "int;int;string")
        3. one or more test expressions using field names 
            surrounded by percentage signs.  For example, to include 
            2 criteria: "%age%&gt;3 and %age%&lt;6 and %sex% == 3","%parent% &lt;= "3992"
    --&gt; a number giving a minimum pop size 
       after subsampling (fewer than the min count of indiv, and the pop is skipped)
    --&gt;a hypnated pair of numbers giving the minimum and maximum i where
i gives the ith population listed in any one of the genepop files. For
example to run the estimator only on the 2nd through the 10th population,
the arg would be: "2-10" (all subject to minimum pop size).  
Using "all" for this arg will result in all pops being included.
    --&gt;a minimum allele frequency (single float)
    --&gt; a number giving the total replicate NeEstimator runs 
to be executed on each file at each subsample percentage 
    --&gt;optional the number of processes to create in performing the
estimates in parallele, on process for each population (as divided up inside a file,
or as given by a signle genepop file with only one population listed)

2016_09_24
    Revisions to add new sampling schemes (No-subsampling, and AgeStructureNe schemes to get
    cohorts and siblings (Relateds).  In addition to the percent, remove, and
    criteria schemes, we add "none", "cohorts", and "relateds" as options for arg 2.
    For arg 3 we now add, if cohorts is the second arg value, comma-delimited
    fields (see mod genepopfilesampler.py, class 
    GenepopFileSamplerParamsAgeStructureCohorts):
    --&gt; int, max individual age
    And, if "relateds" is the 2nd arg value (see mod genepopfilesampler.py, class
    GenepopFileSamplerIndividualsAgeStructureRelateds):
    1. float, percent of relateds each gen
    Also added max pop size as new arg, now arg #
2016_10_20
    Adding to Loci scheme "percent," to subsample a percentage
    of the loci (as circumscribed by existing params min loci 
    position and max loci position.
</pre> </div><h2 class="groupheader">Function Documentation</h2>
<a class="anchor" id="af69b7cf5fb237f79980c1d77c9134e6e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def agestrucne.pgdriveneestimator.add_to_set_of_calls_to_do_estimate </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>o_genepopfile</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>s_sample_scheme</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>f_min_allele_freq</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>b_monogamy</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>s_chromlocifile</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_allele_pairing_scheme</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_min_pop_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_max_pop_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_min_loci_position</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_max_loci_position</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>o_debug_mode</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>llv_args_each_process</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>IDX_NE_ESTIMATOR_OUTPUT_FIELDS_TO_SKIP</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>o_secondary_outfile</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>f_nbne_ratio</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>s_temporary_directory</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_genepop_file_count</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">This def creates ne-estimator caller object and adds it to list of args for a single call
to def do_estimate.  The call is then appended to llv_args_each_process
</pre> 
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l02457">2457</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>
<div class="fragment"><div class="line"><a name="l02457"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html#af69b7cf5fb237f79980c1d77c9134e6e"> 2457</a></span>&#160;                                            i_genepop_file_count ):</div><div class="line"><a name="l02458"></a><span class="lineno"> 2458</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;     </span></div><div class="line"><a name="l02459"></a><span class="lineno"> 2459</span>&#160;<span class="stringliteral">    This def creates ne-estimator caller object and adds it to list of args for a single call</span></div><div class="line"><a name="l02460"></a><span class="lineno"> 2460</span>&#160;<span class="stringliteral">    to def do_estimate.  The call is then appended to llv_args_each_process</span></div><div class="line"><a name="l02461"></a><span class="lineno"> 2461</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l02462"></a><span class="lineno"> 2462</span>&#160;</div><div class="line"><a name="l02463"></a><span class="lineno"> 2463</span>&#160;    <span class="comment">#all these GenepopFileManager objects should now have</span></div><div class="line"><a name="l02464"></a><span class="lineno"> 2464</span>&#160;    <span class="comment">#a single population sample name:</span></div><div class="line"><a name="l02465"></a><span class="lineno"> 2465</span>&#160;    s_population_sample_name=o_genepopfile.population_subsample_tags[ 0 ]</div><div class="line"><a name="l02466"></a><span class="lineno"> 2466</span>&#160;</div><div class="line"><a name="l02467"></a><span class="lineno"> 2467</span>&#160;    ls_indiv_sample_names=o_genepopfile.indiv_subsample_tags</div><div class="line"><a name="l02468"></a><span class="lineno"> 2468</span>&#160;</div><div class="line"><a name="l02469"></a><span class="lineno"> 2469</span>&#160;    li_population_numbers=o_genepopfile.getListPopulationNumbers( s_population_sample_name )</div><div class="line"><a name="l02470"></a><span class="lineno"> 2470</span>&#160;    <span class="comment">#for each sample (i.e. for each replicate of a sampling with a given set of params:</span></div><div class="line"><a name="l02471"></a><span class="lineno"> 2471</span>&#160;</div><div class="line"><a name="l02472"></a><span class="lineno"> 2472</span>&#160;    <span class="comment">#Added this to help create a unique, temporary gp file name</span></div><div class="line"><a name="l02473"></a><span class="lineno"> 2473</span>&#160;    <span class="comment">#using file number individ sample, and loci sample counts:</span></div><div class="line"><a name="l02474"></a><span class="lineno"> 2474</span>&#160;    i_individ_sample_count=0</div><div class="line"><a name="l02475"></a><span class="lineno"> 2475</span>&#160;</div><div class="line"><a name="l02476"></a><span class="lineno"> 2476</span>&#160;    <span class="keywordflow">for</span> s_indiv_sample <span class="keywordflow">in</span> ls_indiv_sample_names:</div><div class="line"><a name="l02477"></a><span class="lineno"> 2477</span>&#160;</div><div class="line"><a name="l02478"></a><span class="lineno"> 2478</span>&#160;        i_individ_sample_count+=1</div><div class="line"><a name="l02479"></a><span class="lineno"> 2479</span>&#160;</div><div class="line"><a name="l02480"></a><span class="lineno"> 2480</span>&#160;        <span class="comment">#we also check that at least one pop has at least one individual</span></div><div class="line"><a name="l02481"></a><span class="lineno"> 2481</span>&#160;        <span class="comment">#listed for this subsample</span></div><div class="line"><a name="l02482"></a><span class="lineno"> 2482</span>&#160;        li_indiv_counts_for_this_sample=o_genepopfile.getIndividualCounts( s_indiv_subsample_tag=s_indiv_sample )</div><div class="line"><a name="l02483"></a><span class="lineno"> 2483</span>&#160;</div><div class="line"><a name="l02484"></a><span class="lineno"> 2484</span>&#160;        <span class="keywordflow">if</span> sum( li_indiv_counts_for_this_sample ) == 0:</div><div class="line"><a name="l02485"></a><span class="lineno"> 2485</span>&#160;            s_sample_value, i_replicate_count=get_sample_val_and_rep_number_from_sample_name( \</div><div class="line"><a name="l02486"></a><span class="lineno"> 2486</span>&#160;                                            s_indiv_sample, s_sample_scheme )</div><div class="line"><a name="l02487"></a><span class="lineno"> 2487</span>&#160;            <span class="comment">#we only want to report the first replicate</span></div><div class="line"><a name="l02488"></a><span class="lineno"> 2488</span>&#160;            <span class="comment">#(by inference of course all replicates will be skipped):</span></div><div class="line"><a name="l02489"></a><span class="lineno"> 2489</span>&#160;            <span class="keywordflow">if</span> i_replicate_count == 0:</div><div class="line"><a name="l02490"></a><span class="lineno"> 2490</span>&#160;                s_msg=<span class="stringliteral">&quot;Skipping samples from: &quot;</span> \</div><div class="line"><a name="l02491"></a><span class="lineno"> 2491</span>&#160;                        + o_genepopfile.original_file_name \</div><div class="line"><a name="l02492"></a><span class="lineno"> 2492</span>&#160;                        + <span class="stringliteral">&quot;.  In pgdriveneestimator.py, all populations &quot;</span> \</div><div class="line"><a name="l02493"></a><span class="lineno"> 2493</span>&#160;                        + <span class="stringliteral">&quot;have zero individuals when sampled using &quot;</span> \</div><div class="line"><a name="l02494"></a><span class="lineno"> 2494</span>&#160;                        + <span class="stringliteral">&quot;scheme: &quot;</span> + s_sample_scheme \</div><div class="line"><a name="l02495"></a><span class="lineno"> 2495</span>&#160;                        + <span class="stringliteral">&quot;, and with sample parameter, &quot;</span> + s_sample_value + <span class="stringliteral">&quot;.&quot;</span></div><div class="line"><a name="l02496"></a><span class="lineno"> 2496</span>&#160;                o_secondary_outfile.write( s_msg + <span class="stringliteral">&quot;\n&quot;</span> )</div><div class="line"><a name="l02497"></a><span class="lineno"> 2497</span>&#160;            <span class="comment">#end if first replicate, report</span></div><div class="line"><a name="l02498"></a><span class="lineno"> 2498</span>&#160;            <span class="keywordflow">continue</span></div><div class="line"><a name="l02499"></a><span class="lineno"> 2499</span>&#160;        <span class="comment">#end skip this subsample if no pops have any individuals</span></div><div class="line"><a name="l02500"></a><span class="lineno"> 2500</span>&#160;</div><div class="line"><a name="l02501"></a><span class="lineno"> 2501</span>&#160;        <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l02502"></a><span class="lineno"> 2502</span>&#160;<span class="stringliteral">        For percent and remove-N schemes, the min/max pop size params are not </span></div><div class="line"><a name="l02503"></a><span class="lineno"> 2503</span>&#160;<span class="stringliteral">        enforced in the sampler, and, as of 2016_09_27, are to be applied</span></div><div class="line"><a name="l02504"></a><span class="lineno"> 2504</span>&#160;<span class="stringliteral">        before sampling.  Thus, for these scheme we skip populations if their </span></div><div class="line"><a name="l02505"></a><span class="lineno"> 2505</span>&#160;<span class="stringliteral">        original, un-subsampled size is outside the pop size range.  Other schemes</span></div><div class="line"><a name="l02506"></a><span class="lineno"> 2506</span>&#160;<span class="stringliteral">        should have applied the criteria when subsampling.  For the latter, this</span></div><div class="line"><a name="l02507"></a><span class="lineno"> 2507</span>&#160;<span class="stringliteral">        would not be an appropriate test::</span></div><div class="line"><a name="l02508"></a><span class="lineno"> 2508</span>&#160;<span class="stringliteral">        &#39;&#39;&#39;</span></div><div class="line"><a name="l02509"></a><span class="lineno"> 2509</span>&#160;</div><div class="line"><a name="l02510"></a><span class="lineno"> 2510</span>&#160;        li_pops_with_invalid_size=[]</div><div class="line"><a name="l02511"></a><span class="lineno"> 2511</span>&#160;        </div><div class="line"><a name="l02512"></a><span class="lineno"> 2512</span>&#160;        <span class="keywordflow">if</span> s_sample_scheme <span class="keywordflow">in</span> SAMPLE_SCHEMES_NON_CRITERIA:</div><div class="line"><a name="l02513"></a><span class="lineno"> 2513</span>&#160;            li_pops_with_invalid_size=<a class="code" href="namespaceagestrucne_1_1pgdriveneestimator.html#abd76f8aa32c542ee3e90799c0e7b166a">get_pops_in_file_outside_valid_size_range</a>( o_genepopfile, </div><div class="line"><a name="l02514"></a><span class="lineno"> 2514</span>&#160;                                                                                i_min_pop_size,</div><div class="line"><a name="l02515"></a><span class="lineno"> 2515</span>&#160;                                                                                i_max_pop_size )</div><div class="line"><a name="l02516"></a><span class="lineno"> 2516</span>&#160;        <span class="comment">#end if we need to check original (non-sampled) pop sizes</span></div><div class="line"><a name="l02517"></a><span class="lineno"> 2517</span>&#160;</div><div class="line"><a name="l02518"></a><span class="lineno"> 2518</span>&#160;        ls_loci_subsample_tags_associated_with_this_indiv_sample= \</div><div class="line"><a name="l02519"></a><span class="lineno"> 2519</span>&#160;                get_loci_subsample_tags_for_this_indiv_subsample( o_genepopfile,</div><div class="line"><a name="l02520"></a><span class="lineno"> 2520</span>&#160;                                                                    s_indiv_sample )</div><div class="line"><a name="l02521"></a><span class="lineno"> 2521</span>&#160;    </div><div class="line"><a name="l02522"></a><span class="lineno"> 2522</span>&#160;        i_loci_subsample_count=0</div><div class="line"><a name="l02523"></a><span class="lineno"> 2523</span>&#160;</div><div class="line"><a name="l02524"></a><span class="lineno"> 2524</span>&#160;        <span class="keywordflow">for</span> s_loci_subsample_tag <span class="keywordflow">in</span> ls_loci_subsample_tags_associated_with_this_indiv_sample:</div><div class="line"><a name="l02525"></a><span class="lineno"> 2525</span>&#160;</div><div class="line"><a name="l02526"></a><span class="lineno"> 2526</span>&#160;            <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l02527"></a><span class="lineno"> 2527</span>&#160;<span class="stringliteral">            To most-efficiently use multi-processes, that is, to assign many</span></div><div class="line"><a name="l02528"></a><span class="lineno"> 2528</span>&#160;<span class="stringliteral">            processes a single file with many populations, we do an</span></div><div class="line"><a name="l02529"></a><span class="lineno"> 2529</span>&#160;<span class="stringliteral">            estimate on each pop in the file separately, and divvy</span></div><div class="line"><a name="l02530"></a><span class="lineno"> 2530</span>&#160;<span class="stringliteral">            the pops up among processes:</span></div><div class="line"><a name="l02531"></a><span class="lineno"> 2531</span>&#160;<span class="stringliteral">            &#39;&#39;&#39;</span></div><div class="line"><a name="l02532"></a><span class="lineno"> 2532</span>&#160;            <span class="keywordflow">for</span> i_population_number <span class="keywordflow">in</span> li_population_numbers:</div><div class="line"><a name="l02533"></a><span class="lineno"> 2533</span>&#160;</div><div class="line"><a name="l02534"></a><span class="lineno"> 2534</span>&#160;                i_loci_subsample_count+=1</div><div class="line"><a name="l02535"></a><span class="lineno"> 2535</span>&#160;</div><div class="line"><a name="l02536"></a><span class="lineno"> 2536</span>&#160;                <span class="keywordflow">if</span> i_population_number <span class="keywordflow">in</span> li_pops_with_invalid_size:</div><div class="line"><a name="l02537"></a><span class="lineno"> 2537</span>&#160;                    s_msg= <span class="stringliteral">&quot;In pgdriveneestimator.py, def add_to_set_of_calls_to_do_estimate, &quot;</span> \</div><div class="line"><a name="l02538"></a><span class="lineno"> 2538</span>&#160;                                    + <span class="stringliteral">&quot; in file, &quot;</span> + o_genepopfile.original_file_name \</div><div class="line"><a name="l02539"></a><span class="lineno"> 2539</span>&#160;                                    + <span class="stringliteral">&quot;, skipping pop number, &quot;</span>  + str( i_population_number ) \</div><div class="line"><a name="l02540"></a><span class="lineno"> 2540</span>&#160;                                    + <span class="stringliteral">&quot;.  It has an indiv count outside of valid &quot;</span> \</div><div class="line"><a name="l02541"></a><span class="lineno"> 2541</span>&#160;                                    + <span class="stringliteral">&quot;range, &quot;</span> + str( i_min_pop_size ) \</div><div class="line"><a name="l02542"></a><span class="lineno"> 2542</span>&#160;                                    + <span class="stringliteral">&quot; - &quot;</span> + str( i_max_pop_size ) + <span class="stringliteral">&quot;.&quot;</span></div><div class="line"><a name="l02543"></a><span class="lineno"> 2543</span>&#160;                    o_secondary_outfile.write( s_msg + <span class="stringliteral">&quot;\n&quot;</span> )</div><div class="line"><a name="l02544"></a><span class="lineno"> 2544</span>&#160;                    <span class="keywordflow">continue</span></div><div class="line"><a name="l02545"></a><span class="lineno"> 2545</span>&#160;                <span class="comment">#end if i_population_number is</span></div><div class="line"><a name="l02546"></a><span class="lineno"> 2546</span>&#160;</div><div class="line"><a name="l02547"></a><span class="lineno"> 2547</span>&#160;                <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l02548"></a><span class="lineno"> 2548</span>&#160;<span class="stringliteral">                We send to def do_estimate the genepop file object, o_genepopfile, and the subsample tag</span></div><div class="line"><a name="l02549"></a><span class="lineno"> 2549</span>&#160;<span class="stringliteral">                that identifies the correct subsample calculated and stored in the object, so we can write</span></div><div class="line"><a name="l02550"></a><span class="lineno"> 2550</span>&#160;<span class="stringliteral">                each subsample genepop file in the same process that will use it as input to the the estimator.</span></div><div class="line"><a name="l02551"></a><span class="lineno"> 2551</span>&#160;<span class="stringliteral">                This should limit the number of input files existing concurrenty in the directory to the </span></div><div class="line"><a name="l02552"></a><span class="lineno"> 2552</span>&#160;<span class="stringliteral">                number of processes in use.  We also send allele freq, and the original genepop file name </span></div><div class="line"><a name="l02553"></a><span class="lineno"> 2553</span>&#160;<span class="stringliteral">                inside the genepop file manager object):</span></div><div class="line"><a name="l02554"></a><span class="lineno"> 2554</span>&#160;<span class="stringliteral">                &#39;&#39;&#39;</span></div><div class="line"><a name="l02555"></a><span class="lineno"> 2555</span>&#160;                s_sample_value, s_replicate_number = \</div><div class="line"><a name="l02556"></a><span class="lineno"> 2556</span>&#160;                        get_sample_val_and_rep_number_from_sample_name( s_indiv_sample, s_sample_scheme )</div><div class="line"><a name="l02557"></a><span class="lineno"> 2557</span>&#160;</div><div class="line"><a name="l02558"></a><span class="lineno"> 2558</span>&#160;                s_loci_sample_value, s_loci_replicate_number = \</div><div class="line"><a name="l02559"></a><span class="lineno"> 2559</span>&#160;                        <a class="code" href="namespaceagestrucne_1_1pgdriveneestimator.html#a6eea3cf1501fe02f36abfa30d6b4b7e3">get_loci_sample_val_and_rep_number_from_loci_sample_tag</a>( s_loci_subsample_tag )</div><div class="line"><a name="l02560"></a><span class="lineno"> 2560</span>&#160;                </div><div class="line"><a name="l02561"></a><span class="lineno"> 2561</span>&#160;</div><div class="line"><a name="l02562"></a><span class="lineno"> 2562</span>&#160;                i_tot_indivs_this_subsample=o_genepopfile.getIndividualCount( i_population_number, </div><div class="line"><a name="l02563"></a><span class="lineno"> 2563</span>&#160;                                                                                        s_indiv_sample )</div><div class="line"><a name="l02564"></a><span class="lineno"> 2564</span>&#160;</div><div class="line"><a name="l02565"></a><span class="lineno"> 2565</span>&#160;                <span class="keywordflow">if</span> i_tot_indivs_this_subsample == 0:</div><div class="line"><a name="l02566"></a><span class="lineno"> 2566</span>&#160;</div><div class="line"><a name="l02567"></a><span class="lineno"> 2567</span>&#160;                    s_msg= <span class="stringliteral">&quot;In pgdriveneestimator.py, def add_to_set_of_calls_to_do_estimate, &quot;</span> \</div><div class="line"><a name="l02568"></a><span class="lineno"> 2568</span>&#160;                                    + <span class="stringliteral">&quot; no estimation was done on pop section number &quot;</span>  \</div><div class="line"><a name="l02569"></a><span class="lineno"> 2569</span>&#160;                                    + str( i_population_number ) \</div><div class="line"><a name="l02570"></a><span class="lineno"> 2570</span>&#160;                                    + <span class="stringliteral">&quot; in file, &quot;</span> + o_genepopfile.original_file_name \</div><div class="line"><a name="l02571"></a><span class="lineno"> 2571</span>&#160;                                    + <span class="stringliteral">&quot;.  The parameters for population or loci &quot;</span> \</div><div class="line"><a name="l02572"></a><span class="lineno"> 2572</span>&#160;                                    + <span class="stringliteral">&quot;resulted in no individuals being sampled.&quot;</span></div><div class="line"><a name="l02573"></a><span class="lineno"> 2573</span>&#160;</div><div class="line"><a name="l02574"></a><span class="lineno"> 2574</span>&#160;                    o_secondary_outfile.write( s_msg + <span class="stringliteral">&quot;\n&quot;</span> )</div><div class="line"><a name="l02575"></a><span class="lineno"> 2575</span>&#160;                    <span class="keywordflow">continue</span></div><div class="line"><a name="l02576"></a><span class="lineno"> 2576</span>&#160;                <span class="comment"># end if this sample for this pop has zero individuals</span></div><div class="line"><a name="l02577"></a><span class="lineno"> 2577</span>&#160;</div><div class="line"><a name="l02578"></a><span class="lineno"> 2578</span>&#160;                <span class="comment">#for downstream analysis, we get the census of total indiv</span></div><div class="line"><a name="l02579"></a><span class="lineno"> 2579</span>&#160;                <span class="comment">#non-sampled, for this pop section:</span></div><div class="line"><a name="l02580"></a><span class="lineno"> 2580</span>&#160;                s_census=str( o_genepopfile.getIndividualCount( i_population_number ) )</div><div class="line"><a name="l02581"></a><span class="lineno"> 2581</span>&#160;</div><div class="line"><a name="l02582"></a><span class="lineno"> 2582</span>&#160;                <span class="comment">#we make population subsample list consisting of only this populations number:</span></div><div class="line"><a name="l02583"></a><span class="lineno"> 2583</span>&#160;                s_this_pop_number=str( i_population_number ) </div><div class="line"><a name="l02584"></a><span class="lineno"> 2584</span>&#160;</div><div class="line"><a name="l02585"></a><span class="lineno"> 2585</span>&#160;                o_genepopfile.subsamplePopulationsByList( [ i_population_number ], s_this_pop_number )</div><div class="line"><a name="l02586"></a><span class="lineno"> 2586</span>&#160;</div><div class="line"><a name="l02587"></a><span class="lineno"> 2587</span>&#160;                <span class="keywordflow">if</span> o_debug_mode.isSet( DebugMode.PRINT_REPLICATE_SELECTIONS ):</div><div class="line"><a name="l02588"></a><span class="lineno"> 2588</span>&#160;                    print_test_list_replicate_selection_indices( o_genepopfile, </div><div class="line"><a name="l02589"></a><span class="lineno"> 2589</span>&#160;                                            s_sample_scheme,    </div><div class="line"><a name="l02590"></a><span class="lineno"> 2590</span>&#160;                                            s_indiv_sample, </div><div class="line"><a name="l02591"></a><span class="lineno"> 2591</span>&#160;                                            s_loci_sample_value,</div><div class="line"><a name="l02592"></a><span class="lineno"> 2592</span>&#160;                                            s_loci_replicate_number,</div><div class="line"><a name="l02593"></a><span class="lineno"> 2593</span>&#160;                                            o_secondary_outfile, </div><div class="line"><a name="l02594"></a><span class="lineno"> 2594</span>&#160;                                            s_population_subsample_tag=s_this_pop_number )</div><div class="line"><a name="l02595"></a><span class="lineno"> 2595</span>&#160;                <span class="comment">#end if we&#39;re testing, print </span></div><div class="line"><a name="l02596"></a><span class="lineno"> 2596</span>&#160;</div><div class="line"><a name="l02597"></a><span class="lineno"> 2597</span>&#160;                <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l02598"></a><span class="lineno"> 2598</span>&#160;<span class="stringliteral">                2017_04_03.  A Windows limit on path length motivates new, shorter names</span></div><div class="line"><a name="l02599"></a><span class="lineno"> 2599</span>&#160;<span class="stringliteral">                for the intermediate genepop files.  Now we simply use the i_genepop_file_count</span></div><div class="line"><a name="l02600"></a><span class="lineno"> 2600</span>&#160;<span class="stringliteral">                number passed to this def from def drive_estimator, and the subsample count</span></div><div class="line"><a name="l02601"></a><span class="lineno"> 2601</span>&#160;<span class="stringliteral">                to get a uniq identifier for this genepop file vs. all others in this run.</span></div><div class="line"><a name="l02602"></a><span class="lineno"> 2602</span>&#160;<span class="stringliteral">                For now we retain the original code remm&#39;d out, and construct the shorter</span></div><div class="line"><a name="l02603"></a><span class="lineno"> 2603</span>&#160;<span class="stringliteral">                name in a new def:</span></div><div class="line"><a name="l02604"></a><span class="lineno"> 2604</span>&#160;<span class="stringliteral">                &#39;&#39;&#39;</span></div><div class="line"><a name="l02605"></a><span class="lineno"> 2605</span>&#160;                <span class="comment">#in naming a the genepop file for the subsampled pop,</span></div><div class="line"><a name="l02606"></a><span class="lineno"> 2606</span>&#160;                <span class="comment">#some processing is needed to avoid NeEstimator file naming errors, and</span></div><div class="line"><a name="l02607"></a><span class="lineno"> 2607</span>&#160;                <span class="comment">#path mangling, so we made a separate def:</span></div><div class="line"><a name="l02608"></a><span class="lineno"> 2608</span>&#160;<span class="comment">#               s_genepop_file_subsample=get_subsample_genepop_file_name( o_genepopfile.original_file_name, </span></div><div class="line"><a name="l02609"></a><span class="lineno"> 2609</span>&#160;<span class="comment">#                                                       s_loci_subsample_tag, </span></div><div class="line"><a name="l02610"></a><span class="lineno"> 2610</span>&#160;<span class="comment">#                                                       s_this_pop_number,</span></div><div class="line"><a name="l02611"></a><span class="lineno"> 2611</span>&#160;<span class="comment">#                                                       s_temporary_directory )</span></div><div class="line"><a name="l02612"></a><span class="lineno"> 2612</span>&#160;                </div><div class="line"><a name="l02613"></a><span class="lineno"> 2613</span>&#160;                s_genepop_file_subsample=<a class="code" href="namespaceagestrucne_1_1pgdriveneestimator.html#a2ea5983385f4ef857bcf935d904c59b1">make_subsample_genepop_file_name</a>( o_genepopfile.original_file_name,</div><div class="line"><a name="l02614"></a><span class="lineno"> 2614</span>&#160;                                                                                            s_temporary_directory, </div><div class="line"><a name="l02615"></a><span class="lineno"> 2615</span>&#160;                                                                                                i_genepop_file_count, </div><div class="line"><a name="l02616"></a><span class="lineno"> 2616</span>&#160;                                                                                                    i_individ_sample_count,</div><div class="line"><a name="l02617"></a><span class="lineno"> 2617</span>&#160;                                                                                                        i_loci_subsample_count )</div><div class="line"><a name="l02618"></a><span class="lineno"> 2618</span>&#160;</div><div class="line"><a name="l02619"></a><span class="lineno"> 2619</span>&#160;                s_run_output_file=s_genepop_file_subsample + <span class="stringliteral">&quot;_ne_run.txt&quot;</span></div><div class="line"><a name="l02620"></a><span class="lineno"> 2620</span>&#160;</div><div class="line"><a name="l02621"></a><span class="lineno"> 2621</span>&#160;                o_neinput=pgin.PGInputNeEstimator( s_genepop_file_subsample )</div><div class="line"><a name="l02622"></a><span class="lineno"> 2622</span>&#160;                </div><div class="line"><a name="l02623"></a><span class="lineno"> 2623</span>&#160;                <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l02624"></a><span class="lineno"> 2624</span>&#160;<span class="stringliteral">                2018_03_15.  Note that we have added a  new parameter &quot;monogamy&quot;, to </span></div><div class="line"><a name="l02625"></a><span class="lineno"> 2625</span>&#160;<span class="stringliteral">                be passed to LDNe2 (see pginputneestimator class ).</span></div><div class="line"><a name="l02626"></a><span class="lineno"> 2626</span>&#160;<span class="stringliteral">                &#39;&#39;&#39;</span></div><div class="line"><a name="l02627"></a><span class="lineno"> 2627</span>&#160;                o_neinput.run_params={ <span class="stringliteral">&quot;crits&quot;</span>:[ f_min_allele_freq ] , <span class="stringliteral">&quot;monogamy&quot;</span>: b_monogamy }</div><div class="line"><a name="l02628"></a><span class="lineno"> 2628</span>&#160;</div><div class="line"><a name="l02629"></a><span class="lineno"> 2629</span>&#160;                <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l02630"></a><span class="lineno"> 2630</span>&#160;<span class="stringliteral">                2018_04_28.  New parameter used by LDNe2.</span></div><div class="line"><a name="l02631"></a><span class="lineno"> 2631</span>&#160;<span class="stringliteral">                &#39;&#39;&#39;</span></div><div class="line"><a name="l02632"></a><span class="lineno"> 2632</span>&#160;                <span class="keywordflow">if</span> ESTIMATOR_TO_USE==LDNE2:</div><div class="line"><a name="l02633"></a><span class="lineno"> 2633</span>&#160;                    o_neinput.ldne2_only_params= {<span class="stringliteral">&quot;chromlocifile&quot;</span>:s_chromlocifile, </div><div class="line"><a name="l02634"></a><span class="lineno"> 2634</span>&#160;                                            <span class="stringliteral">&quot;allele_pairing_scheme&quot;</span>:i_allele_pairing_scheme }</div><div class="line"><a name="l02635"></a><span class="lineno"> 2635</span>&#160;                <span class="comment">#end if we&#39;re using ldne2, add the chromlocifile value</span></div><div class="line"><a name="l02636"></a><span class="lineno"> 2636</span>&#160;</div><div class="line"><a name="l02637"></a><span class="lineno"> 2637</span>&#160;                o_neoutput=pgout.PGOutputNeEstimator( s_genepop_file_subsample, </div><div class="line"><a name="l02638"></a><span class="lineno"> 2638</span>&#160;                                                        s_run_output_file, </div><div class="line"><a name="l02639"></a><span class="lineno"> 2639</span>&#160;                                                        s_estimator_to_use=ESTIMATOR_TO_USE )   </div><div class="line"><a name="l02640"></a><span class="lineno"> 2640</span>&#160;                </div><div class="line"><a name="l02641"></a><span class="lineno"> 2641</span>&#160;                o_ne_estimator=pgne.PGOpNeEstimator( o_neinput, </div><div class="line"><a name="l02642"></a><span class="lineno"> 2642</span>&#160;                                                        o_neoutput, </div><div class="line"><a name="l02643"></a><span class="lineno"> 2643</span>&#160;                                                        s_estimator_name=ESTIMATOR_TO_USE,</div><div class="line"><a name="l02644"></a><span class="lineno"> 2644</span>&#160;                                                        s_parent_dir_for_workspace=s_temporary_directory ) </div><div class="line"><a name="l02645"></a><span class="lineno"> 2645</span>&#160;                </div><div class="line"><a name="l02646"></a><span class="lineno"> 2646</span>&#160;                <span class="keywordflow">if</span> ESTIMATOR_TO_USE == LDNE2 <span class="keywordflow">and</span> PATH_TO_LDNE2 <span class="keywordflow">is</span> <span class="keywordflow">not</span> <span class="keywordtype">None</span>:</div><div class="line"><a name="l02647"></a><span class="lineno"> 2647</span>&#160;                    o_ne_estimator.ldne_path=PATH_TO_LDNE2</div><div class="line"><a name="l02648"></a><span class="lineno"> 2648</span>&#160;                <span class="comment">#end if we are using LDNe2 and we have a path to the executable</span></div><div class="line"><a name="l02649"></a><span class="lineno"> 2649</span>&#160;</div><div class="line"><a name="l02650"></a><span class="lineno"> 2650</span>&#160;                lv_these_args = [ o_genepopfile,  </div><div class="line"><a name="l02651"></a><span class="lineno"> 2651</span>&#160;                                    o_ne_estimator, </div><div class="line"><a name="l02652"></a><span class="lineno"> 2652</span>&#160;                                    s_sample_value, </div><div class="line"><a name="l02653"></a><span class="lineno"> 2653</span>&#160;                                    s_loci_sample_value,</div><div class="line"><a name="l02654"></a><span class="lineno"> 2654</span>&#160;                                    f_min_allele_freq, </div><div class="line"><a name="l02655"></a><span class="lineno"> 2655</span>&#160;                                    b_monogamy,</div><div class="line"><a name="l02656"></a><span class="lineno"> 2656</span>&#160;                                    s_indiv_sample, </div><div class="line"><a name="l02657"></a><span class="lineno"> 2657</span>&#160;                                    s_loci_subsample_tag,</div><div class="line"><a name="l02658"></a><span class="lineno"> 2658</span>&#160;                                    s_this_pop_number, </div><div class="line"><a name="l02659"></a><span class="lineno"> 2659</span>&#160;                                    s_census,</div><div class="line"><a name="l02660"></a><span class="lineno"> 2660</span>&#160;                                    s_replicate_number, </div><div class="line"><a name="l02661"></a><span class="lineno"> 2661</span>&#160;                                    s_loci_replicate_number,</div><div class="line"><a name="l02662"></a><span class="lineno"> 2662</span>&#160;                                    o_debug_mode,</div><div class="line"><a name="l02663"></a><span class="lineno"> 2663</span>&#160;                                    IDX_NE_ESTIMATOR_OUTPUT_FIELDS_TO_SKIP,</div><div class="line"><a name="l02664"></a><span class="lineno"> 2664</span>&#160;                                    f_nbne_ratio,</div><div class="line"><a name="l02665"></a><span class="lineno"> 2665</span>&#160;                                    i_min_loci_position,</div><div class="line"><a name="l02666"></a><span class="lineno"> 2666</span>&#160;                                    i_max_loci_position ]</div><div class="line"><a name="l02667"></a><span class="lineno"> 2667</span>&#160;</div><div class="line"><a name="l02668"></a><span class="lineno"> 2668</span>&#160;                llv_args_each_process.append( lv_these_args )</div><div class="line"><a name="l02669"></a><span class="lineno"> 2669</span>&#160;            <span class="comment">#end for each population</span></div><div class="line"><a name="l02670"></a><span class="lineno"> 2670</span>&#160;        <span class="comment">#end for each loci sample name assoc with this indiv sample name</span></div><div class="line"><a name="l02671"></a><span class="lineno"> 2671</span>&#160;    <span class="comment">#end for each indiv sample name</span></div><div class="line"><a name="l02672"></a><span class="lineno"> 2672</span>&#160;    <span class="keywordflow">return</span></div><div class="line"><a name="l02673"></a><span class="lineno"> 2673</span>&#160;<span class="comment">#end add_to_set_of_calls_to_do_estimate</span></div><div class="line"><a name="l02674"></a><span class="lineno"> 2674</span>&#160;</div><div class="ttc" id="namespaceagestrucne_1_1pgdriveneestimator_html_a6eea3cf1501fe02f36abfa30d6b4b7e3"><div class="ttname"><a href="namespaceagestrucne_1_1pgdriveneestimator.html#a6eea3cf1501fe02f36abfa30d6b4b7e3">agestrucne.pgdriveneestimator.get_loci_sample_val_and_rep_number_from_loci_sample_tag</a></div><div class="ttdeci">def get_loci_sample_val_and_rep_number_from_loci_sample_tag(s_loci_subsample_tag)</div><div class="ttdef"><b>Definition:</b> <a href="pgdriveneestimator_8py_source.html#l02675">pgdriveneestimator.py:2675</a></div></div>
<div class="ttc" id="namespaceagestrucne_1_1pgdriveneestimator_html_abd76f8aa32c542ee3e90799c0e7b166a"><div class="ttname"><a href="namespaceagestrucne_1_1pgdriveneestimator.html#abd76f8aa32c542ee3e90799c0e7b166a">agestrucne.pgdriveneestimator.get_pops_in_file_outside_valid_size_range</a></div><div class="ttdeci">def get_pops_in_file_outside_valid_size_range(o_genepopfile, i_min_pop_size, i_max_pop_size)</div><div class="ttdef"><b>Definition:</b> <a href="pgdriveneestimator_8py_source.html#l01207">pgdriveneestimator.py:1207</a></div></div>
<div class="ttc" id="namespaceagestrucne_1_1pgdriveneestimator_html_a2ea5983385f4ef857bcf935d904c59b1"><div class="ttname"><a href="namespaceagestrucne_1_1pgdriveneestimator.html#a2ea5983385f4ef857bcf935d904c59b1">agestrucne.pgdriveneestimator.make_subsample_genepop_file_name</a></div><div class="ttdeci">def make_subsample_genepop_file_name(s_original_genepop_file, s_temporary_directory, i_genepop_file_count, i_indiv_sample_count, i_loci_subsample_count)</div><div class="ttdef"><b>Definition:</b> <a href="pgdriveneestimator_8py_source.html#l02404">pgdriveneestimator.py:2404</a></div></div>
</div><!-- fragment -->
</div>
</div>
<a class="anchor" id="ae574fb9ce112652e14c9b6e97942f634"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def agestrucne.pgdriveneestimator.do_estimate </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>xxx_todo_changeme</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">2017_07_07.  The paramaters min_loci_position
and max_loci_position are added to the arglist,
so that this def can compute the heterozygosity
only using the loci in the range specified by the
caller.
</pre> 
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l01359">1359</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>
<div class="fragment"><div class="line"><a name="l01359"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html#ae574fb9ce112652e14c9b6e97942f634"> 1359</a></span>&#160;<span class="keyword">def </span><a class="code" href="namespaceagestrucne_1_1pgdriveneestimator.html#ae574fb9ce112652e14c9b6e97942f634">do_estimate</a>(xxx_todo_changeme ):</div><div class="line"><a name="l01360"></a><span class="lineno"> 1360</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l01361"></a><span class="lineno"> 1361</span>&#160;<span class="stringliteral">    2017_07_07.  The paramaters min_loci_position</span></div><div class="line"><a name="l01362"></a><span class="lineno"> 1362</span>&#160;<span class="stringliteral">    and max_loci_position are added to the arglist,</span></div><div class="line"><a name="l01363"></a><span class="lineno"> 1363</span>&#160;<span class="stringliteral">    so that this def can compute the heterozygosity</span></div><div class="line"><a name="l01364"></a><span class="lineno"> 1364</span>&#160;<span class="stringliteral">    only using the loci in the range specified by the</span></div><div class="line"><a name="l01365"></a><span class="lineno"> 1365</span>&#160;<span class="stringliteral">    caller.</span></div><div class="line"><a name="l01366"></a><span class="lineno"> 1366</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l01367"></a><span class="lineno"> 1367</span>&#160;    ( o_genepopfile, o_ne_estimator, </div><div class="line"><a name="l01368"></a><span class="lineno"> 1368</span>&#160;                    s_sample_param_val, s_loci_sample_value,</div><div class="line"><a name="l01369"></a><span class="lineno"> 1369</span>&#160;                    f_min_allele_freq, b_monogamy,</div><div class="line"><a name="l01370"></a><span class="lineno"> 1370</span>&#160;                    s_subsample_tag, </div><div class="line"><a name="l01371"></a><span class="lineno"> 1371</span>&#160;                    s_loci_sample_tag, s_population_number, </div><div class="line"><a name="l01372"></a><span class="lineno"> 1372</span>&#160;                    s_census, i_replicate_number, </div><div class="line"><a name="l01373"></a><span class="lineno"> 1373</span>&#160;                    s_loci_replicate_number, o_debug_mode, </div><div class="line"><a name="l01374"></a><span class="lineno"> 1374</span>&#160;                        IDX_NE_ESTIMATOR_OUTPUT_FIELDS_TO_SKIP,</div><div class="line"><a name="l01375"></a><span class="lineno"> 1375</span>&#160;                        f_nbne_ratio,</div><div class="line"><a name="l01376"></a><span class="lineno"> 1376</span>&#160;                        i_min_loci_position,</div><div class="line"><a name="l01377"></a><span class="lineno"> 1377</span>&#160;                        i_max_loci_position) = xxx_todo_changeme</div><div class="line"><a name="l01378"></a><span class="lineno"> 1378</span>&#160;    <span class="keywordflow">try</span>:</div><div class="line"><a name="l01379"></a><span class="lineno"> 1379</span>&#160;</div><div class="line"><a name="l01380"></a><span class="lineno"> 1380</span>&#160;        s_genepop_file_subsample=o_ne_estimator.input.genepop_file</div><div class="line"><a name="l01381"></a><span class="lineno"> 1381</span>&#160;</div><div class="line"><a name="l01382"></a><span class="lineno"> 1382</span>&#160;        o_genepopfile.writeGenePopFile( s_genepop_file_subsample, </div><div class="line"><a name="l01383"></a><span class="lineno"> 1383</span>&#160;                                s_indiv_subsample_tag=s_subsample_tag,</div><div class="line"><a name="l01384"></a><span class="lineno"> 1384</span>&#160;                                s_pop_subsample_tag=s_population_number,</div><div class="line"><a name="l01385"></a><span class="lineno"> 1385</span>&#160;                                s_loci_subsample_tag=s_loci_sample_tag ) </div><div class="line"><a name="l01386"></a><span class="lineno"> 1386</span>&#160;</div><div class="line"><a name="l01387"></a><span class="lineno"> 1387</span>&#160;        o_ne_estimator.doOp()</div><div class="line"><a name="l01388"></a><span class="lineno"> 1388</span>&#160;</div><div class="line"><a name="l01389"></a><span class="lineno"> 1389</span>&#160;        llv_output=o_ne_estimator.deliverResults()</div><div class="line"><a name="l01390"></a><span class="lineno"> 1390</span>&#160;</div><div class="line"><a name="l01391"></a><span class="lineno"> 1391</span>&#160;        li_sample_indiv_count=o_genepopfile.getIndividualCounts( s_indiv_subsample_tag = s_subsample_tag, </div><div class="line"><a name="l01392"></a><span class="lineno"> 1392</span>&#160;                                                                    s_pop_subsample_tag = s_population_number )</div><div class="line"><a name="l01393"></a><span class="lineno"> 1393</span>&#160;</div><div class="line"><a name="l01394"></a><span class="lineno"> 1394</span>&#160;        s_sample_indiv_count = str( li_sample_indiv_count[ 0 ] )</div><div class="line"><a name="l01395"></a><span class="lineno"> 1395</span>&#160;</div><div class="line"><a name="l01396"></a><span class="lineno"> 1396</span>&#160;        ls_runinfo=[ o_genepopfile.original_file_name, </div><div class="line"><a name="l01397"></a><span class="lineno"> 1397</span>&#160;                                    s_population_number, </div><div class="line"><a name="l01398"></a><span class="lineno"> 1398</span>&#160;                                    s_census,</div><div class="line"><a name="l01399"></a><span class="lineno"> 1399</span>&#160;                                    s_sample_indiv_count, </div><div class="line"><a name="l01400"></a><span class="lineno"> 1400</span>&#160;                                    s_sample_param_val, </div><div class="line"><a name="l01401"></a><span class="lineno"> 1401</span>&#160;                                    str( i_replicate_number), </div><div class="line"><a name="l01402"></a><span class="lineno"> 1402</span>&#160;                                    s_loci_sample_value,</div><div class="line"><a name="l01403"></a><span class="lineno"> 1403</span>&#160;                                    s_loci_replicate_number,</div><div class="line"><a name="l01404"></a><span class="lineno"> 1404</span>&#160;                                    str( f_min_allele_freq ) ]</div><div class="line"><a name="l01405"></a><span class="lineno"> 1405</span>&#160;        </div><div class="line"><a name="l01406"></a><span class="lineno"> 1406</span>&#160;        ls_stdout=[]</div><div class="line"><a name="l01407"></a><span class="lineno"> 1407</span>&#160;        <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l01408"></a><span class="lineno"> 1408</span>&#160;<span class="stringliteral">        2017_06_30.  We now append a mean heterozygosity value based on allele</span></div><div class="line"><a name="l01409"></a><span class="lineno"> 1409</span>&#160;<span class="stringliteral">        freqs of the loci sampled in this run of the Ne estimator.</span></div><div class="line"><a name="l01410"></a><span class="lineno"> 1410</span>&#160;<span class="stringliteral">        </span></div><div class="line"><a name="l01411"></a><span class="lineno"> 1411</span>&#160;<span class="stringliteral">        2017_07_07. We add a loci sample that includes all in the range</span></div><div class="line"><a name="l01412"></a><span class="lineno"> 1412</span>&#160;<span class="stringliteral">        specified by the caller.  This allows het values, for example,</span></div><div class="line"><a name="l01413"></a><span class="lineno"> 1413</span>&#160;<span class="stringliteral">        to include only SNPs when the caller specifies range n to m, where</span></div><div class="line"><a name="l01414"></a><span class="lineno"> 1414</span>&#160;<span class="stringliteral">        loci n-1 is the last microsat and loci n is the first SNP.</span></div><div class="line"><a name="l01415"></a><span class="lineno"> 1415</span>&#160;<span class="stringliteral">        &#39;&#39;&#39;</span></div><div class="line"><a name="l01416"></a><span class="lineno"> 1416</span>&#160;</div><div class="line"><a name="l01417"></a><span class="lineno"> 1417</span>&#160;        o_genepopfile.subsampleLociByRangeAndMax( i_min_loci_position, i_max_loci_position, <span class="stringliteral">&quot;locirange&quot;</span> )</div><div class="line"><a name="l01418"></a><span class="lineno"> 1418</span>&#160;        </div><div class="line"><a name="l01419"></a><span class="lineno"> 1419</span>&#160;        df_het_values_by_pop_number=\</div><div class="line"><a name="l01420"></a><span class="lineno"> 1420</span>&#160;                <a class="code" href="namespaceagestrucne_1_1pgdriveneestimator.html#a2efaad3207ead8feb0abb42b921ab402">get_mean_het_based_on_allele_freqs</a>( \</div><div class="line"><a name="l01421"></a><span class="lineno"> 1421</span>&#160;                                            o_genepopfile,</div><div class="line"><a name="l01422"></a><span class="lineno"> 1422</span>&#160;                                            s_pop_subsample_tag=s_population_number,</div><div class="line"><a name="l01423"></a><span class="lineno"> 1423</span>&#160;                                            s_loci_subsample_tag=<span class="stringliteral">&quot;locirange&quot;</span> )</div><div class="line"><a name="l01424"></a><span class="lineno"> 1424</span>&#160;</div><div class="line"><a name="l01425"></a><span class="lineno"> 1425</span>&#160;        <span class="comment">#Becuase this def should always be operating on a single pop,</span></div><div class="line"><a name="l01426"></a><span class="lineno"> 1426</span>&#160;        <span class="comment">#We do this check for consistency:</span></div><div class="line"><a name="l01427"></a><span class="lineno"> 1427</span>&#160;        <span class="keywordflow">if</span> len( df_het_values_by_pop_number ) != 1:</div><div class="line"><a name="l01428"></a><span class="lineno"> 1428</span>&#160;            s_gp_file=o_genepopfile.original_file_name</div><div class="line"><a name="l01429"></a><span class="lineno"> 1429</span>&#160;            s_msg=<span class="stringliteral">&quot;In pgdriveneestimator.py, def do_estimate, &quot;</span> \</div><div class="line"><a name="l01430"></a><span class="lineno"> 1430</span>&#160;                        + <span class="stringliteral">&quot;the number of heterozygosity values for &quot;</span> \</div><div class="line"><a name="l01431"></a><span class="lineno"> 1431</span>&#160;                        + <span class="stringliteral">&quot;the estimate on population number &quot;</span> \</div><div class="line"><a name="l01432"></a><span class="lineno"> 1432</span>&#160;                        + s_population_number + <span class="stringliteral">&quot;, &quot;</span> \</div><div class="line"><a name="l01433"></a><span class="lineno"> 1433</span>&#160;                        + <span class="stringliteral">&quot;from file, &quot;</span> + s_gp_file + <span class="stringliteral">&quot;, &quot;</span> \</div><div class="line"><a name="l01434"></a><span class="lineno"> 1434</span>&#160;                        + <span class="stringliteral">&quot;should be a single number, &quot;</span> \</div><div class="line"><a name="l01435"></a><span class="lineno"> 1435</span>&#160;                        + <span class="stringliteral">&quot;but instead is, &quot;</span> + str( df_het_values_by_pop_number ) \</div><div class="line"><a name="l01436"></a><span class="lineno"> 1436</span>&#160;                        + <span class="stringliteral">&quot;.&quot;</span></div><div class="line"><a name="l01437"></a><span class="lineno"> 1437</span>&#160;            <span class="keywordflow">raise</span> Exception( s_msg )</div><div class="line"><a name="l01438"></a><span class="lineno"> 1438</span>&#160;        <span class="comment">#end if we did not get a single het value</span></div><div class="line"><a name="l01439"></a><span class="lineno"> 1439</span>&#160;</div><div class="line"><a name="l01440"></a><span class="lineno"> 1440</span>&#160;        i_pop_number_key_for_het = int( s_population_number )</div><div class="line"><a name="l01441"></a><span class="lineno"> 1441</span>&#160;        f_this_het=round( df_het_values_by_pop_number[ i_pop_number_key_for_het ], 4 )</div><div class="line"><a name="l01442"></a><span class="lineno"> 1442</span>&#160;        s_this_het=str( f_this_het )</div><div class="line"><a name="l01443"></a><span class="lineno"> 1443</span>&#160;        <span class="keywordflow">if</span> len( llv_output ) == 0:</div><div class="line"><a name="l01444"></a><span class="lineno"> 1444</span>&#160;            <span class="comment">#In this case we stub in &quot;NA&quot; vals for the missing estimator and bias adjust values</span></div><div class="line"><a name="l01445"></a><span class="lineno"> 1445</span>&#160;            i_total_estimator_fields=<a class="code" href="namespaceagestrucne_1_1pgdriveneestimator.html#aaba26add66aaee9119650828c2803cf7">get_count_estimator_fields</a>()</div><div class="line"><a name="l01446"></a><span class="lineno"> 1446</span>&#160;</div><div class="line"><a name="l01447"></a><span class="lineno"> 1447</span>&#160;            <span class="comment">#Besides the estimator fields, We add two fields for the Nb/Ne and bias adjustment</span></div><div class="line"><a name="l01448"></a><span class="lineno"> 1448</span>&#160;            i_number_nb_bias_fields=2</div><div class="line"><a name="l01449"></a><span class="lineno"> 1449</span>&#160;</div><div class="line"><a name="l01450"></a><span class="lineno"> 1450</span>&#160;            ls_output_vals_as_none=[ <span class="stringliteral">&quot;NA&quot;</span> <span class="keywordflow">for</span> idx \</div><div class="line"><a name="l01451"></a><span class="lineno"> 1451</span>&#160;                        <span class="keywordflow">in</span> range( i_total_estimator_fields  + i_number_nb_bias_fields ) ] </div><div class="line"><a name="l01452"></a><span class="lineno"> 1452</span>&#160;</div><div class="line"><a name="l01453"></a><span class="lineno"> 1453</span>&#160;            ls_stdout.append( OUTPUT_DELIMITER.join( ls_runinfo + ls_output_vals_as_none + [s_this_het] ) )</div><div class="line"><a name="l01454"></a><span class="lineno"> 1454</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l01455"></a><span class="lineno"> 1455</span>&#160;</div><div class="line"><a name="l01456"></a><span class="lineno"> 1456</span>&#160;            <span class="keywordflow">for</span> lv_output <span class="keywordflow">in</span> llv_output:</div><div class="line"><a name="l01457"></a><span class="lineno"> 1457</span>&#160;</div><div class="line"><a name="l01458"></a><span class="lineno"> 1458</span>&#160;                ls_fields_to_report=[]  </div><div class="line"><a name="l01459"></a><span class="lineno"> 1459</span>&#160;</div><div class="line"><a name="l01460"></a><span class="lineno"> 1460</span>&#160;                ls_output_vals_as_strings=[ str( v_val ) <span class="keywordflow">for</span> v_val <span class="keywordflow">in</span> lv_output ]</div><div class="line"><a name="l01461"></a><span class="lineno"> 1461</span>&#160;</div><div class="line"><a name="l01462"></a><span class="lineno"> 1462</span>&#160;                <span class="keywordflow">for</span> idx <span class="keywordflow">in</span> range( len( ls_output_vals_as_strings ) ):</div><div class="line"><a name="l01463"></a><span class="lineno"> 1463</span>&#160;                    <span class="keywordflow">if</span> idx <span class="keywordflow">not</span> <span class="keywordflow">in</span> IDX_NE_ESTIMATOR_OUTPUT_FIELDS_TO_SKIP:</div><div class="line"><a name="l01464"></a><span class="lineno"> 1464</span>&#160;                        ls_fields_to_report.append( ls_output_vals_as_strings[ idx ] )</div><div class="line"><a name="l01465"></a><span class="lineno"> 1465</span>&#160;                    <span class="comment">#end if idx is not for a field we&#39;re skipping</span></div><div class="line"><a name="l01466"></a><span class="lineno"> 1466</span>&#160;                <span class="comment">#end for each field&#39;s index</span></div><div class="line"><a name="l01467"></a><span class="lineno"> 1467</span>&#160;</div><div class="line"><a name="l01468"></a><span class="lineno"> 1468</span>&#160;                <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l01469"></a><span class="lineno"> 1469</span>&#160;<span class="stringliteral">                2017_04_17.  The bias adjustment code in this def is now simplified,</span></div><div class="line"><a name="l01470"></a><span class="lineno"> 1470</span>&#160;<span class="stringliteral">                so that it is passed an f_nbne_ratio, that is either None, or a float,</span></div><div class="line"><a name="l01471"></a><span class="lineno"> 1471</span>&#160;<span class="stringliteral">                so that this def can simply pass it and the objects needed to get the</span></div><div class="line"><a name="l01472"></a><span class="lineno"> 1472</span>&#160;<span class="stringliteral">                un-adjusted estimate.  Formerly, we passed the genepop file object to a</span></div><div class="line"><a name="l01473"></a><span class="lineno"> 1473</span>&#160;<span class="stringliteral">                allow the called def to search  the header for a ratio, if a flag</span></div><div class="line"><a name="l01474"></a><span class="lineno"> 1474</span>&#160;<span class="stringliteral">                (also used to be passed to this def) was true.  Now this ratio finding</span></div><div class="line"><a name="l01475"></a><span class="lineno"> 1475</span>&#160;<span class="stringliteral">                is done on a per-original-genepop file basis in def run_estimator.</span></div><div class="line"><a name="l01476"></a><span class="lineno"> 1476</span>&#160;<span class="stringliteral">                &#39;&#39;&#39;</span></div><div class="line"><a name="l01477"></a><span class="lineno"> 1477</span>&#160;</div><div class="line"><a name="l01478"></a><span class="lineno"> 1478</span>&#160;                s_ratio_used_for_adjustment, s_bias_adjusted_ne_value, s_bias_adjusted_ci_hi_value, s_bias_adjusted_ci_low_value=\</div><div class="line"><a name="l01479"></a><span class="lineno"> 1479</span>&#160;                            <a class="code" href="namespaceagestrucne_1_1pgdriveneestimator.html#a78e7def8930f60b0f5021f16b356e1e3">get_string_values_ldne_ratio_and_bias_adjustment</a>( \</div><div class="line"><a name="l01480"></a><span class="lineno"> 1480</span>&#160;                                                                    o_ne_estimator,</div><div class="line"><a name="l01481"></a><span class="lineno"> 1481</span>&#160;                                                                    lv_output,</div><div class="line"><a name="l01482"></a><span class="lineno"> 1482</span>&#160;                                                                    f_nbne_ratio )</div><div class="line"><a name="l01483"></a><span class="lineno"> 1483</span>&#160;</div><div class="line"><a name="l01484"></a><span class="lineno"> 1484</span>&#160;                </div><div class="line"><a name="l01485"></a><span class="lineno"> 1485</span>&#160;                ls_fields_to_report.append( s_ratio_used_for_adjustment )</div><div class="line"><a name="l01486"></a><span class="lineno"> 1486</span>&#160;                ls_fields_to_report.append( s_bias_adjusted_ne_value )</div><div class="line"><a name="l01487"></a><span class="lineno"> 1487</span>&#160;                ls_fields_to_report.append( s_bias_adjusted_ci_low_value )</div><div class="line"><a name="l01488"></a><span class="lineno"> 1488</span>&#160;                ls_fields_to_report.append( s_bias_adjusted_ci_hi_value )</div><div class="line"><a name="l01489"></a><span class="lineno"> 1489</span>&#160;                ls_fields_to_report.append( s_this_het )</div><div class="line"><a name="l01490"></a><span class="lineno"> 1490</span>&#160;                ls_stdout.append( OUTPUT_DELIMITER.join(  ls_runinfo + ls_fields_to_report )  )</div><div class="line"><a name="l01491"></a><span class="lineno"> 1491</span>&#160;</div><div class="line"><a name="l01492"></a><span class="lineno"> 1492</span>&#160;            <span class="comment">#end for each line of parsed NeEstimator Output</span></div><div class="line"><a name="l01493"></a><span class="lineno"> 1493</span>&#160;        <span class="comment">#end if we have no estimation output, else we do</span></div><div class="line"><a name="l01494"></a><span class="lineno"> 1494</span>&#160;        </div><div class="line"><a name="l01495"></a><span class="lineno"> 1495</span>&#160;        <span class="comment">#Make one long string, endline-delimited, for the stdout.</span></div><div class="line"><a name="l01496"></a><span class="lineno"> 1496</span>&#160;        <span class="comment">#Note:  if the inpout genepop file had only one</span></div><div class="line"><a name="l01497"></a><span class="lineno"> 1497</span>&#160;        <span class="comment">#population, this should be a single line of text.</span></div><div class="line"><a name="l01498"></a><span class="lineno"> 1498</span>&#160;</div><div class="line"><a name="l01499"></a><span class="lineno"> 1499</span>&#160;        s_stdout=ENDLINE_SEQ.join( ls_stdout )</div><div class="line"><a name="l01500"></a><span class="lineno"> 1500</span>&#160;</div><div class="line"><a name="l01501"></a><span class="lineno"> 1501</span>&#160;        s_stderr=<span class="keywordtype">None</span></div><div class="line"><a name="l01502"></a><span class="lineno"> 1502</span>&#160;</div><div class="line"><a name="l01503"></a><span class="lineno"> 1503</span>&#160;        ls_indiv_list=<span class="keywordtype">None</span></div><div class="line"><a name="l01504"></a><span class="lineno"> 1504</span>&#160;</div><div class="line"><a name="l01505"></a><span class="lineno"> 1505</span>&#160;        s_nodat_file=o_ne_estimator.output.getNoDatFileName()</div><div class="line"><a name="l01506"></a><span class="lineno"> 1506</span>&#160;</div><div class="line"><a name="l01507"></a><span class="lineno"> 1507</span>&#160;        <span class="keywordflow">if</span> o_debug_mode.isSet( DebugMode.MAKE_INDIV_TABLE ):</div><div class="line"><a name="l01508"></a><span class="lineno"> 1508</span>&#160;            ls_indiv_list = o_genepopfile.getListIndividuals( </div><div class="line"><a name="l01509"></a><span class="lineno"> 1509</span>&#160;                    i_pop_number=int( s_population_number ), </div><div class="line"><a name="l01510"></a><span class="lineno"> 1510</span>&#160;                    s_indiv_subsample_tag = s_subsample_tag )   </div><div class="line"><a name="l01511"></a><span class="lineno"> 1511</span>&#160;        <span class="comment">#end if return list indiv</span></div><div class="line"><a name="l01512"></a><span class="lineno"> 1512</span>&#160;        </div><div class="line"><a name="l01513"></a><span class="lineno"> 1513</span>&#160;        <span class="keywordflow">if</span> not( o_debug_mode.isSet( DebugMode.KEEP_ESTIMATOR_FILES) ):</div><div class="line"><a name="l01514"></a><span class="lineno"> 1514</span>&#160;            <span class="keywordflow">if</span> os.path.exists( o_ne_estimator.output.run_output_file ):</div><div class="line"><a name="l01515"></a><span class="lineno"> 1515</span>&#160;                os.remove( o_ne_estimator.output.run_output_file )</div><div class="line"><a name="l01516"></a><span class="lineno"> 1516</span>&#160;            <span class="comment">#end if NeEstimator&#39;s orig output file exists, delete it</span></div><div class="line"><a name="l01517"></a><span class="lineno"> 1517</span>&#160;        <span class="comment">#end if flag to keep output file is false</span></div><div class="line"><a name="l01518"></a><span class="lineno"> 1518</span>&#160;</div><div class="line"><a name="l01519"></a><span class="lineno"> 1519</span>&#160;        <span class="keywordflow">if</span> not( o_debug_mode.isSet( DebugMode.KEEP_REPLICATE_GENEPOP_FILES ) ):</div><div class="line"><a name="l01520"></a><span class="lineno"> 1520</span>&#160;            <span class="keywordflow">if</span> os.path.exists( o_ne_estimator.output.run_input_file ):</div><div class="line"><a name="l01521"></a><span class="lineno"> 1521</span>&#160;                os.remove( o_ne_estimator.output.run_input_file )</div><div class="line"><a name="l01522"></a><span class="lineno"> 1522</span>&#160;            <span class="comment">#end if subsample genepop file exists, delete it</span></div><div class="line"><a name="l01523"></a><span class="lineno"> 1523</span>&#160;        <span class="comment">#end if flag to keep input file is false</span></div><div class="line"><a name="l01524"></a><span class="lineno"> 1524</span>&#160;</div><div class="line"><a name="l01525"></a><span class="lineno"> 1525</span>&#160;        <span class="keywordflow">if</span> not( o_debug_mode.isSet( DebugMode.KEEP_NODAT_FILES ) ):</div><div class="line"><a name="l01526"></a><span class="lineno"> 1526</span>&#160;            s_nodat_file=o_ne_estimator.output.getNoDatFileName()</div><div class="line"><a name="l01527"></a><span class="lineno"> 1527</span>&#160;            <span class="keywordflow">if</span> s_nodat_file <span class="keywordflow">is</span> <span class="keywordflow">not</span> <span class="keywordtype">None</span>:</div><div class="line"><a name="l01528"></a><span class="lineno"> 1528</span>&#160;                os.remove( s_nodat_file )</div><div class="line"><a name="l01529"></a><span class="lineno"> 1529</span>&#160;            <span class="comment">#end for each nodatfile</span></div><div class="line"><a name="l01530"></a><span class="lineno"> 1530</span>&#160;        <span class="comment">#end if debug mode says remove nodats</span></div><div class="line"><a name="l01531"></a><span class="lineno"> 1531</span>&#160;</div><div class="line"><a name="l01532"></a><span class="lineno"> 1532</span>&#160;        <span class="keywordflow">if</span> not( o_debug_mode.isSet( DebugMode.KEEP_NE_EXTRA_FILES) ):</div><div class="line"><a name="l01533"></a><span class="lineno"> 1533</span>&#160;</div><div class="line"><a name="l01534"></a><span class="lineno"> 1534</span>&#160;            ls_extra_files=glob.glob( \</div><div class="line"><a name="l01535"></a><span class="lineno"> 1535</span>&#160;                    o_ne_estimator.output.run_output_file \</div><div class="line"><a name="l01536"></a><span class="lineno"> 1536</span>&#160;                    + <span class="stringliteral">&quot;*x??.txt&quot;</span> )</div><div class="line"><a name="l01537"></a><span class="lineno"> 1537</span>&#160;</div><div class="line"><a name="l01538"></a><span class="lineno"> 1538</span>&#160;            <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l01539"></a><span class="lineno"> 1539</span>&#160;<span class="stringliteral">            2017_03_21. To accomodate the use of the  LDNe2 estimator, </span></div><div class="line"><a name="l01540"></a><span class="lineno"> 1540</span>&#160;<span class="stringliteral">            to cleanup we add to glob the LDNe2 tabular output</span></div><div class="line"><a name="l01541"></a><span class="lineno"> 1541</span>&#160;<span class="stringliteral">            file extention, &quot;x.txt&quot;.</span></div><div class="line"><a name="l01542"></a><span class="lineno"> 1542</span>&#160;<span class="stringliteral">            &#39;&#39;&#39;</span></div><div class="line"><a name="l01543"></a><span class="lineno"> 1543</span>&#160;            ls_extra_files+=glob.glob( \</div><div class="line"><a name="l01544"></a><span class="lineno"> 1544</span>&#160;                    o_ne_estimator.output.run_output_file \</div><div class="line"><a name="l01545"></a><span class="lineno"> 1545</span>&#160;                    + <span class="stringliteral">&quot;x.txt&quot;</span> )</div><div class="line"><a name="l01546"></a><span class="lineno"> 1546</span>&#160;</div><div class="line"><a name="l01547"></a><span class="lineno"> 1547</span>&#160;            <span class="keywordflow">for</span> o_extra_file <span class="keywordflow">in</span> ls_extra_files:</div><div class="line"><a name="l01548"></a><span class="lineno"> 1548</span>&#160;                os.remove( o_extra_file )</div><div class="line"><a name="l01549"></a><span class="lineno"> 1549</span>&#160;            <span class="comment">#end for each nextpfile</span></div><div class="line"><a name="l01550"></a><span class="lineno"> 1550</span>&#160;        <span class="comment">#end if debug mode says remove nodats</span></div><div class="line"><a name="l01551"></a><span class="lineno"> 1551</span>&#160;</div><div class="line"><a name="l01552"></a><span class="lineno"> 1552</span>&#160;        </div><div class="line"><a name="l01553"></a><span class="lineno"> 1553</span>&#160;    <span class="keywordflow">except</span> Exception <span class="keyword">as</span> oex:</div><div class="line"><a name="l01554"></a><span class="lineno"> 1554</span>&#160;        o_traceback=sys.exc_info()[ 2 ]</div><div class="line"><a name="l01555"></a><span class="lineno"> 1555</span>&#160;        s_trace_msg=pgut.get_traceback_info_about_offending_code( o_traceback ) </div><div class="line"><a name="l01556"></a><span class="lineno"> 1556</span>&#160;        <span class="keywordflow">raise</span> Exception( <span class="stringliteral">&quot;An exception was raised in &quot;</span> \</div><div class="line"><a name="l01557"></a><span class="lineno"> 1557</span>&#160;                             + <span class="stringliteral">&quot;module pgdriveneestimator.py, &quot;</span> \</div><div class="line"><a name="l01558"></a><span class="lineno"> 1558</span>&#160;                             + <span class="stringliteral">&quot;def do_estimate, with message: &quot;</span> \</div><div class="line"><a name="l01559"></a><span class="lineno"> 1559</span>&#160;                             + str( oex ) + <span class="stringliteral">&quot;\n&quot;</span> + <span class="stringliteral">&quot;Exeption origin from &quot;</span> \</div><div class="line"><a name="l01560"></a><span class="lineno"> 1560</span>&#160;                             + s_trace_msg )</div><div class="line"><a name="l01561"></a><span class="lineno"> 1561</span>&#160;    <span class="comment">#end try...except</span></div><div class="line"><a name="l01562"></a><span class="lineno"> 1562</span>&#160;    <span class="keywordflow">return</span> { <span class="stringliteral">&quot;for_stdout&quot;</span> : s_stdout, <span class="stringliteral">&quot;for_stderr&quot;</span> : s_stderr, </div><div class="line"><a name="l01563"></a><span class="lineno"> 1563</span>&#160;            <span class="stringliteral">&quot;for_indiv_table&quot;</span>: <span class="keywordtype">None</span> <span class="keywordflow">if</span> ls_indiv_list <span class="keywordflow">is</span> <span class="keywordtype">None</span> \</div><div class="line"><a name="l01564"></a><span class="lineno"> 1564</span>&#160;                    <span class="keywordflow">else</span> { <span class="stringliteral">&quot;file&quot;</span> : o_genepopfile.original_file_name, </div><div class="line"><a name="l01565"></a><span class="lineno"> 1565</span>&#160;                        <span class="stringliteral">&quot;sample_val&quot;</span>: s_sample_param_val,</div><div class="line"><a name="l01566"></a><span class="lineno"> 1566</span>&#160;                        <span class="stringliteral">&quot;loci_val&quot;</span>:s_loci_sample_value,</div><div class="line"><a name="l01567"></a><span class="lineno"> 1567</span>&#160;                        <span class="stringliteral">&quot;rep&quot;</span> : i_replicate_number,</div><div class="line"><a name="l01568"></a><span class="lineno"> 1568</span>&#160;                        <span class="stringliteral">&quot;loci_rep&quot;</span>:s_loci_replicate_number,</div><div class="line"><a name="l01569"></a><span class="lineno"> 1569</span>&#160;                        <span class="stringliteral">&quot;list&quot;</span> : ls_indiv_list,</div><div class="line"><a name="l01570"></a><span class="lineno"> 1570</span>&#160;                        <span class="stringliteral">&quot;pop&quot;</span> : s_population_number } }</div><div class="line"><a name="l01571"></a><span class="lineno"> 1571</span>&#160;</div><div class="line"><a name="l01572"></a><span class="lineno"> 1572</span>&#160;<span class="comment">#end do_estimate</span></div><div class="line"><a name="l01573"></a><span class="lineno"> 1573</span>&#160;</div><div class="ttc" id="namespaceagestrucne_1_1pgdriveneestimator_html_a2efaad3207ead8feb0abb42b921ab402"><div class="ttname"><a href="namespaceagestrucne_1_1pgdriveneestimator.html#a2efaad3207ead8feb0abb42b921ab402">agestrucne.pgdriveneestimator.get_mean_het_based_on_allele_freqs</a></div><div class="ttdeci">def get_mean_het_based_on_allele_freqs(o_genepopfile, s_pop_subsample_tag=None, s_individual_subsample_tag=None, s_loci_subsample_tag=None)</div><div class="ttdef"><b>Definition:</b> <a href="pgdriveneestimator_8py_source.html#l01319">pgdriveneestimator.py:1319</a></div></div>
<div class="ttc" id="namespaceagestrucne_1_1pgdriveneestimator_html_a78e7def8930f60b0f5021f16b356e1e3"><div class="ttname"><a href="namespaceagestrucne_1_1pgdriveneestimator.html#a78e7def8930f60b0f5021f16b356e1e3">agestrucne.pgdriveneestimator.get_string_values_ldne_ratio_and_bias_adjustment</a></div><div class="ttdeci">def get_string_values_ldne_ratio_and_bias_adjustment(o_ne_estimator, lv_results_list, f_nbne_ratio)</div><div class="ttdef"><b>Definition:</b> <a href="pgdriveneestimator_8py_source.html#l01235">pgdriveneestimator.py:1235</a></div></div>
<div class="ttc" id="namespaceagestrucne_1_1pgdriveneestimator_html_ae574fb9ce112652e14c9b6e97942f634"><div class="ttname"><a href="namespaceagestrucne_1_1pgdriveneestimator.html#ae574fb9ce112652e14c9b6e97942f634">agestrucne.pgdriveneestimator.do_estimate</a></div><div class="ttdeci">def do_estimate(xxx_todo_changeme)</div><div class="ttdef"><b>Definition:</b> <a href="pgdriveneestimator_8py_source.html#l01359">pgdriveneestimator.py:1359</a></div></div>
<div class="ttc" id="namespaceagestrucne_1_1pgdriveneestimator_html_aaba26add66aaee9119650828c2803cf7"><div class="ttname"><a href="namespaceagestrucne_1_1pgdriveneestimator.html#aaba26add66aaee9119650828c2803cf7">agestrucne.pgdriveneestimator.get_count_estimator_fields</a></div><div class="ttdeci">def get_count_estimator_fields()</div><div class="ttdef"><b>Definition:</b> <a href="pgdriveneestimator_8py_source.html#l02959">pgdriveneestimator.py:2959</a></div></div>
</div><!-- fragment -->
</div>
</div>
<a class="anchor" id="aaf2e5f67e91fb3bff9ad32c7dd0ccf53"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def agestrucne.pgdriveneestimator.do_ldne_bias_adjustment </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>f_ldne_estimate</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>f_nbne_ratio</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">2017_02_11. Implements, through the LDNENbBiasAdjustor
class, the bias adjustment to LDNE estimates
via eq (8) in Waples 2014 (see class description).
</pre> 
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l01349">1349</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>
<div class="fragment"><div class="line"><a name="l01349"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html#aaf2e5f67e91fb3bff9ad32c7dd0ccf53"> 1349</a></span>&#160;<span class="keyword">def </span><a class="code" href="namespaceagestrucne_1_1pgdriveneestimator.html#aaf2e5f67e91fb3bff9ad32c7dd0ccf53">do_ldne_bias_adjustment</a>( f_ldne_estimate, f_nbne_ratio ):</div><div class="line"><a name="l01350"></a><span class="lineno"> 1350</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l01351"></a><span class="lineno"> 1351</span>&#160;<span class="stringliteral">    2017_02_11. Implements, through the LDNENbBiasAdjustor</span></div><div class="line"><a name="l01352"></a><span class="lineno"> 1352</span>&#160;<span class="stringliteral">    class, the bias adjustment to LDNE estimates</span></div><div class="line"><a name="l01353"></a><span class="lineno"> 1353</span>&#160;<span class="stringliteral">    via eq (8) in Waples 2014 (see class description).</span></div><div class="line"><a name="l01354"></a><span class="lineno"> 1354</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l01355"></a><span class="lineno"> 1355</span>&#160;    o_adjustor=<a class="code" href="classagestrucne_1_1pgutilityclasses_1_1LDNENbBiasAdjustor.html">LDNENbBiasAdjustor</a>( f_ldne_estimate, f_nbne_ratio )</div><div class="line"><a name="l01356"></a><span class="lineno"> 1356</span>&#160;    <span class="keywordflow">return</span> o_adjustor.adjusted_nb</div><div class="line"><a name="l01357"></a><span class="lineno"> 1357</span>&#160;<span class="comment">#end def do_ldne_bias_adjustment</span></div><div class="line"><a name="l01358"></a><span class="lineno"> 1358</span>&#160;</div><div class="ttc" id="namespaceagestrucne_1_1pgdriveneestimator_html_aaf2e5f67e91fb3bff9ad32c7dd0ccf53"><div class="ttname"><a href="namespaceagestrucne_1_1pgdriveneestimator.html#aaf2e5f67e91fb3bff9ad32c7dd0ccf53">agestrucne.pgdriveneestimator.do_ldne_bias_adjustment</a></div><div class="ttdeci">def do_ldne_bias_adjustment(f_ldne_estimate, f_nbne_ratio)</div><div class="ttdef"><b>Definition:</b> <a href="pgdriveneestimator_8py_source.html#l01349">pgdriveneestimator.py:1349</a></div></div>
<div class="ttc" id="classagestrucne_1_1pgutilityclasses_1_1LDNENbBiasAdjustor_html"><div class="ttname"><a href="classagestrucne_1_1pgutilityclasses_1_1LDNENbBiasAdjustor.html">agestrucne.pgutilityclasses.LDNENbBiasAdjustor</a></div><div class="ttdef"><b>Definition:</b> <a href="pgutilityclasses_8py_source.html#l01082">pgutilityclasses.py:1082</a></div></div>
</div><!-- fragment -->
</div>
</div>
<a class="anchor" id="a1d38a43a7280fa4c608458abd861bacc"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def agestrucne.pgdriveneestimator.do_sample </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>o_genepopfile</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_min_pop_range</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_max_pop_range</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_min_pop_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_max_pop_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>s_sample_scheme</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>lv_sample_values</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_replicates</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>s_loci_sampling_scheme</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>v_loci_sampling_scheme_param</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_min_loci_position</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_max_loci_position</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_min_total_loci</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_max_total_loci</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_loci_replicates</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">2017_05_29.  We add a return value, so that the calling
def (drive_estimator), will know whether any pops were
sampled. If the call to do_sample_individuals has no
pops in range we return 0, otherwise, we do the loci
sampling and return the value returned, which will be
the pop list length, or zero if no loci are in range,
as given by min_loci_number and max_loci_number.
</pre> 
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l01870">1870</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>
<div class="fragment"><div class="line"><a name="l01870"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html#a1d38a43a7280fa4c608458abd861bacc"> 1870</a></span>&#160;                i_loci_replicates ):</div><div class="line"><a name="l01871"></a><span class="lineno"> 1871</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l01872"></a><span class="lineno"> 1872</span>&#160;<span class="stringliteral">    2017_05_29.  We add a return value, so that the calling</span></div><div class="line"><a name="l01873"></a><span class="lineno"> 1873</span>&#160;<span class="stringliteral">    def (drive_estimator), will know whether any pops were</span></div><div class="line"><a name="l01874"></a><span class="lineno"> 1874</span>&#160;<span class="stringliteral">    sampled. If the call to do_sample_individuals has no</span></div><div class="line"><a name="l01875"></a><span class="lineno"> 1875</span>&#160;<span class="stringliteral">    pops in range we return 0, otherwise, we do the loci</span></div><div class="line"><a name="l01876"></a><span class="lineno"> 1876</span>&#160;<span class="stringliteral">    sampling and return the value returned, which will be</span></div><div class="line"><a name="l01877"></a><span class="lineno"> 1877</span>&#160;<span class="stringliteral">    the pop list length, or zero if no loci are in range,</span></div><div class="line"><a name="l01878"></a><span class="lineno"> 1878</span>&#160;<span class="stringliteral">    as given by min_loci_number and max_loci_number.</span></div><div class="line"><a name="l01879"></a><span class="lineno"> 1879</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l01880"></a><span class="lineno"> 1880</span>&#160;</div><div class="line"><a name="l01881"></a><span class="lineno"> 1881</span>&#160;    li_population_list=<a class="code" href="namespaceagestrucne_1_1pgdriveneestimator.html#a8d8109b98382d21b972e701323440f95">do_sample_individuals</a>( o_genepopfile, </div><div class="line"><a name="l01882"></a><span class="lineno"> 1882</span>&#160;                    i_min_pop_range,</div><div class="line"><a name="l01883"></a><span class="lineno"> 1883</span>&#160;                    i_max_pop_range,</div><div class="line"><a name="l01884"></a><span class="lineno"> 1884</span>&#160;                    i_min_pop_size,</div><div class="line"><a name="l01885"></a><span class="lineno"> 1885</span>&#160;                    i_max_pop_size,</div><div class="line"><a name="l01886"></a><span class="lineno"> 1886</span>&#160;                    s_sample_scheme, </div><div class="line"><a name="l01887"></a><span class="lineno"> 1887</span>&#160;                    lv_sample_values, </div><div class="line"><a name="l01888"></a><span class="lineno"> 1888</span>&#160;                    i_replicates )</div><div class="line"><a name="l01889"></a><span class="lineno"> 1889</span>&#160;</div><div class="line"><a name="l01890"></a><span class="lineno"> 1890</span>&#160;    <span class="keywordflow">if</span> li_population_list <span class="keywordflow">is</span> <span class="keywordtype">None</span>:</div><div class="line"><a name="l01891"></a><span class="lineno"> 1891</span>&#160;        <span class="keywordflow">return</span> 0</div><div class="line"><a name="l01892"></a><span class="lineno"> 1892</span>&#160;    <span class="comment">#end if no pops were sampled (i.e. min pop &gt; total pops).  </span></div><div class="line"><a name="l01893"></a><span class="lineno"> 1893</span>&#160;</div><div class="line"><a name="l01894"></a><span class="lineno"> 1894</span>&#160;    <span class="comment">#we also want to sample loci one loci sampling</span></div><div class="line"><a name="l01895"></a><span class="lineno"> 1895</span>&#160;    <span class="comment">#replicate for each each individual replicate,</span></div><div class="line"><a name="l01896"></a><span class="lineno"> 1896</span>&#160;    <span class="comment">#the latter now included in the genepopfile object:</span></div><div class="line"><a name="l01897"></a><span class="lineno"> 1897</span>&#160;    i_total_pops_sampled=<a class="code" href="namespaceagestrucne_1_1pgdriveneestimator.html#a9544ec0a253ab2bcc1fcfbe1edb66120">do_sample_loci</a>( o_genepopfile,</div><div class="line"><a name="l01898"></a><span class="lineno"> 1898</span>&#160;                    s_loci_sampling_scheme,</div><div class="line"><a name="l01899"></a><span class="lineno"> 1899</span>&#160;                    v_loci_sampling_scheme_param,</div><div class="line"><a name="l01900"></a><span class="lineno"> 1900</span>&#160;                    i_min_loci_position,</div><div class="line"><a name="l01901"></a><span class="lineno"> 1901</span>&#160;                    i_max_loci_position,</div><div class="line"><a name="l01902"></a><span class="lineno"> 1902</span>&#160;                    i_min_total_loci,</div><div class="line"><a name="l01903"></a><span class="lineno"> 1903</span>&#160;                    i_max_total_loci,</div><div class="line"><a name="l01904"></a><span class="lineno"> 1904</span>&#160;                    li_population_list,</div><div class="line"><a name="l01905"></a><span class="lineno"> 1905</span>&#160;                    i_loci_replicates )</div><div class="line"><a name="l01906"></a><span class="lineno"> 1906</span>&#160;</div><div class="line"><a name="l01907"></a><span class="lineno"> 1907</span>&#160;    <span class="keywordflow">return</span> i_total_pops_sampled</div><div class="line"><a name="l01908"></a><span class="lineno"> 1908</span>&#160;<span class="comment">#end do_sample</span></div><div class="line"><a name="l01909"></a><span class="lineno"> 1909</span>&#160;</div><div class="ttc" id="namespaceagestrucne_1_1pgdriveneestimator_html_a8d8109b98382d21b972e701323440f95"><div class="ttname"><a href="namespaceagestrucne_1_1pgdriveneestimator.html#a8d8109b98382d21b972e701323440f95">agestrucne.pgdriveneestimator.do_sample_individuals</a></div><div class="ttdeci">def do_sample_individuals(o_genepopfile, i_min_pop_range, i_max_pop_range, i_min_pop_size, i_max_pop_size, s_sample_scheme, lv_sample_values, i_replicates)</div><div class="ttdef"><b>Definition:</b> <a href="pgdriveneestimator_8py_source.html#l01917">pgdriveneestimator.py:1917</a></div></div>
<div class="ttc" id="namespaceagestrucne_1_1pgdriveneestimator_html_a9544ec0a253ab2bcc1fcfbe1edb66120"><div class="ttname"><a href="namespaceagestrucne_1_1pgdriveneestimator.html#a9544ec0a253ab2bcc1fcfbe1edb66120">agestrucne.pgdriveneestimator.do_sample_loci</a></div><div class="ttdeci">def do_sample_loci(o_genepopfile, s_loci_sampling_scheme, v_loci_sampling_scheme_param, i_min_loci_position, i_max_loci_position, i_min_total_loci, i_max_total_loci, li_population_list, i_loci_replicates=1)</div><div class="ttdef"><b>Definition:</b> <a href="pgdriveneestimator_8py_source.html#l02190">pgdriveneestimator.py:2190</a></div></div>
</div><!-- fragment -->
</div>
</div>
<a class="anchor" id="a8d8109b98382d21b972e701323440f95"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def agestrucne.pgdriveneestimator.do_sample_individuals </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>o_genepopfile</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_min_pop_range</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_max_pop_range</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_min_pop_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_max_pop_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>s_sample_scheme</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>lv_sample_values</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_replicates</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">do_sample adds subsample info to the o_genepopfile GenepopFileManager object,
and so changes it in place, returns no value.  The o_genepopfile can then
be accessed to write a new genepop file that contains the subsampled data.
sampler requires a population list (in case we want to sample fewer than
all the pops in the file). If the client passed in "all" parse_args
will have assigned None to both min and max.

2017_05_29.  We add check to see if the min pop range number is &gt;
the total pops int he genepop file, in which case we return None.
</pre> 
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l01917">1917</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>
<div class="fragment"><div class="line"><a name="l01917"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html#a8d8109b98382d21b972e701323440f95"> 1917</a></span>&#160;                i_replicates ):</div><div class="line"><a name="l01918"></a><span class="lineno"> 1918</span>&#160;</div><div class="line"><a name="l01919"></a><span class="lineno"> 1919</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l01920"></a><span class="lineno"> 1920</span>&#160;<span class="stringliteral">    do_sample adds subsample info to the o_genepopfile GenepopFileManager object,</span></div><div class="line"><a name="l01921"></a><span class="lineno"> 1921</span>&#160;<span class="stringliteral">    and so changes it in place, returns no value.  The o_genepopfile can then</span></div><div class="line"><a name="l01922"></a><span class="lineno"> 1922</span>&#160;<span class="stringliteral">    be accessed to write a new genepop file that contains the subsampled data.</span></div><div class="line"><a name="l01923"></a><span class="lineno"> 1923</span>&#160;<span class="stringliteral">    sampler requires a population list (in case we want to sample fewer than</span></div><div class="line"><a name="l01924"></a><span class="lineno"> 1924</span>&#160;<span class="stringliteral">    all the pops in the file). If the client passed in &quot;all&quot; parse_args</span></div><div class="line"><a name="l01925"></a><span class="lineno"> 1925</span>&#160;<span class="stringliteral">    will have assigned None to both min and max.</span></div><div class="line"><a name="l01926"></a><span class="lineno"> 1926</span>&#160;<span class="stringliteral"></span></div><div class="line"><a name="l01927"></a><span class="lineno"> 1927</span>&#160;<span class="stringliteral">    2017_05_29.  We add check to see if the min pop range number is &gt;</span></div><div class="line"><a name="l01928"></a><span class="lineno"> 1928</span>&#160;<span class="stringliteral">    the total pops int he genepop file, in which case we return None.</span></div><div class="line"><a name="l01929"></a><span class="lineno"> 1929</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l01930"></a><span class="lineno"> 1930</span>&#160;</div><div class="line"><a name="l01931"></a><span class="lineno"> 1931</span>&#160;    i_total_pops_in_file=o_genepopfile.pop_total</div><div class="line"><a name="l01932"></a><span class="lineno"> 1932</span>&#160;    li_population_list=<span class="keywordtype">None</span></div><div class="line"><a name="l01933"></a><span class="lineno"> 1933</span>&#160;</div><div class="line"><a name="l01934"></a><span class="lineno"> 1934</span>&#160;    <span class="keywordflow">if</span> i_min_pop_range <span class="keywordflow">is</span> <span class="keywordtype">None</span>:</div><div class="line"><a name="l01935"></a><span class="lineno"> 1935</span>&#160;        <span class="comment">#make sure both are None, else</span></div><div class="line"><a name="l01936"></a><span class="lineno"> 1936</span>&#160;        <span class="comment">#arg-parsing error:</span></div><div class="line"><a name="l01937"></a><span class="lineno"> 1937</span>&#160;        <span class="keywordflow">if</span> i_max_pop_range <span class="keywordflow">is</span> <span class="keywordtype">None</span>:</div><div class="line"><a name="l01938"></a><span class="lineno"> 1938</span>&#160;            li_population_list=list( range( 1, i_total_pops_in_file + 1 ) )</div><div class="line"><a name="l01939"></a><span class="lineno"> 1939</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l01940"></a><span class="lineno"> 1940</span>&#160;            s_msg=<span class="stringliteral">&quot;In pgdriveneestimator.py, def do_sample, &quot;</span> \</div><div class="line"><a name="l01941"></a><span class="lineno"> 1941</span>&#160;                        + <span class="stringliteral">&quot; found arg min pop range is None, &quot;</span> \</div><div class="line"><a name="l01942"></a><span class="lineno"> 1942</span>&#160;                        + <span class="stringliteral">&quot; but max pop range is not None, &quot;</span> \</div><div class="line"><a name="l01943"></a><span class="lineno"> 1943</span>&#160;                        + str( i_max_pop_range ) \</div><div class="line"><a name="l01944"></a><span class="lineno"> 1944</span>&#160;                        + <span class="stringliteral">&quot;.  Invalid argument pair.&quot;</span></div><div class="line"><a name="l01945"></a><span class="lineno"> 1945</span>&#160;            <span class="keywordflow">raise</span> Exception( s_msg )</div><div class="line"><a name="l01946"></a><span class="lineno"> 1946</span>&#160;        <span class="comment">#end if max is None, then get all pops into list, else error</span></div><div class="line"><a name="l01947"></a><span class="lineno"> 1947</span>&#160;    <span class="keywordflow">elif</span> i_min_pop_range &lt; 1 <span class="keywordflow">or</span> i_min_pop_range &gt; i_max_pop_range:</div><div class="line"><a name="l01948"></a><span class="lineno"> 1948</span>&#160;            s_msg=<span class="stringliteral">&quot;In pgdriveneestimator.py, def do_sample, &quot;</span> \</div><div class="line"><a name="l01949"></a><span class="lineno"> 1949</span>&#160;                        + <span class="stringliteral">&quot;min pop number range, &quot;</span> \</div><div class="line"><a name="l01950"></a><span class="lineno"> 1950</span>&#160;                        + str( i_min_pop_range ) \</div><div class="line"><a name="l01951"></a><span class="lineno"> 1951</span>&#160;                        + <span class="stringliteral">&quot;, is less than one or &quot;</span> \</div><div class="line"><a name="l01952"></a><span class="lineno"> 1952</span>&#160;                        + <span class="stringliteral">&quot;greaterh than max pop number, &quot;</span> \</div><div class="line"><a name="l01953"></a><span class="lineno"> 1953</span>&#160;                        + str( i_max_pop_range ) + <span class="stringliteral">&quot;.&quot;</span> </div><div class="line"><a name="l01954"></a><span class="lineno"> 1954</span>&#160;            <span class="keywordflow">raise</span> Exception( s_msg )</div><div class="line"><a name="l01955"></a><span class="lineno"> 1955</span>&#160;        <span class="comment">#end if min or max out of range</span></div><div class="line"><a name="l01956"></a><span class="lineno"> 1956</span>&#160;    <span class="keywordflow">elif</span> i_min_pop_range &gt; i_total_pops_in_file:</div><div class="line"><a name="l01957"></a><span class="lineno"> 1957</span>&#160;        <span class="keywordflow">return</span> <span class="keywordtype">None</span></div><div class="line"><a name="l01958"></a><span class="lineno"> 1958</span>&#160;    <span class="keywordflow">else</span>:</div><div class="line"><a name="l01959"></a><span class="lineno"> 1959</span>&#160;</div><div class="line"><a name="l01960"></a><span class="lineno"> 1960</span>&#160;        i_max_pop_range=min( i_max_pop_range, i_total_pops_in_file )</div><div class="line"><a name="l01961"></a><span class="lineno"> 1961</span>&#160;        li_population_list=list(range( i_min_pop_range, i_max_pop_range + 1))</div><div class="line"><a name="l01962"></a><span class="lineno"> 1962</span>&#160;    <span class="comment">#end if min pop range is None, else if range invalid, else if out of range, </span></div><div class="line"><a name="l01963"></a><span class="lineno"> 1963</span>&#160;    <span class="comment">#else good range, get corresponding pop number list </span></div><div class="line"><a name="l01964"></a><span class="lineno"> 1964</span>&#160;</div><div class="line"><a name="l01965"></a><span class="lineno"> 1965</span>&#160;    o_sampler=<span class="keywordtype">None</span>  </div><div class="line"><a name="l01966"></a><span class="lineno"> 1966</span>&#160;</div><div class="line"><a name="l01967"></a><span class="lineno"> 1967</span>&#160;    s_population_subsample_tag=s_sample_scheme</div><div class="line"><a name="l01968"></a><span class="lineno"> 1968</span>&#160;</div><div class="line"><a name="l01969"></a><span class="lineno"> 1969</span>&#160;    <span class="keywordflow">if</span> s_sample_scheme==SAMPLE_BY_NONE:</div><div class="line"><a name="l01970"></a><span class="lineno"> 1970</span>&#160;        o_sample_params=gps.GenepopFileSampleParamsNone( li_population_numbers=li_population_list,</div><div class="line"><a name="l01971"></a><span class="lineno"> 1971</span>&#160;                                                            i_min_pop_size=i_min_pop_size,</div><div class="line"><a name="l01972"></a><span class="lineno"> 1972</span>&#160;                                                            i_max_pop_size=i_max_pop_size,</div><div class="line"><a name="l01973"></a><span class="lineno"> 1973</span>&#160;                                                            i_replicates=i_replicates,</div><div class="line"><a name="l01974"></a><span class="lineno"> 1974</span>&#160;                                                            s_population_subsample_name=s_population_subsample_tag )    </div><div class="line"><a name="l01975"></a><span class="lineno"> 1975</span>&#160;</div><div class="line"><a name="l01976"></a><span class="lineno"> 1976</span>&#160;        o_sampler=gps.GenepopFileSamplerNone( o_genepopfile, o_sample_params )</div><div class="line"><a name="l01977"></a><span class="lineno"> 1977</span>&#160;</div><div class="line"><a name="l01978"></a><span class="lineno"> 1978</span>&#160;    <span class="keywordflow">elif</span> s_sample_scheme==SAMPLE_BY_PERCENTAGES:</div><div class="line"><a name="l01979"></a><span class="lineno"> 1979</span>&#160;</div><div class="line"><a name="l01980"></a><span class="lineno"> 1980</span>&#160;        <span class="comment">#our &quot;proportion&quot; sampler object needs proportions instead of percentages:</span></div><div class="line"><a name="l01981"></a><span class="lineno"> 1981</span>&#160;</div><div class="line"><a name="l01982"></a><span class="lineno"> 1982</span>&#160;        o_sample_params=gps.GenepopFileSampleParamsProportion( li_population_numbers=li_population_list,</div><div class="line"><a name="l01983"></a><span class="lineno"> 1983</span>&#160;                                                lf_proportions=lv_sample_values,</div><div class="line"><a name="l01984"></a><span class="lineno"> 1984</span>&#160;                                                i_replicates=i_replicates,</div><div class="line"><a name="l01985"></a><span class="lineno"> 1985</span>&#160;                                                s_population_subsample_name=s_population_subsample_tag )</div><div class="line"><a name="l01986"></a><span class="lineno"> 1986</span>&#160;</div><div class="line"><a name="l01987"></a><span class="lineno"> 1987</span>&#160;        o_sampler=gps.GenepopFileSamplerIndividualsByProportion( o_genepopfile, o_sample_params )</div><div class="line"><a name="l01988"></a><span class="lineno"> 1988</span>&#160;</div><div class="line"><a name="l01989"></a><span class="lineno"> 1989</span>&#160;    <span class="keywordflow">elif</span> s_sample_scheme==SAMPLE_BY_REMOVAL:</div><div class="line"><a name="l01990"></a><span class="lineno"> 1990</span>&#160;</div><div class="line"><a name="l01991"></a><span class="lineno"> 1991</span>&#160;        o_sample_params=gps.GenepopFileSampleParamsRemoval( li_population_numbers=li_population_list,</div><div class="line"><a name="l01992"></a><span class="lineno"> 1992</span>&#160;                                                li_n_to_remove=lv_sample_values,</div><div class="line"><a name="l01993"></a><span class="lineno"> 1993</span>&#160;                                                i_replicates=i_replicates,</div><div class="line"><a name="l01994"></a><span class="lineno"> 1994</span>&#160;                                                s_population_subsample_name=s_population_subsample_tag )</div><div class="line"><a name="l01995"></a><span class="lineno"> 1995</span>&#160;</div><div class="line"><a name="l01996"></a><span class="lineno"> 1996</span>&#160;        o_sampler=gps.GenepopFileSamplerIndividualsByRemoval( o_genepopfile, o_sample_params )</div><div class="line"><a name="l01997"></a><span class="lineno"> 1997</span>&#160;</div><div class="line"><a name="l01998"></a><span class="lineno"> 1998</span>&#160;    <span class="keywordflow">elif</span> s_sample_scheme==SAMPLE_BY_CRITERIA:</div><div class="line"><a name="l01999"></a><span class="lineno"> 1999</span>&#160;</div><div class="line"><a name="l02000"></a><span class="lineno"> 2000</span>&#160;        o_indiv_fields, o_criteria = get_criteria_sampling_params( lv_sample_values )</div><div class="line"><a name="l02001"></a><span class="lineno"> 2001</span>&#160;</div><div class="line"><a name="l02002"></a><span class="lineno"> 2002</span>&#160;        o_sample_params=gps.GenepopFileSampleParamsCriteria( o_genepop_indiv_id_fields=o_indiv_fields,</div><div class="line"><a name="l02003"></a><span class="lineno"> 2003</span>&#160;                                                                    o_genepop_indiv_id_critera=o_criteria,</div><div class="line"><a name="l02004"></a><span class="lineno"> 2004</span>&#160;                                                                    li_population_numbers=li_population_list,</div><div class="line"><a name="l02005"></a><span class="lineno"> 2005</span>&#160;                                                                    i_min_sampled_pop_size=i_min_pop_size,</div><div class="line"><a name="l02006"></a><span class="lineno"> 2006</span>&#160;                                                                    i_max_sampled_pop_size=i_max_pop_size,</div><div class="line"><a name="l02007"></a><span class="lineno"> 2007</span>&#160;                                                                    s_population_subsample_name = s_population_subsample_tag,</div><div class="line"><a name="l02008"></a><span class="lineno"> 2008</span>&#160;                                                                    i_replicates=i_replicates )</div><div class="line"><a name="l02009"></a><span class="lineno"> 2009</span>&#160;</div><div class="line"><a name="l02010"></a><span class="lineno"> 2010</span>&#160;        o_sampler=gps.GenepopFileSamplerIndividualsByCriteria( o_genepopfile, o_sample_params )</div><div class="line"><a name="l02011"></a><span class="lineno"> 2011</span>&#160;</div><div class="line"><a name="l02012"></a><span class="lineno"> 2012</span>&#160;    <span class="keywordflow">elif</span> s_sample_scheme==SAMPLE_BY_COHORT:</div><div class="line"><a name="l02013"></a><span class="lineno"> 2013</span>&#160;</div><div class="line"><a name="l02014"></a><span class="lineno"> 2014</span>&#160;        ( o_indiv_fields,</div><div class="line"><a name="l02015"></a><span class="lineno"> 2015</span>&#160;            i_max_age ) = get_cohorts_sampling_params( lv_sample_values )</div><div class="line"><a name="l02016"></a><span class="lineno"> 2016</span>&#160;</div><div class="line"><a name="l02017"></a><span class="lineno"> 2017</span>&#160;        <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l02018"></a><span class="lineno"> 2018</span>&#160;<span class="stringliteral">        2017_07_21.  We are now using the newer classes GenepopFileSampleParamsCohorts</span></div><div class="line"><a name="l02019"></a><span class="lineno"> 2019</span>&#160;<span class="stringliteral">        and GenepopFileSamplerCohorts.</span></div><div class="line"><a name="l02020"></a><span class="lineno"> 2020</span>&#160;<span class="stringliteral">        &#39;&#39;&#39;</span></div><div class="line"><a name="l02021"></a><span class="lineno"> 2021</span>&#160;        o_sample_params=gps.GenepopFileParamsCohorts( \</div><div class="line"><a name="l02022"></a><span class="lineno"> 2022</span>&#160;                                        o_genepop_indiv_id_fields = o_indiv_fields,</div><div class="line"><a name="l02023"></a><span class="lineno"> 2023</span>&#160;                                        li_population_numbers = li_population_list,</div><div class="line"><a name="l02024"></a><span class="lineno"> 2024</span>&#160;                                        i_max_age=i_max_age,</div><div class="line"><a name="l02025"></a><span class="lineno"> 2025</span>&#160;                                        i_min_individuals_per_gen=i_min_pop_size,</div><div class="line"><a name="l02026"></a><span class="lineno"> 2026</span>&#160;                                        i_max_individuals_per_gen=i_max_pop_size,</div><div class="line"><a name="l02027"></a><span class="lineno"> 2027</span>&#160;                                        i_replicates=i_replicates )</div><div class="line"><a name="l02028"></a><span class="lineno"> 2028</span>&#160;</div><div class="line"><a name="l02029"></a><span class="lineno"> 2029</span>&#160;        o_sampler=gps.GenepopFileSamplerCohorts( o_genepopfile, o_sample_params )</div><div class="line"><a name="l02030"></a><span class="lineno"> 2030</span>&#160;    <span class="keywordflow">elif</span> s_sample_scheme==SAMPLE_BY_COHORT_PERCENTAGE:</div><div class="line"><a name="l02031"></a><span class="lineno"> 2031</span>&#160;</div><div class="line"><a name="l02032"></a><span class="lineno"> 2032</span>&#160;        <span class="comment">#this scheme has params identical to the regular &quot;cohorts&quot; scheme,</span></div><div class="line"><a name="l02033"></a><span class="lineno"> 2033</span>&#160;        <span class="comment">#with a hyphenated list percentage values appended to the max-age param, </span></div><div class="line"><a name="l02034"></a><span class="lineno"> 2034</span>&#160;        <span class="comment">#so we extract the last two values of the last param list:  </span></div><div class="line"><a name="l02035"></a><span class="lineno"> 2035</span>&#160;        </div><div class="line"><a name="l02036"></a><span class="lineno"> 2036</span>&#160;        COHORT_PARAM_LIST_DELIMITER=<span class="stringliteral">&quot;;&quot;</span></div><div class="line"><a name="l02037"></a><span class="lineno"> 2037</span>&#160;        COHORT_PERC_DELIM=<span class="stringliteral">&quot;-&quot;</span></div><div class="line"><a name="l02038"></a><span class="lineno"> 2038</span>&#160;        IDX_MAX_AGE=0</div><div class="line"><a name="l02039"></a><span class="lineno"> 2039</span>&#160;        IDX_PERC=1</div><div class="line"><a name="l02040"></a><span class="lineno"> 2040</span>&#160;</div><div class="line"><a name="l02041"></a><span class="lineno"> 2041</span>&#160;        s_last_param=lv_sample_values[ -1 ]</div><div class="line"><a name="l02042"></a><span class="lineno"> 2042</span>&#160;</div><div class="line"><a name="l02043"></a><span class="lineno"> 2043</span>&#160;        ls_age_and_percentage=s_last_param.split( \</div><div class="line"><a name="l02044"></a><span class="lineno"> 2044</span>&#160;                        COHORT_PARAM_LIST_DELIMITER )</div><div class="line"><a name="l02045"></a><span class="lineno"> 2045</span>&#160;</div><div class="line"><a name="l02046"></a><span class="lineno"> 2046</span>&#160;        s_max_age=ls_age_and_percentage[ IDX_MAX_AGE ]</div><div class="line"><a name="l02047"></a><span class="lineno"> 2047</span>&#160;        s_percentages=ls_age_and_percentage[ IDX_PERC ]</div><div class="line"><a name="l02048"></a><span class="lineno"> 2048</span>&#160;        </div><div class="line"><a name="l02049"></a><span class="lineno"> 2049</span>&#160;        <span class="comment">#we reassemble the list to match the expected</span></div><div class="line"><a name="l02050"></a><span class="lineno"> 2050</span>&#160;        <span class="comment">#cohort sampling params (without the percentage):</span></div><div class="line"><a name="l02051"></a><span class="lineno"> 2051</span>&#160;        lv_cohort_param_fields=lv_sample_values[ 0:len( lv_sample_values ) - 1 ] \</div><div class="line"><a name="l02052"></a><span class="lineno"> 2052</span>&#160;                                                            + [ s_max_age ]     </div><div class="line"><a name="l02053"></a><span class="lineno"> 2053</span>&#160;        <span class="comment">#Convert the percentages to a list of floats giving proportion:</span></div><div class="line"><a name="l02054"></a><span class="lineno"> 2054</span>&#160;        lf_proportions=[ old_div(float( s_percentage ),100.0) <span class="keywordflow">for</span> s_percentage  \</div><div class="line"><a name="l02055"></a><span class="lineno"> 2055</span>&#160;                            <span class="keywordflow">in</span> s_percentages.split( COHORT_PERC_DELIM ) ]</div><div class="line"><a name="l02056"></a><span class="lineno"> 2056</span>&#160;</div><div class="line"><a name="l02057"></a><span class="lineno"> 2057</span>&#160;        ( o_indiv_fields,</div><div class="line"><a name="l02058"></a><span class="lineno"> 2058</span>&#160;                i_max_age ) = get_cohorts_sampling_params( lv_cohort_param_fields )</div><div class="line"><a name="l02059"></a><span class="lineno"> 2059</span>&#160;</div><div class="line"><a name="l02060"></a><span class="lineno"> 2060</span>&#160;        <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l02061"></a><span class="lineno"> 2061</span>&#160;<span class="stringliteral">        2017_07_21. We are now using the new classes GenepopFileSampleParamsCohorts</span></div><div class="line"><a name="l02062"></a><span class="lineno"> 2062</span>&#160;<span class="stringliteral">        and GenepopFileSamplerCohorts.  The new params class requires that any</span></div><div class="line"><a name="l02063"></a><span class="lineno"> 2063</span>&#160;<span class="stringliteral">        subsample value list be submitted as a list of new objects of type</span></div><div class="line"><a name="l02064"></a><span class="lineno"> 2064</span>&#160;<span class="stringliteral">        CohortSamplingValue.  Note, too that the class GenepopFileSamplerCohorts</span></div><div class="line"><a name="l02065"></a><span class="lineno"> 2065</span>&#160;<span class="stringliteral">        handles both plain cohorts sampling and sampling with a list of proportions.</span></div><div class="line"><a name="l02066"></a><span class="lineno"> 2066</span>&#160;<span class="stringliteral">        &#39;&#39;&#39;</span></div><div class="line"><a name="l02067"></a><span class="lineno"> 2067</span>&#160;</div><div class="line"><a name="l02068"></a><span class="lineno"> 2068</span>&#160;        lo_cohort_sample_value_objects=[ \</div><div class="line"><a name="l02069"></a><span class="lineno"> 2069</span>&#160;                    gps.CohortSamplingValue( i_type=gps.CohortSamplingValue.TYPE_PROPORTION,</div><div class="line"><a name="l02070"></a><span class="lineno"> 2070</span>&#160;                                         v_value=f_proportion ) \</div><div class="line"><a name="l02071"></a><span class="lineno"> 2071</span>&#160;                                                 <span class="keywordflow">for</span> f_proportion <span class="keywordflow">in</span> lf_proportions ]</div><div class="line"><a name="l02072"></a><span class="lineno"> 2072</span>&#160;        o_sample_params=gps.GenepopFileSampleParamsCohorts( \</div><div class="line"><a name="l02073"></a><span class="lineno"> 2073</span>&#160;                                            o_genepop_indiv_id_fields = o_indiv_fields,</div><div class="line"><a name="l02074"></a><span class="lineno"> 2074</span>&#160;                                            li_population_numbers = li_population_list,</div><div class="line"><a name="l02075"></a><span class="lineno"> 2075</span>&#160;                                            i_max_age=i_max_age,</div><div class="line"><a name="l02076"></a><span class="lineno"> 2076</span>&#160;                                            lo_list_of_value_objects=lo_cohort_sample_value_objects,</div><div class="line"><a name="l02077"></a><span class="lineno"> 2077</span>&#160;                                            i_min_individuals_per_gen=i_min_pop_size,</div><div class="line"><a name="l02078"></a><span class="lineno"> 2078</span>&#160;                                            i_max_individuals_per_gen=i_max_pop_size,</div><div class="line"><a name="l02079"></a><span class="lineno"> 2079</span>&#160;                                            i_replicates=i_replicates )</div><div class="line"><a name="l02080"></a><span class="lineno"> 2080</span>&#160;</div><div class="line"><a name="l02081"></a><span class="lineno"> 2081</span>&#160;        o_sampler=gps.GenepopFileSamplerCohorts( o_genepopfile, o_sample_params )</div><div class="line"><a name="l02082"></a><span class="lineno"> 2082</span>&#160;</div><div class="line"><a name="l02083"></a><span class="lineno"> 2083</span>&#160;    <span class="keywordflow">elif</span> s_sample_scheme==SAMPLE_BY_COHORT_COUNT \</div><div class="line"><a name="l02084"></a><span class="lineno"> 2084</span>&#160;            <span class="keywordflow">or</span> s_sample_scheme==SAMPLE_BY_COHORT_COUNT_MAX: </div><div class="line"><a name="l02085"></a><span class="lineno"> 2085</span>&#160;        <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l02086"></a><span class="lineno"> 2086</span>&#160;<span class="stringliteral">        2017_07_21.  This scheme is added to allow a user to request</span></div><div class="line"><a name="l02087"></a><span class="lineno"> 2087</span>&#160;<span class="stringliteral">        a sample with a constant &quot;t&quot; such that each cohort in range,</span></div><div class="line"><a name="l02088"></a><span class="lineno"> 2088</span>&#160;<span class="stringliteral">        0 to max_age, will have &quot;t&quot; individuals (instead of the default</span></div><div class="line"><a name="l02089"></a><span class="lineno"> 2089</span>&#160;<span class="stringliteral">        even sampling where &quot;t&quot; = min( sum_invid( total_indiv) ).</span></div><div class="line"><a name="l02090"></a><span class="lineno"> 2090</span>&#160;<span class="stringliteral"></span></div><div class="line"><a name="l02091"></a><span class="lineno"> 2091</span>&#160;<span class="stringliteral">        2017_11_04 Adding scheme cohortsmax, which samples like cohortcount</span></div><div class="line"><a name="l02092"></a><span class="lineno"> 2092</span>&#160;<span class="stringliteral">        but when the smallest cohort to be samples has fewer indiv than</span></div><div class="line"><a name="l02093"></a><span class="lineno"> 2093</span>&#160;<span class="stringliteral">        the count, the entire cohort is sampled (instead of throwing error).</span></div><div class="line"><a name="l02094"></a><span class="lineno"> 2094</span>&#160;<span class="stringliteral">        &#39;&#39;&#39;</span></div><div class="line"><a name="l02095"></a><span class="lineno"> 2095</span>&#160;        COHORT_PARAM_LIST_DELIMITER=<span class="stringliteral">&quot;;&quot;</span></div><div class="line"><a name="l02096"></a><span class="lineno"> 2096</span>&#160;        COHORT_COUNTS_DELIM=<span class="stringliteral">&quot;-&quot;</span></div><div class="line"><a name="l02097"></a><span class="lineno"> 2097</span>&#160;        IDX_MAX_AGE=0</div><div class="line"><a name="l02098"></a><span class="lineno"> 2098</span>&#160;        IDX_COUNT=1</div><div class="line"><a name="l02099"></a><span class="lineno"> 2099</span>&#160;</div><div class="line"><a name="l02100"></a><span class="lineno"> 2100</span>&#160;        s_last_param=lv_sample_values[ -1 ]</div><div class="line"><a name="l02101"></a><span class="lineno"> 2101</span>&#160;</div><div class="line"><a name="l02102"></a><span class="lineno"> 2102</span>&#160;        ls_age_and_counts=s_last_param.split( \</div><div class="line"><a name="l02103"></a><span class="lineno"> 2103</span>&#160;                        COHORT_PARAM_LIST_DELIMITER )</div><div class="line"><a name="l02104"></a><span class="lineno"> 2104</span>&#160;</div><div class="line"><a name="l02105"></a><span class="lineno"> 2105</span>&#160;        s_max_age=ls_age_and_counts[ IDX_MAX_AGE ]</div><div class="line"><a name="l02106"></a><span class="lineno"> 2106</span>&#160;        s_counts=ls_age_and_counts[ IDX_COUNT ]</div><div class="line"><a name="l02107"></a><span class="lineno"> 2107</span>&#160;        </div><div class="line"><a name="l02108"></a><span class="lineno"> 2108</span>&#160;        <span class="comment">#we reassemble the list to match the expected</span></div><div class="line"><a name="l02109"></a><span class="lineno"> 2109</span>&#160;        <span class="comment">#cohort sampling params (without the counts):</span></div><div class="line"><a name="l02110"></a><span class="lineno"> 2110</span>&#160;        lv_cohort_param_fields=lv_sample_values[ 0:len( lv_sample_values ) - 1 ] \</div><div class="line"><a name="l02111"></a><span class="lineno"> 2111</span>&#160;                                                            + [ s_max_age ]     </div><div class="line"><a name="l02112"></a><span class="lineno"> 2112</span>&#160;        <span class="comment">#Convert the counts to a list of ints giving tot indivs per age class:</span></div><div class="line"><a name="l02113"></a><span class="lineno"> 2113</span>&#160;        li_counts=[ int( s_count ) <span class="keywordflow">for</span> s_count  \</div><div class="line"><a name="l02114"></a><span class="lineno"> 2114</span>&#160;                            <span class="keywordflow">in</span> s_counts.split( COHORT_COUNTS_DELIM ) ]</div><div class="line"><a name="l02115"></a><span class="lineno"> 2115</span>&#160;</div><div class="line"><a name="l02116"></a><span class="lineno"> 2116</span>&#160;        ( o_indiv_fields,</div><div class="line"><a name="l02117"></a><span class="lineno"> 2117</span>&#160;                i_max_age ) = get_cohorts_sampling_params( lv_cohort_param_fields )</div><div class="line"><a name="l02118"></a><span class="lineno"> 2118</span>&#160;</div><div class="line"><a name="l02119"></a><span class="lineno"> 2119</span>&#160;        <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l02120"></a><span class="lineno"> 2120</span>&#160;<span class="stringliteral">        2017_07_21. We are now using the new classes GenepopFileSampleParamsCohorts</span></div><div class="line"><a name="l02121"></a><span class="lineno"> 2121</span>&#160;<span class="stringliteral">        and GenepopFileSamplerCohorts.  The new params class requires that any</span></div><div class="line"><a name="l02122"></a><span class="lineno"> 2122</span>&#160;<span class="stringliteral">        subsample value list be submitted as a list of new objects of type</span></div><div class="line"><a name="l02123"></a><span class="lineno"> 2123</span>&#160;<span class="stringliteral">        CohortSamplingValue.  Note, too that the class GenepopFileSamplerCohorts</span></div><div class="line"><a name="l02124"></a><span class="lineno"> 2124</span>&#160;<span class="stringliteral">        handles both plain cohorts sampling and sampling with a list of proportions</span></div><div class="line"><a name="l02125"></a><span class="lineno"> 2125</span>&#160;<span class="stringliteral">        and/or counts.</span></div><div class="line"><a name="l02126"></a><span class="lineno"> 2126</span>&#160;<span class="stringliteral">        &#39;&#39;&#39;</span></div><div class="line"><a name="l02127"></a><span class="lineno"> 2127</span>&#160;</div><div class="line"><a name="l02128"></a><span class="lineno"> 2128</span>&#160;        lo_cohort_sample_value_objects=[]</div><div class="line"><a name="l02129"></a><span class="lineno"> 2129</span>&#160;        i_sampval_type=gps.CohortSamplingValue.TYPE_COUNT</div><div class="line"><a name="l02130"></a><span class="lineno"> 2130</span>&#160;        <span class="keywordflow">if</span> s_sample_scheme==SAMPLE_BY_COHORT_COUNT_MAX:</div><div class="line"><a name="l02131"></a><span class="lineno"> 2131</span>&#160;            i_sampval_type=gps.CohortSamplingValue.TYPE_COUNT_MAX</div><div class="line"><a name="l02132"></a><span class="lineno"> 2132</span>&#160;        <span class="comment">#end if we&#39;re using min( countval, available-count) instead</span></div><div class="line"><a name="l02133"></a><span class="lineno"> 2133</span>&#160;        <span class="comment">#of strict countval as sample size per ageclass.</span></div><div class="line"><a name="l02134"></a><span class="lineno"> 2134</span>&#160;</div><div class="line"><a name="l02135"></a><span class="lineno"> 2135</span>&#160;        <span class="keywordflow">for</span> i_count <span class="keywordflow">in</span> li_counts:</div><div class="line"><a name="l02136"></a><span class="lineno"> 2136</span>&#160;            o_this_sample_object=gps.CohortSamplingValue( \</div><div class="line"><a name="l02137"></a><span class="lineno"> 2137</span>&#160;                            i_type=i_sampval_type,</div><div class="line"><a name="l02138"></a><span class="lineno"> 2138</span>&#160;                            v_value=i_count )</div><div class="line"><a name="l02139"></a><span class="lineno"> 2139</span>&#160;            lo_cohort_sample_value_objects.append( o_this_sample_object )</div><div class="line"><a name="l02140"></a><span class="lineno"> 2140</span>&#160;        <span class="comment">#end for each count valuie</span></div><div class="line"><a name="l02141"></a><span class="lineno"> 2141</span>&#160;</div><div class="line"><a name="l02142"></a><span class="lineno"> 2142</span>&#160;        o_sample_params=gps.GenepopFileSampleParamsCohorts( \</div><div class="line"><a name="l02143"></a><span class="lineno"> 2143</span>&#160;                                            o_genepop_indiv_id_fields = o_indiv_fields,</div><div class="line"><a name="l02144"></a><span class="lineno"> 2144</span>&#160;                                            li_population_numbers = li_population_list,</div><div class="line"><a name="l02145"></a><span class="lineno"> 2145</span>&#160;                                            i_max_age=i_max_age,</div><div class="line"><a name="l02146"></a><span class="lineno"> 2146</span>&#160;                                            lo_list_of_value_objects=lo_cohort_sample_value_objects,</div><div class="line"><a name="l02147"></a><span class="lineno"> 2147</span>&#160;                                            i_min_individuals_per_gen=i_min_pop_size,</div><div class="line"><a name="l02148"></a><span class="lineno"> 2148</span>&#160;                                            i_max_individuals_per_gen=i_max_pop_size,</div><div class="line"><a name="l02149"></a><span class="lineno"> 2149</span>&#160;                                            i_replicates=i_replicates )</div><div class="line"><a name="l02150"></a><span class="lineno"> 2150</span>&#160;</div><div class="line"><a name="l02151"></a><span class="lineno"> 2151</span>&#160;        o_sampler=gps.GenepopFileSamplerCohorts( o_genepopfile, o_sample_params )</div><div class="line"><a name="l02152"></a><span class="lineno"> 2152</span>&#160;</div><div class="line"><a name="l02153"></a><span class="lineno"> 2153</span>&#160;                                </div><div class="line"><a name="l02154"></a><span class="lineno"> 2154</span>&#160;    <span class="keywordflow">elif</span> s_sample_scheme==SAMPLE_BY_RELATEDS:</div><div class="line"><a name="l02155"></a><span class="lineno"> 2155</span>&#160;</div><div class="line"><a name="l02156"></a><span class="lineno"> 2156</span>&#160;            ( o_indiv_fields,</div><div class="line"><a name="l02157"></a><span class="lineno"> 2157</span>&#160;                f_percent_relateds) = get_relateds_sampling_params( lv_sample_values )</div><div class="line"><a name="l02158"></a><span class="lineno"> 2158</span>&#160;</div><div class="line"><a name="l02159"></a><span class="lineno"> 2159</span>&#160;            o_sample_params=gps.GenepopFileSampleParamsAgeStructureRelateds( \</div><div class="line"><a name="l02160"></a><span class="lineno"> 2160</span>&#160;                                        o_genepop_indiv_id_fields = o_indiv_fields,</div><div class="line"><a name="l02161"></a><span class="lineno"> 2161</span>&#160;                                        li_population_numbers = li_population_list,</div><div class="line"><a name="l02162"></a><span class="lineno"> 2162</span>&#160;                                        f_percent_relateds_per_gen=f_percent_relateds,</div><div class="line"><a name="l02163"></a><span class="lineno"> 2163</span>&#160;                                        i_min_individuals_per_gen=i_min_pop_size,</div><div class="line"><a name="l02164"></a><span class="lineno"> 2164</span>&#160;                                        i_max_individuals_per_gen=i_max_pop_size,</div><div class="line"><a name="l02165"></a><span class="lineno"> 2165</span>&#160;                                        i_replicates=i_replicates )</div><div class="line"><a name="l02166"></a><span class="lineno"> 2166</span>&#160;</div><div class="line"><a name="l02167"></a><span class="lineno"> 2167</span>&#160;            o_sampler=gps.GenepopFileSamplerIndividualsAgeStructureRelateds( o_genepopfile,</div><div class="line"><a name="l02168"></a><span class="lineno"> 2168</span>&#160;                                                                    o_sample_params )</div><div class="line"><a name="l02169"></a><span class="lineno"> 2169</span>&#160;    <span class="keywordflow">else</span>:</div><div class="line"><a name="l02170"></a><span class="lineno"> 2170</span>&#160;</div><div class="line"><a name="l02171"></a><span class="lineno"> 2171</span>&#160;        s_msg=<span class="stringliteral">&quot;In pgdriveneestimator.py, def do_sample, &quot;</span> \</div><div class="line"><a name="l02172"></a><span class="lineno"> 2172</span>&#160;                + <span class="stringliteral">&quot;unknown sampling scheme: &quot;</span> \</div><div class="line"><a name="l02173"></a><span class="lineno"> 2173</span>&#160;                + s_sample_scheme + <span class="stringliteral">&quot;.&quot;</span></div><div class="line"><a name="l02174"></a><span class="lineno"> 2174</span>&#160;        <span class="keywordflow">raise</span> Exception( s_msg )</div><div class="line"><a name="l02175"></a><span class="lineno"> 2175</span>&#160;    <span class="comment">#end if percent, else removal, else criteria, else cohort, else relateds, else error</span></div><div class="line"><a name="l02176"></a><span class="lineno"> 2176</span>&#160;</div><div class="line"><a name="l02177"></a><span class="lineno"> 2177</span>&#160;    o_sampler.doSample()</div><div class="line"><a name="l02178"></a><span class="lineno"> 2178</span>&#160;</div><div class="line"><a name="l02179"></a><span class="lineno"> 2179</span>&#160;    <span class="keywordflow">return</span> li_population_list</div><div class="line"><a name="l02180"></a><span class="lineno"> 2180</span>&#160;<span class="comment">#end do_sample_individuals</span></div><div class="line"><a name="l02181"></a><span class="lineno"> 2181</span>&#160;</div></div><!-- fragment -->
</div>
</div>
<a class="anchor" id="a9544ec0a253ab2bcc1fcfbe1edb66120"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def agestrucne.pgdriveneestimator.do_sample_loci </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>o_genepopfile</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>s_loci_sampling_scheme</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>v_loci_sampling_scheme_param</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_min_loci_position</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_max_loci_position</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_min_total_loci</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_max_total_loci</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>li_population_list</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_loci_replicates</em> = <code>1</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">As of 2016_10_06, we have no loci sampling scheme
but we still subsample by loci order, start index, stop index
and a max total -- randomly drawn if the start/stop range
results in total loci over the max. 

Further, we currently only permit one loci subsample for
each individual subsampling session -- i.e. for all 
of a given sample replicate.

As of 2016_10_20, we've added new scheme "percent", but
still doing one sample in the genepop file per call
to this def, which now still means one loci sample will
be applied to all replicate indiv samples.

2016_11_06, revising to do multiple loci sampling.  The
scheme is, for each existing individual subsample,
we subsample the loci for each loci-subsampling parameter
value, and for each of these, for the number of loci
replicates given by def param i_loci_replicates.

2017_05_29.  We add a check at the start of this def
to validate the min and max loci position arguments.
if the range is invalid, we throw and error.  If the
min is greater than the total number of loci, we return
a 0, to indicate no pops were sampled.  Otherwise, 
we return the number of items in the li_population_numbers.</pre> 
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l02190">2190</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>
<div class="fragment"><div class="line"><a name="l02190"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html#a9544ec0a253ab2bcc1fcfbe1edb66120"> 2190</a></span>&#160;                        i_loci_replicates=1 ):</div><div class="line"><a name="l02191"></a><span class="lineno"> 2191</span>&#160;        </div><div class="line"><a name="l02192"></a><span class="lineno"> 2192</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l02193"></a><span class="lineno"> 2193</span>&#160;<span class="stringliteral">    As of 2016_10_06, we have no loci sampling scheme</span></div><div class="line"><a name="l02194"></a><span class="lineno"> 2194</span>&#160;<span class="stringliteral">    but we still subsample by loci order, start index, stop index</span></div><div class="line"><a name="l02195"></a><span class="lineno"> 2195</span>&#160;<span class="stringliteral">    and a max total -- randomly drawn if the start/stop range</span></div><div class="line"><a name="l02196"></a><span class="lineno"> 2196</span>&#160;<span class="stringliteral">    results in total loci over the max. </span></div><div class="line"><a name="l02197"></a><span class="lineno"> 2197</span>&#160;<span class="stringliteral"></span></div><div class="line"><a name="l02198"></a><span class="lineno"> 2198</span>&#160;<span class="stringliteral">    Further, we currently only permit one loci subsample for</span></div><div class="line"><a name="l02199"></a><span class="lineno"> 2199</span>&#160;<span class="stringliteral">    each individual subsampling session -- i.e. for all </span></div><div class="line"><a name="l02200"></a><span class="lineno"> 2200</span>&#160;<span class="stringliteral">    of a given sample replicate.</span></div><div class="line"><a name="l02201"></a><span class="lineno"> 2201</span>&#160;<span class="stringliteral"></span></div><div class="line"><a name="l02202"></a><span class="lineno"> 2202</span>&#160;<span class="stringliteral">    As of 2016_10_20, we&#39;ve added new scheme &quot;percent&quot;, but</span></div><div class="line"><a name="l02203"></a><span class="lineno"> 2203</span>&#160;<span class="stringliteral">    still doing one sample in the genepop file per call</span></div><div class="line"><a name="l02204"></a><span class="lineno"> 2204</span>&#160;<span class="stringliteral">    to this def, which now still means one loci sample will</span></div><div class="line"><a name="l02205"></a><span class="lineno"> 2205</span>&#160;<span class="stringliteral">    be applied to all replicate indiv samples.</span></div><div class="line"><a name="l02206"></a><span class="lineno"> 2206</span>&#160;<span class="stringliteral"></span></div><div class="line"><a name="l02207"></a><span class="lineno"> 2207</span>&#160;<span class="stringliteral">    2016_11_06, revising to do multiple loci sampling.  The</span></div><div class="line"><a name="l02208"></a><span class="lineno"> 2208</span>&#160;<span class="stringliteral">    scheme is, for each existing individual subsample,</span></div><div class="line"><a name="l02209"></a><span class="lineno"> 2209</span>&#160;<span class="stringliteral">    we subsample the loci for each loci-subsampling parameter</span></div><div class="line"><a name="l02210"></a><span class="lineno"> 2210</span>&#160;<span class="stringliteral">    value, and for each of these, for the number of loci</span></div><div class="line"><a name="l02211"></a><span class="lineno"> 2211</span>&#160;<span class="stringliteral">    replicates given by def param i_loci_replicates.</span></div><div class="line"><a name="l02212"></a><span class="lineno"> 2212</span>&#160;<span class="stringliteral"></span></div><div class="line"><a name="l02213"></a><span class="lineno"> 2213</span>&#160;<span class="stringliteral">    2017_05_29.  We add a check at the start of this def</span></div><div class="line"><a name="l02214"></a><span class="lineno"> 2214</span>&#160;<span class="stringliteral">    to validate the min and max loci position arguments.</span></div><div class="line"><a name="l02215"></a><span class="lineno"> 2215</span>&#160;<span class="stringliteral">    if the range is invalid, we throw and error.  If the</span></div><div class="line"><a name="l02216"></a><span class="lineno"> 2216</span>&#160;<span class="stringliteral">    min is greater than the total number of loci, we return</span></div><div class="line"><a name="l02217"></a><span class="lineno"> 2217</span>&#160;<span class="stringliteral">    a 0, to indicate no pops were sampled.  Otherwise, </span></div><div class="line"><a name="l02218"></a><span class="lineno"> 2218</span>&#160;<span class="stringliteral">    we return the number of items in the li_population_numbers.</span></div><div class="line"><a name="l02219"></a><span class="lineno"> 2219</span>&#160;<span class="stringliteral"></span></div><div class="line"><a name="l02220"></a><span class="lineno"> 2220</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l02221"></a><span class="lineno"> 2221</span>&#160;</div><div class="line"><a name="l02222"></a><span class="lineno"> 2222</span>&#160;    <span class="keywordflow">if</span> i_min_loci_position &lt; 1  \</div><div class="line"><a name="l02223"></a><span class="lineno"> 2223</span>&#160;            <span class="keywordflow">or</span> i_min_loci_position &gt; i_max_loci_position:</div><div class="line"><a name="l02224"></a><span class="lineno"> 2224</span>&#160;        s_msg=<span class="stringliteral">&quot;In pgdriveneestimator.py, def do_sample_loci, &quot;</span> \</div><div class="line"><a name="l02225"></a><span class="lineno"> 2225</span>&#160;                        + <span class="stringliteral">&quot;min loci number range, &quot;</span> \</div><div class="line"><a name="l02226"></a><span class="lineno"> 2226</span>&#160;                        + str( i_min_loci_position ) \</div><div class="line"><a name="l02227"></a><span class="lineno"> 2227</span>&#160;                        + <span class="stringliteral">&quot;, is less than one or &quot;</span> \</div><div class="line"><a name="l02228"></a><span class="lineno"> 2228</span>&#160;                        + <span class="stringliteral">&quot;greater than max pop number, &quot;</span> \</div><div class="line"><a name="l02229"></a><span class="lineno"> 2229</span>&#160;                        + str( i_max_loci_position ) + <span class="stringliteral">&quot;.&quot;</span> </div><div class="line"><a name="l02230"></a><span class="lineno"> 2230</span>&#160;        <span class="keywordflow">raise</span> Exception( s_msg )</div><div class="line"><a name="l02231"></a><span class="lineno"> 2231</span>&#160;    <span class="keywordflow">elif</span> i_min_loci_position &gt; o_genepopfile.loci_total:</div><div class="line"><a name="l02232"></a><span class="lineno"> 2232</span>&#160;            <span class="keywordflow">return</span> 0</div><div class="line"><a name="l02233"></a><span class="lineno"> 2233</span>&#160;    <span class="comment">#end if invalid range, else no loci in range</span></div><div class="line"><a name="l02234"></a><span class="lineno"> 2234</span>&#160;</div><div class="line"><a name="l02235"></a><span class="lineno"> 2235</span>&#160;</div><div class="line"><a name="l02236"></a><span class="lineno"> 2236</span>&#160;    <span class="keywordflow">for</span> s_indiv_subsample_tag <span class="keywordflow">in</span> o_genepopfile.indiv_subsample_tags:</div><div class="line"><a name="l02237"></a><span class="lineno"> 2237</span>&#160;</div><div class="line"><a name="l02238"></a><span class="lineno"> 2238</span>&#160;        <span class="keywordflow">if</span> s_loci_sampling_scheme == SAMPLE_LOCI_SCHEME_NONE:</div><div class="line"><a name="l02239"></a><span class="lineno"> 2239</span>&#160;</div><div class="line"><a name="l02240"></a><span class="lineno"> 2240</span>&#160;            <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l02241"></a><span class="lineno"> 2241</span>&#160;<span class="stringliteral">            Loci subsample tags consist of an individual subsample tag</span></div><div class="line"><a name="l02242"></a><span class="lineno"> 2242</span>&#160;<span class="stringliteral">            followed by &quot;_l_t_v_r&quot; where t gives the scheme type, v gives</span></div><div class="line"><a name="l02243"></a><span class="lineno"> 2243</span>&#160;<span class="stringliteral">            the salient param value for the scheme, and r gives the replicate.</span></div><div class="line"><a name="l02244"></a><span class="lineno"> 2244</span>&#160;<span class="stringliteral"></span></div><div class="line"><a name="l02245"></a><span class="lineno"> 2245</span>&#160;<span class="stringliteral">            </span></div><div class="line"><a name="l02246"></a><span class="lineno"> 2246</span>&#160;<span class="stringliteral">            2017_05_24. Bug fix.  The subsampling for this scheme was being</span></div><div class="line"><a name="l02247"></a><span class="lineno"> 2247</span>&#160;<span class="stringliteral">            done in a loop, for idx in range( i_replicates), but the</span></div><div class="line"><a name="l02248"></a><span class="lineno"> 2248</span>&#160;<span class="stringliteral">            i_replicates parameter in the sampler was hard-coded to 1.</span></div><div class="line"><a name="l02249"></a><span class="lineno"> 2249</span>&#160;<span class="stringliteral">            Hence only a single replicate was being done, with one</span></div><div class="line"><a name="l02250"></a><span class="lineno"> 2250</span>&#160;<span class="stringliteral">            loci subsample tag, but overwritten i_loci_replicates</span></div><div class="line"><a name="l02251"></a><span class="lineno"> 2251</span>&#160;<span class="stringliteral">            times.  The tsv file then had only one replicate of the sampling,</span></div><div class="line"><a name="l02252"></a><span class="lineno"> 2252</span>&#160;<span class="stringliteral">            corresponding to the last sampling.  Since the instance of </span></div><div class="line"><a name="l02253"></a><span class="lineno"> 2253</span>&#160;<span class="stringliteral">            GenepopFileSamplerLociByRangeAndTotal, in its doSample()</span></div><div class="line"><a name="l02254"></a><span class="lineno"> 2254</span>&#160;<span class="stringliteral">            def, if passed i_loci_replicates by its GenepopFileSampleParamsLoci</span></div><div class="line"><a name="l02255"></a><span class="lineno"> 2255</span>&#160;<span class="stringliteral">            object, will do each replicate and give it a uniq tag (as with</span></div><div class="line"><a name="l02256"></a><span class="lineno"> 2256</span>&#160;<span class="stringliteral">            other schemes).</span></div><div class="line"><a name="l02257"></a><span class="lineno"> 2257</span>&#160;<span class="stringliteral">            &#39;&#39;&#39;</span></div><div class="line"><a name="l02258"></a><span class="lineno"> 2258</span>&#160;            s_loci_subsample_tag_base=s_indiv_subsample_tag</div><div class="line"><a name="l02259"></a><span class="lineno"> 2259</span>&#160;</div><div class="line"><a name="l02260"></a><span class="lineno"> 2260</span>&#160;            o_sample_params=gps.GenepopFileSampleParamsLoci( \</div><div class="line"><a name="l02261"></a><span class="lineno"> 2261</span>&#160;                                        li_population_numbers=li_population_list,</div><div class="line"><a name="l02262"></a><span class="lineno"> 2262</span>&#160;                                        i_min_loci_position=i_min_loci_position,</div><div class="line"><a name="l02263"></a><span class="lineno"> 2263</span>&#160;                                        i_max_loci_position=i_max_loci_position,</div><div class="line"><a name="l02264"></a><span class="lineno"> 2264</span>&#160;                                        i_min_total_loci=i_min_total_loci,</div><div class="line"><a name="l02265"></a><span class="lineno"> 2265</span>&#160;                                        i_max_total_loci=i_max_total_loci,</div><div class="line"><a name="l02266"></a><span class="lineno"> 2266</span>&#160;                                        i_replicates=i_loci_replicates,</div><div class="line"><a name="l02267"></a><span class="lineno"> 2267</span>&#160;                                        s_sample_tag_base=s_loci_subsample_tag_base )</div><div class="line"><a name="l02268"></a><span class="lineno"> 2268</span>&#160;</div><div class="line"><a name="l02269"></a><span class="lineno"> 2269</span>&#160;            o_sampler=gps.GenepopFileSamplerLociByRangeAndTotal( o_genepopfile,</div><div class="line"><a name="l02270"></a><span class="lineno"> 2270</span>&#160;                                                                    o_sample_params )</div><div class="line"><a name="l02271"></a><span class="lineno"> 2271</span>&#160;</div><div class="line"><a name="l02272"></a><span class="lineno"> 2272</span>&#160;            o_sampler.doSample()</div><div class="line"><a name="l02273"></a><span class="lineno"> 2273</span>&#160;</div><div class="line"><a name="l02274"></a><span class="lineno"> 2274</span>&#160;        <span class="keywordflow">elif</span> s_loci_sampling_scheme == SAMPLE_LOCI_SCHEME_PERCENT:</div><div class="line"><a name="l02275"></a><span class="lineno"> 2275</span>&#160;</div><div class="line"><a name="l02276"></a><span class="lineno"> 2276</span>&#160;            s_loci_subsample_tag_base=s_indiv_subsample_tag</div><div class="line"><a name="l02277"></a><span class="lineno"> 2277</span>&#160;</div><div class="line"><a name="l02278"></a><span class="lineno"> 2278</span>&#160;            o_sample_params=gps.GenepopFileSampleParamsLoci( \</div><div class="line"><a name="l02279"></a><span class="lineno"> 2279</span>&#160;                                                    li_population_numbers=li_population_list,</div><div class="line"><a name="l02280"></a><span class="lineno"> 2280</span>&#160;                                                    i_min_loci_position=i_min_loci_position,</div><div class="line"><a name="l02281"></a><span class="lineno"> 2281</span>&#160;                                                    i_max_loci_position=i_max_loci_position,</div><div class="line"><a name="l02282"></a><span class="lineno"> 2282</span>&#160;                                                    i_min_total_loci=i_min_total_loci,</div><div class="line"><a name="l02283"></a><span class="lineno"> 2283</span>&#160;                                                    i_max_total_loci=i_max_total_loci,</div><div class="line"><a name="l02284"></a><span class="lineno"> 2284</span>&#160;                                                    lf_proportions=v_loci_sampling_scheme_param,</div><div class="line"><a name="l02285"></a><span class="lineno"> 2285</span>&#160;                                                    i_replicates=i_loci_replicates,</div><div class="line"><a name="l02286"></a><span class="lineno"> 2286</span>&#160;                                                    s_sample_tag_base=s_loci_subsample_tag_base )</div><div class="line"><a name="l02287"></a><span class="lineno"> 2287</span>&#160;</div><div class="line"><a name="l02288"></a><span class="lineno"> 2288</span>&#160;            o_sampler=gps.GenepopFileSamplerLociByRangeAndPercentage( o_genepopfile,</div><div class="line"><a name="l02289"></a><span class="lineno"> 2289</span>&#160;                                                                    o_sample_params )</div><div class="line"><a name="l02290"></a><span class="lineno"> 2290</span>&#160;            o_sampler.doSample()</div><div class="line"><a name="l02291"></a><span class="lineno"> 2291</span>&#160;</div><div class="line"><a name="l02292"></a><span class="lineno"> 2292</span>&#160;        <span class="keywordflow">elif</span> s_loci_sampling_scheme == SAMPLE_LOCI_CONSTANT_TOTAL:</div><div class="line"><a name="l02293"></a><span class="lineno"> 2293</span>&#160;</div><div class="line"><a name="l02294"></a><span class="lineno"> 2294</span>&#160;            s_loci_subsample_tag_base=s_indiv_subsample_tag</div><div class="line"><a name="l02295"></a><span class="lineno"> 2295</span>&#160;        </div><div class="line"><a name="l02296"></a><span class="lineno"> 2296</span>&#160;            <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l02297"></a><span class="lineno"> 2297</span>&#160;<span class="stringliteral">            For the constant sampling totals, the sampler ignores </span></div><div class="line"><a name="l02298"></a><span class="lineno"> 2298</span>&#160;<span class="stringliteral">            the passed values #for min and max loci total, and </span></div><div class="line"><a name="l02299"></a><span class="lineno"> 2299</span>&#160;<span class="stringliteral">            substitutes the current total for both min and max. Thus,</span></div><div class="line"><a name="l02300"></a><span class="lineno"> 2300</span>&#160;<span class="stringliteral">            for each N in the list of Ns, (and also for each replicate),  </span></div><div class="line"><a name="l02301"></a><span class="lineno"> 2301</span>&#160;<span class="stringliteral">            a random N will be drawn from the loci (positions). </span></div><div class="line"><a name="l02302"></a><span class="lineno"> 2302</span>&#160;<span class="stringliteral">            &#39;&#39;&#39;</span></div><div class="line"><a name="l02303"></a><span class="lineno"> 2303</span>&#160;</div><div class="line"><a name="l02304"></a><span class="lineno"> 2304</span>&#160;            o_sample_params=gps.GenepopFileSampleParamsLoci( \</div><div class="line"><a name="l02305"></a><span class="lineno"> 2305</span>&#160;                                        li_population_numbers=li_population_list,</div><div class="line"><a name="l02306"></a><span class="lineno"> 2306</span>&#160;                                        i_min_loci_position=i_min_loci_position,</div><div class="line"><a name="l02307"></a><span class="lineno"> 2307</span>&#160;                                        i_max_loci_position=i_max_loci_position,</div><div class="line"><a name="l02308"></a><span class="lineno"> 2308</span>&#160;                                        i_min_total_loci=i_min_total_loci,</div><div class="line"><a name="l02309"></a><span class="lineno"> 2309</span>&#160;                                        i_max_total_loci=i_max_total_loci,</div><div class="line"><a name="l02310"></a><span class="lineno"> 2310</span>&#160;                                        li_sample_totals=v_loci_sampling_scheme_param,</div><div class="line"><a name="l02311"></a><span class="lineno"> 2311</span>&#160;                                        i_replicates=i_loci_replicates,</div><div class="line"><a name="l02312"></a><span class="lineno"> 2312</span>&#160;                                        s_sample_tag_base=s_loci_subsample_tag_base )</div><div class="line"><a name="l02313"></a><span class="lineno"> 2313</span>&#160;</div><div class="line"><a name="l02314"></a><span class="lineno"> 2314</span>&#160;            o_sampler=gps.GenepopFileSamplerLociByRangeAndSampleTotalList( o_genepopfile,</div><div class="line"><a name="l02315"></a><span class="lineno"> 2315</span>&#160;                                                                            o_sample_params )</div><div class="line"><a name="l02316"></a><span class="lineno"> 2316</span>&#160;            o_sampler.doSample()</div><div class="line"><a name="l02317"></a><span class="lineno"> 2317</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l02318"></a><span class="lineno"> 2318</span>&#160;            s_msg=<span class="stringliteral">&quot;In pgdriveneestimator.py, def do_sample_loci, &quot;</span> \</div><div class="line"><a name="l02319"></a><span class="lineno"> 2319</span>&#160;                        + <span class="stringliteral">&quot;unknown loci sampling scheme: &quot;</span> \</div><div class="line"><a name="l02320"></a><span class="lineno"> 2320</span>&#160;                        + s_loci_sampling_scheme + <span class="stringliteral">&quot;.&quot;</span></div><div class="line"><a name="l02321"></a><span class="lineno"> 2321</span>&#160;            <span class="keywordflow">raise</span> Exception( s_msg )</div><div class="line"><a name="l02322"></a><span class="lineno"> 2322</span>&#160;        <span class="comment">#end if sampling scheme none else unknown</span></div><div class="line"><a name="l02323"></a><span class="lineno"> 2323</span>&#160;    <span class="comment">#end for each individual subsample tag</span></div><div class="line"><a name="l02324"></a><span class="lineno"> 2324</span>&#160;</div><div class="line"><a name="l02325"></a><span class="lineno"> 2325</span>&#160;    <span class="keywordflow">return</span> len( li_population_list )</div><div class="line"><a name="l02326"></a><span class="lineno"> 2326</span>&#160;<span class="comment">#end do_sample_loci</span></div><div class="line"><a name="l02327"></a><span class="lineno"> 2327</span>&#160;</div></div><!-- fragment -->
</div>
</div>
<a class="anchor" id="aaba26add66aaee9119650828c2803cf7"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def agestrucne.pgdriveneestimator.get_count_estimator_fields </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">2017_06_22.   This def is called by do_estimate 
when the estimator produces no output, so that
an entry can be made in the tsv file (ex. with NA
values for estimator values) with correct field count.
</pre> 
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l02959">2959</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>
<div class="fragment"><div class="line"><a name="l02959"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html#aaba26add66aaee9119650828c2803cf7"> 2959</a></span>&#160;<span class="keyword">def </span><a class="code" href="namespaceagestrucne_1_1pgdriveneestimator.html#aaba26add66aaee9119650828c2803cf7">get_count_estimator_fields</a>():</div><div class="line"><a name="l02960"></a><span class="lineno"> 2960</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l02961"></a><span class="lineno"> 2961</span>&#160;<span class="stringliteral">    2017_06_22.   This def is called by do_estimate </span></div><div class="line"><a name="l02962"></a><span class="lineno"> 2962</span>&#160;<span class="stringliteral">    when the estimator produces no output, so that</span></div><div class="line"><a name="l02963"></a><span class="lineno"> 2963</span>&#160;<span class="stringliteral">    an entry can be made in the tsv file (ex. with NA</span></div><div class="line"><a name="l02964"></a><span class="lineno"> 2964</span>&#160;<span class="stringliteral">    values for estimator values) with correct field count.</span></div><div class="line"><a name="l02965"></a><span class="lineno"> 2965</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l02966"></a><span class="lineno"> 2966</span>&#160;    i_count=0</div><div class="line"><a name="l02967"></a><span class="lineno"> 2967</span>&#160;    <span class="keywordflow">for</span> s_field <span class="keywordflow">in</span> pgout.PGOutputNeEstimator.OUTPUT_FIELDS:</div><div class="line"><a name="l02968"></a><span class="lineno"> 2968</span>&#160;        <span class="keywordflow">if</span> s_field <span class="keywordflow">not</span> <span class="keywordflow">in</span> NE_ESTIMATOR_OUTPUT_FIELDS_TO_SKIP:</div><div class="line"><a name="l02969"></a><span class="lineno"> 2969</span>&#160;            i_count += 1</div><div class="line"><a name="l02970"></a><span class="lineno"> 2970</span>&#160;        <span class="comment">#end if not a skipped field</span></div><div class="line"><a name="l02971"></a><span class="lineno"> 2971</span>&#160;    <span class="comment">#end for each output field</span></div><div class="line"><a name="l02972"></a><span class="lineno"> 2972</span>&#160;    <span class="keywordflow">return</span> i_count</div><div class="line"><a name="l02973"></a><span class="lineno"> 2973</span>&#160;<span class="comment">#end get_count_estimator_fields</span></div><div class="line"><a name="l02974"></a><span class="lineno"> 2974</span>&#160;</div><div class="ttc" id="namespaceagestrucne_1_1pgdriveneestimator_html_aaba26add66aaee9119650828c2803cf7"><div class="ttname"><a href="namespaceagestrucne_1_1pgdriveneestimator.html#aaba26add66aaee9119650828c2803cf7">agestrucne.pgdriveneestimator.get_count_estimator_fields</a></div><div class="ttdeci">def get_count_estimator_fields()</div><div class="ttdef"><b>Definition:</b> <a href="pgdriveneestimator_8py_source.html#l02959">pgdriveneestimator.py:2959</a></div></div>
</div><!-- fragment -->
</div>
</div>
<a class="anchor" id="aceaee36e8e89f8873e39642960bac265"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def agestrucne.pgdriveneestimator.get_genepop_file_list </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>s_genepop_files_arg</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">if invoked from the console the arg is a glob expression, but
but if invoked from the gui interface in pgguineestimator.py,
the arg is a list, such that eval will genreate a python
list of strings.

This def assumes that no user will invoke this mod from 
the command line with a glob expression, includeing a file name,
that can be evaluated as a python list (i.e. "[ \"item\", \"item\" ... ]</pre> 
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l00848">848</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>
<div class="fragment"><div class="line"><a name="l00848"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html#aceaee36e8e89f8873e39642960bac265">  848</a></span>&#160;<span class="keyword">def </span><a class="code" href="namespaceagestrucne_1_1pgdriveneestimator.html#aceaee36e8e89f8873e39642960bac265">get_genepop_file_list</a>( s_genepop_files_arg ):</div><div class="line"><a name="l00849"></a><span class="lineno">  849</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l00850"></a><span class="lineno">  850</span>&#160;<span class="stringliteral">    if invoked from the console the arg is a glob expression, but</span></div><div class="line"><a name="l00851"></a><span class="lineno">  851</span>&#160;<span class="stringliteral">    but if invoked from the gui interface in pgguineestimator.py,</span></div><div class="line"><a name="l00852"></a><span class="lineno">  852</span>&#160;<span class="stringliteral">    the arg is a list, such that eval will genreate a python</span></div><div class="line"><a name="l00853"></a><span class="lineno">  853</span>&#160;<span class="stringliteral">    list of strings.</span></div><div class="line"><a name="l00854"></a><span class="lineno">  854</span>&#160;<span class="stringliteral"></span></div><div class="line"><a name="l00855"></a><span class="lineno">  855</span>&#160;<span class="stringliteral">    This def assumes that no user will invoke this mod from </span></div><div class="line"><a name="l00856"></a><span class="lineno">  856</span>&#160;<span class="stringliteral">    the command line with a glob expression, includeing a file name,</span></div><div class="line"><a name="l00857"></a><span class="lineno">  857</span>&#160;<span class="stringliteral">    that can be evaluated as a python list (i.e. &quot;[ \&quot;item\&quot;, \&quot;item\&quot; ... ]</span></div><div class="line"><a name="l00858"></a><span class="lineno">  858</span>&#160;<span class="stringliteral">    </span></div><div class="line"><a name="l00859"></a><span class="lineno">  859</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l00860"></a><span class="lineno">  860</span>&#160;</div><div class="line"><a name="l00861"></a><span class="lineno">  861</span>&#160;    ls_files=<span class="keywordtype">None</span></div><div class="line"><a name="l00862"></a><span class="lineno">  862</span>&#160;</div><div class="line"><a name="l00863"></a><span class="lineno">  863</span>&#160;    <span class="keywordflow">try</span>:</div><div class="line"><a name="l00864"></a><span class="lineno">  864</span>&#160;</div><div class="line"><a name="l00865"></a><span class="lineno">  865</span>&#160;        ls_files=eval( s_genepop_files_arg )</div><div class="line"><a name="l00866"></a><span class="lineno">  866</span>&#160;</div><div class="line"><a name="l00867"></a><span class="lineno">  867</span>&#160;        <span class="keywordflow">if</span> type( ls_files ) != list:</div><div class="line"><a name="l00868"></a><span class="lineno">  868</span>&#160;            <span class="comment">#still could be a glob expression, for example,</span></div><div class="line"><a name="l00869"></a><span class="lineno">  869</span>&#160;            <span class="comment">#of a file name that python can eval as an int or float,</span></div><div class="line"><a name="l00870"></a><span class="lineno">  870</span>&#160;            <span class="comment">#so we&#39;ll just glob the orig:</span></div><div class="line"><a name="l00871"></a><span class="lineno">  871</span>&#160;</div><div class="line"><a name="l00872"></a><span class="lineno">  872</span>&#160;            ls_files=glob.glob( s_genepop_files_arg )</div><div class="line"><a name="l00873"></a><span class="lineno">  873</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l00874"></a><span class="lineno">  874</span>&#160;            <span class="comment">#make sure it&#39;s a list of strings</span></div><div class="line"><a name="l00875"></a><span class="lineno">  875</span>&#160;            <span class="keywordflow">if</span> not( file_name_list_is_all_strings( ls_files ) ):</div><div class="line"><a name="l00876"></a><span class="lineno">  876</span>&#160;                s_msg=<span class="stringliteral">&quot;In pgdriveneestimator.py, def get_genepop_file_list, &quot;</span> \</div><div class="line"><a name="l00877"></a><span class="lineno">  877</span>&#160;                        + <span class="stringliteral">&quot;genepop file arg evaluated as a list, but, not composed &quot;</span> \</div><div class="line"><a name="l00878"></a><span class="lineno">  878</span>&#160;                        + <span class="stringliteral">&quot;of strings: &quot;</span> + str( list ) + <span class="stringliteral">&quot;.&quot;</span></div><div class="line"><a name="l00879"></a><span class="lineno">  879</span>&#160;                <span class="keywordflow">raise</span> Exception( s_msg )</div><div class="line"><a name="l00880"></a><span class="lineno">  880</span>&#160;            <span class="comment">#end if not all strings in list</span></div><div class="line"><a name="l00881"></a><span class="lineno">  881</span>&#160;    <span class="keywordflow">except</span> ( NameError, SyntaxError ) <span class="keyword">as</span> evalfail:</div><div class="line"><a name="l00882"></a><span class="lineno">  882</span>&#160;</div><div class="line"><a name="l00883"></a><span class="lineno">  883</span>&#160;        <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l00884"></a><span class="lineno">  884</span>&#160;<span class="stringliteral">        this is caught if the eval fails due to </span></div><div class="line"><a name="l00885"></a><span class="lineno">  885</span>&#160;<span class="stringliteral">        the arg being a string.  Further, a syntax</span></div><div class="line"><a name="l00886"></a><span class="lineno">  886</span>&#160;<span class="stringliteral">        error results if the arg begins/ends in a wildcard</span></div><div class="line"><a name="l00887"></a><span class="lineno">  887</span>&#160;<span class="stringliteral">        &quot;*&quot; char -- as will happen often with a glob.</span></div><div class="line"><a name="l00888"></a><span class="lineno">  888</span>&#160;<span class="stringliteral"></span></div><div class="line"><a name="l00889"></a><span class="lineno">  889</span>&#160;<span class="stringliteral">        So for these eval failures we assume the arg </span></div><div class="line"><a name="l00890"></a><span class="lineno">  890</span>&#160;<span class="stringliteral">        is a glob expression</span></div><div class="line"><a name="l00891"></a><span class="lineno">  891</span>&#160;<span class="stringliteral">        &#39;&#39;&#39;</span></div><div class="line"><a name="l00892"></a><span class="lineno">  892</span>&#160;</div><div class="line"><a name="l00893"></a><span class="lineno">  893</span>&#160;        ls_files=glob.glob( s_genepop_files_arg )</div><div class="line"><a name="l00894"></a><span class="lineno">  894</span>&#160;</div><div class="line"><a name="l00895"></a><span class="lineno">  895</span>&#160;    <span class="keywordflow">except</span> Exception <span class="keyword">as</span> unknown_fail:</div><div class="line"><a name="l00896"></a><span class="lineno">  896</span>&#160;        s_errortype=type( unknown_fail ).__name__</div><div class="line"><a name="l00897"></a><span class="lineno">  897</span>&#160;        s_msg=<span class="stringliteral">&quot;In pgdriveneestimator.py, def get_genepop_file_list, &quot;</span> \</div><div class="line"><a name="l00898"></a><span class="lineno">  898</span>&#160;                        + <span class="stringliteral">&quot;gene pop file arg, &quot;</span> \</div><div class="line"><a name="l00899"></a><span class="lineno">  899</span>&#160;                        + str( s_genepop_files_arg ) \</div><div class="line"><a name="l00900"></a><span class="lineno">  900</span>&#160;                        + <span class="stringliteral">&quot;, eval operation produced &quot;</span> \</div><div class="line"><a name="l00901"></a><span class="lineno">  901</span>&#160;                        + <span class="stringliteral">&quot;error of type &quot;</span> + s_errortype \</div><div class="line"><a name="l00902"></a><span class="lineno">  902</span>&#160;                        + <span class="stringliteral">&quot;, with message, &quot;</span> + str( unknown_fail ) + <span class="stringliteral">&quot;.&quot;</span> </div><div class="line"><a name="l00903"></a><span class="lineno">  903</span>&#160;        <span class="keywordflow">raise</span> Exception( s_msg )</div><div class="line"><a name="l00904"></a><span class="lineno">  904</span>&#160;    <span class="comment">#end try eval, except name error, except other</span></div><div class="line"><a name="l00905"></a><span class="lineno">  905</span>&#160;</div><div class="line"><a name="l00906"></a><span class="lineno">  906</span>&#160;    <span class="keywordflow">return</span> ls_files</div><div class="line"><a name="l00907"></a><span class="lineno">  907</span>&#160;<span class="comment">#end get_genepop_file_list</span></div><div class="line"><a name="l00908"></a><span class="lineno">  908</span>&#160;</div><div class="ttc" id="namespaceagestrucne_1_1pgdriveneestimator_html_aceaee36e8e89f8873e39642960bac265"><div class="ttname"><a href="namespaceagestrucne_1_1pgdriveneestimator.html#aceaee36e8e89f8873e39642960bac265">agestrucne.pgdriveneestimator.get_genepop_file_list</a></div><div class="ttdeci">def get_genepop_file_list(s_genepop_files_arg)</div><div class="ttdef"><b>Definition:</b> <a href="pgdriveneestimator_8py_source.html#l00848">pgdriveneestimator.py:848</a></div></div>
</div><!-- fragment -->
</div>
</div>
<a class="anchor" id="a2f5f9109904c6af1b938aa800aba93df"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def agestrucne.pgdriveneestimator.get_invalid_file_names </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>ls_file_names</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">Currently only check is for file name
length (see comments above declaration of const 
MAX_GENEPOP_FILE_NAME_LENGTH.
</pre> 
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l00807">807</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>
<div class="fragment"><div class="line"><a name="l00807"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html#a2f5f9109904c6af1b938aa800aba93df">  807</a></span>&#160;<span class="keyword">def </span><a class="code" href="namespaceagestrucne_1_1pgdriveneestimator.html#a2f5f9109904c6af1b938aa800aba93df">get_invalid_file_names</a>( ls_file_names ):</div><div class="line"><a name="l00808"></a><span class="lineno">  808</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l00809"></a><span class="lineno">  809</span>&#160;<span class="stringliteral">    Currently only check is for file name</span></div><div class="line"><a name="l00810"></a><span class="lineno">  810</span>&#160;<span class="stringliteral">    length (see comments above declaration of const </span></div><div class="line"><a name="l00811"></a><span class="lineno">  811</span>&#160;<span class="stringliteral">    MAX_GENEPOP_FILE_NAME_LENGTH.</span></div><div class="line"><a name="l00812"></a><span class="lineno">  812</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l00813"></a><span class="lineno">  813</span>&#160;    MSG_FILE_TOO_LONG=<span class="stringliteral">&quot;File name longer than &quot;</span> \</div><div class="line"><a name="l00814"></a><span class="lineno">  814</span>&#160;            + str( MAX_GENEPOP_FILE_NAME_LENGTH ) \</div><div class="line"><a name="l00815"></a><span class="lineno">  815</span>&#160;            + <span class="stringliteral">&quot; characters.&quot;</span></div><div class="line"><a name="l00816"></a><span class="lineno">  816</span>&#160;</div><div class="line"><a name="l00817"></a><span class="lineno">  817</span>&#160;    ls_invalid_file_names_and_messages=[]</div><div class="line"><a name="l00818"></a><span class="lineno">  818</span>&#160;</div><div class="line"><a name="l00819"></a><span class="lineno">  819</span>&#160;    <span class="keywordflow">for</span> s_file_name <span class="keywordflow">in</span> ls_file_names:</div><div class="line"><a name="l00820"></a><span class="lineno">  820</span>&#160;        s_file_name_no_path=os.path.basename( s_file_name )</div><div class="line"><a name="l00821"></a><span class="lineno">  821</span>&#160;        i_len_file_name=len( s_file_name_no_path )</div><div class="line"><a name="l00822"></a><span class="lineno">  822</span>&#160;</div><div class="line"><a name="l00823"></a><span class="lineno">  823</span>&#160;        <span class="keywordflow">if</span> i_len_file_name &gt; MAX_GENEPOP_FILE_NAME_LENGTH:</div><div class="line"><a name="l00824"></a><span class="lineno">  824</span>&#160;            ls_invalid_file_names_and_messages.append( \</div><div class="line"><a name="l00825"></a><span class="lineno">  825</span>&#160;                    <span class="stringliteral">&quot;, &quot;</span>.join( [ s_file_name, MSG_FILE_TOO_LONG ] ) )</div><div class="line"><a name="l00826"></a><span class="lineno">  826</span>&#160;        <span class="comment">#and if file name too long</span></div><div class="line"><a name="l00827"></a><span class="lineno">  827</span>&#160;    <span class="comment">#end for each file name</span></div><div class="line"><a name="l00828"></a><span class="lineno">  828</span>&#160;</div><div class="line"><a name="l00829"></a><span class="lineno">  829</span>&#160;    <span class="keywordflow">return</span> ls_invalid_file_names_and_messages</div><div class="line"><a name="l00830"></a><span class="lineno">  830</span>&#160;<span class="comment">#end get_invalid_file_names</span></div><div class="line"><a name="l00831"></a><span class="lineno">  831</span>&#160;</div><div class="ttc" id="namespaceagestrucne_1_1pgdriveneestimator_html_a2f5f9109904c6af1b938aa800aba93df"><div class="ttname"><a href="namespaceagestrucne_1_1pgdriveneestimator.html#a2f5f9109904c6af1b938aa800aba93df">agestrucne.pgdriveneestimator.get_invalid_file_names</a></div><div class="ttdeci">def get_invalid_file_names(ls_file_names)</div><div class="ttdef"><b>Definition:</b> <a href="pgdriveneestimator_8py_source.html#l00807">pgdriveneestimator.py:807</a></div></div>
</div><!-- fragment -->
</div>
</div>
<a class="anchor" id="a6eea3cf1501fe02f36abfa30d6b4b7e3"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def agestrucne.pgdriveneestimator.get_loci_sample_val_and_rep_number_from_loci_sample_tag </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>s_loci_subsample_tag</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">The loci info in the tag follows an indiv tag prefix,
and has form "_l_t_v_r", with t giving the scheme,
v giving the scheme salient param value (ex percent val),
and r giving the replicate number.
</pre> 
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l02675">2675</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>
<div class="fragment"><div class="line"><a name="l02675"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html#a6eea3cf1501fe02f36abfa30d6b4b7e3"> 2675</a></span>&#160;<span class="keyword">def </span><a class="code" href="namespaceagestrucne_1_1pgdriveneestimator.html#a6eea3cf1501fe02f36abfa30d6b4b7e3">get_loci_sample_val_and_rep_number_from_loci_sample_tag</a>( s_loci_subsample_tag ):</div><div class="line"><a name="l02676"></a><span class="lineno"> 2676</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l02677"></a><span class="lineno"> 2677</span>&#160;<span class="stringliteral">    The loci info in the tag follows an indiv tag prefix,</span></div><div class="line"><a name="l02678"></a><span class="lineno"> 2678</span>&#160;<span class="stringliteral">    and has form &quot;_l_t_v_r&quot;, with t giving the scheme,</span></div><div class="line"><a name="l02679"></a><span class="lineno"> 2679</span>&#160;<span class="stringliteral">    v giving the scheme salient param value (ex percent val),</span></div><div class="line"><a name="l02680"></a><span class="lineno"> 2680</span>&#160;<span class="stringliteral">    and r giving the replicate number.</span></div><div class="line"><a name="l02681"></a><span class="lineno"> 2681</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l02682"></a><span class="lineno"> 2682</span>&#160;    <span class="comment">#Sample param val is 2nd to last item</span></div><div class="line"><a name="l02683"></a><span class="lineno"> 2683</span>&#160;    <span class="comment">#in the tag fields, while the replicate</span></div><div class="line"><a name="l02684"></a><span class="lineno"> 2684</span>&#160;    <span class="comment">#number is the last:</span></div><div class="line"><a name="l02685"></a><span class="lineno"> 2685</span>&#160;    IDX_LOCI_SAMPLE_VAL=-2</div><div class="line"><a name="l02686"></a><span class="lineno"> 2686</span>&#160;    IDX_LOCI_REP=-1</div><div class="line"><a name="l02687"></a><span class="lineno"> 2687</span>&#160;</div><div class="line"><a name="l02688"></a><span class="lineno"> 2688</span>&#160;    s_sample_val=<span class="keywordtype">None</span></div><div class="line"><a name="l02689"></a><span class="lineno"> 2689</span>&#160;    s_replicate_number=<span class="keywordtype">None</span></div><div class="line"><a name="l02690"></a><span class="lineno"> 2690</span>&#160;</div><div class="line"><a name="l02691"></a><span class="lineno"> 2691</span>&#160;    s_tag_delim=gps.TAG_DELIMITER</div><div class="line"><a name="l02692"></a><span class="lineno"> 2692</span>&#160;    </div><div class="line"><a name="l02693"></a><span class="lineno"> 2693</span>&#160;    ls_tag_fields=s_loci_subsample_tag.split( s_tag_delim )</div><div class="line"><a name="l02694"></a><span class="lineno"> 2694</span>&#160;</div><div class="line"><a name="l02695"></a><span class="lineno"> 2695</span>&#160;    s_sample_val=ls_tag_fields[ IDX_LOCI_SAMPLE_VAL ]</div><div class="line"><a name="l02696"></a><span class="lineno"> 2696</span>&#160;    s_replicate_number=ls_tag_fields[ IDX_LOCI_REP ]</div><div class="line"><a name="l02697"></a><span class="lineno"> 2697</span>&#160;</div><div class="line"><a name="l02698"></a><span class="lineno"> 2698</span>&#160;    <span class="keywordflow">return</span> s_sample_val, s_replicate_number</div><div class="line"><a name="l02699"></a><span class="lineno"> 2699</span>&#160;<span class="comment">#end get_loci_sample_val_and_rep_number_from_loci_sample_tag</span></div><div class="line"><a name="l02700"></a><span class="lineno"> 2700</span>&#160;</div><div class="ttc" id="namespaceagestrucne_1_1pgdriveneestimator_html_a6eea3cf1501fe02f36abfa30d6b4b7e3"><div class="ttname"><a href="namespaceagestrucne_1_1pgdriveneestimator.html#a6eea3cf1501fe02f36abfa30d6b4b7e3">agestrucne.pgdriveneestimator.get_loci_sample_val_and_rep_number_from_loci_sample_tag</a></div><div class="ttdeci">def get_loci_sample_val_and_rep_number_from_loci_sample_tag(s_loci_subsample_tag)</div><div class="ttdef"><b>Definition:</b> <a href="pgdriveneestimator_8py_source.html#l02675">pgdriveneestimator.py:2675</a></div></div>
</div><!-- fragment -->
</div>
</div>
<a class="anchor" id="a2efaad3207ead8feb0abb42b921ab402"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def agestrucne.pgdriveneestimator.get_mean_het_based_on_allele_freqs </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>o_genepopfile</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>s_pop_subsample_tag</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>s_individual_subsample_tag</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>s_loci_subsample_tag</em> = <code>None</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">2017_06_30.  Adds the mean heterozygosity calculated using allele frequencies of all loci,
for the pop (or pops) given by the pop subsample tag, which, as of this date, will always
be just a single pop, as this is the unit of estimation used in def do_estimate, and all loci
will be called using a subsample of the loci number ranges that the caller provided.
</pre> 
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l01319">1319</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>
<div class="fragment"><div class="line"><a name="l01319"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html#a2efaad3207ead8feb0abb42b921ab402"> 1319</a></span>&#160;                                                s_loci_subsample_tag=<span class="keywordtype">None</span> ):</div><div class="line"><a name="l01320"></a><span class="lineno"> 1320</span>&#160;</div><div class="line"><a name="l01321"></a><span class="lineno"> 1321</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l01322"></a><span class="lineno"> 1322</span>&#160;<span class="stringliteral">    2017_06_30.  Adds the mean heterozygosity calculated using allele frequencies of all loci,</span></div><div class="line"><a name="l01323"></a><span class="lineno"> 1323</span>&#160;<span class="stringliteral">    for the pop (or pops) given by the pop subsample tag, which, as of this date, will always</span></div><div class="line"><a name="l01324"></a><span class="lineno"> 1324</span>&#160;<span class="stringliteral">    be just a single pop, as this is the unit of estimation used in def do_estimate, and all loci</span></div><div class="line"><a name="l01325"></a><span class="lineno"> 1325</span>&#160;<span class="stringliteral">    will be called using a subsample of the loci number ranges that the caller provided.</span></div><div class="line"><a name="l01326"></a><span class="lineno"> 1326</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l01327"></a><span class="lineno"> 1327</span>&#160;    df_mean_hets_by_pop={}</div><div class="line"><a name="l01328"></a><span class="lineno"> 1328</span>&#160;</div><div class="line"><a name="l01329"></a><span class="lineno"> 1329</span>&#160;    o_loci_info=gpl.GenepopFileLociInfo( o_genepopfile, </div><div class="line"><a name="l01330"></a><span class="lineno"> 1330</span>&#160;                                            s_population_subsample_tag=s_pop_subsample_tag,</div><div class="line"><a name="l01331"></a><span class="lineno"> 1331</span>&#160;                                            s_individual_subsample_tag=s_individual_subsample_tag,</div><div class="line"><a name="l01332"></a><span class="lineno"> 1332</span>&#160;                                            s_loci_subsample_tag=s_loci_subsample_tag )</div><div class="line"><a name="l01333"></a><span class="lineno"> 1333</span>&#160;</div><div class="line"><a name="l01334"></a><span class="lineno"> 1334</span>&#160;    ddf_het_per_pop_per_loci=o_loci_info.getHeterozygosity()</div><div class="line"><a name="l01335"></a><span class="lineno"> 1335</span>&#160;</div><div class="line"><a name="l01336"></a><span class="lineno"> 1336</span>&#160;    <span class="keywordflow">for</span> i_pop <span class="keywordflow">in</span> ddf_het_per_pop_per_loci:</div><div class="line"><a name="l01337"></a><span class="lineno"> 1337</span>&#160;</div><div class="line"><a name="l01338"></a><span class="lineno"> 1338</span>&#160;        lf_het_values_this_pop=list( ddf_het_per_pop_per_loci[ i_pop ].values() )</div><div class="line"><a name="l01339"></a><span class="lineno"> 1339</span>&#160;        f_mean_het_value_this_pop=np.mean( lf_het_values_this_pop )</div><div class="line"><a name="l01340"></a><span class="lineno"> 1340</span>&#160;</div><div class="line"><a name="l01341"></a><span class="lineno"> 1341</span>&#160;        df_mean_hets_by_pop[ i_pop ]=f_mean_het_value_this_pop</div><div class="line"><a name="l01342"></a><span class="lineno"> 1342</span>&#160;        </div><div class="line"><a name="l01343"></a><span class="lineno"> 1343</span>&#160;    <span class="comment">#end for this pop</span></div><div class="line"><a name="l01344"></a><span class="lineno"> 1344</span>&#160;</div><div class="line"><a name="l01345"></a><span class="lineno"> 1345</span>&#160;    <span class="keywordflow">return</span> df_mean_hets_by_pop</div><div class="line"><a name="l01346"></a><span class="lineno"> 1346</span>&#160;</div><div class="line"><a name="l01347"></a><span class="lineno"> 1347</span>&#160;<span class="comment">#end get_mean_het_based_on_allele_freqs</span></div><div class="line"><a name="l01348"></a><span class="lineno"> 1348</span>&#160;</div></div><!-- fragment -->
</div>
</div>
<a class="anchor" id="abd76f8aa32c542ee3e90799c0e7b166a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def agestrucne.pgdriveneestimator.get_pops_in_file_outside_valid_size_range </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>o_genepopfile</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_min_pop_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_max_pop_size</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">checks size of every pop in the file
in case we need to skip some pops and 
accept others.

Returns a list of pop numbers, 1-based
indexed (as caller expects).
</pre> 
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l01207">1207</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>
<div class="fragment"><div class="line"><a name="l01207"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html#abd76f8aa32c542ee3e90799c0e7b166a"> 1207</a></span>&#160;                                                i_max_pop_size ):</div><div class="line"><a name="l01208"></a><span class="lineno"> 1208</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l01209"></a><span class="lineno"> 1209</span>&#160;<span class="stringliteral">    checks size of every pop in the file</span></div><div class="line"><a name="l01210"></a><span class="lineno"> 1210</span>&#160;<span class="stringliteral">    in case we need to skip some pops and </span></div><div class="line"><a name="l01211"></a><span class="lineno"> 1211</span>&#160;<span class="stringliteral">    accept others.</span></div><div class="line"><a name="l01212"></a><span class="lineno"> 1212</span>&#160;<span class="stringliteral"></span></div><div class="line"><a name="l01213"></a><span class="lineno"> 1213</span>&#160;<span class="stringliteral">    Returns a list of pop numbers, 1-based</span></div><div class="line"><a name="l01214"></a><span class="lineno"> 1214</span>&#160;<span class="stringliteral">    indexed (as caller expects).</span></div><div class="line"><a name="l01215"></a><span class="lineno"> 1215</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l01216"></a><span class="lineno"> 1216</span>&#160;</div><div class="line"><a name="l01217"></a><span class="lineno"> 1217</span>&#160;    li_popsizes=o_genepopfile.getIndividualCounts()</div><div class="line"><a name="l01218"></a><span class="lineno"> 1218</span>&#160;</div><div class="line"><a name="l01219"></a><span class="lineno"> 1219</span>&#160;    li_pops_outside_valid_size_range = []</div><div class="line"><a name="l01220"></a><span class="lineno"> 1220</span>&#160;    <span class="keywordflow">for</span> idx_pop <span class="keywordflow">in</span> range( len ( li_popsizes ) ):</div><div class="line"><a name="l01221"></a><span class="lineno"> 1221</span>&#160;        i_size_of_pop=li_popsizes[ idx_pop ] </div><div class="line"><a name="l01222"></a><span class="lineno"> 1222</span>&#160;</div><div class="line"><a name="l01223"></a><span class="lineno"> 1223</span>&#160;        <span class="keywordflow">if</span> i_size_of_pop &lt; i_min_pop_size \</div><div class="line"><a name="l01224"></a><span class="lineno"> 1224</span>&#160;                <span class="keywordflow">or</span> i_size_of_pop &gt; i_max_pop_size:</div><div class="line"><a name="l01225"></a><span class="lineno"> 1225</span>&#160;        </div><div class="line"><a name="l01226"></a><span class="lineno"> 1226</span>&#160;            li_pops_outside_valid_size_range.append( idx_pop + 1 )</div><div class="line"><a name="l01227"></a><span class="lineno"> 1227</span>&#160;</div><div class="line"><a name="l01228"></a><span class="lineno"> 1228</span>&#160;        <span class="comment">#end if subsample size too small</span></div><div class="line"><a name="l01229"></a><span class="lineno"> 1229</span>&#160;    <span class="comment">#end for each pop number</span></div><div class="line"><a name="l01230"></a><span class="lineno"> 1230</span>&#160;    <span class="keywordflow">return</span> li_pops_outside_valid_size_range</div><div class="line"><a name="l01231"></a><span class="lineno"> 1231</span>&#160;<span class="comment">#end get_pops_in_file_outside_valid_range</span></div><div class="line"><a name="l01232"></a><span class="lineno"> 1232</span>&#160;</div></div><!-- fragment -->
</div>
</div>
<a class="anchor" id="a78e7def8930f60b0f5021f16b356e1e3"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def agestrucne.pgdriveneestimator.get_string_values_ldne_ratio_and_bias_adjustment </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>o_ne_estimator</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>lv_results_list</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>f_nbne_ratio</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">Revised on 2017_04_17, so that the nbne ratio passed here has 
already been vetted in run_estimator, so the we can test it
and, if it is None, we return "None" and the original LDNe
estimate.  Otherwise, we 
</pre> 
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l01235">1235</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>
<div class="fragment"><div class="line"><a name="l01235"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html#a78e7def8930f60b0f5021f16b356e1e3"> 1235</a></span>&#160;                                                            f_nbne_ratio ):</div><div class="line"><a name="l01236"></a><span class="lineno"> 1236</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l01237"></a><span class="lineno"> 1237</span>&#160;<span class="stringliteral">    Revised on 2017_04_17, so that the nbne ratio passed here has </span></div><div class="line"><a name="l01238"></a><span class="lineno"> 1238</span>&#160;<span class="stringliteral">    already been vetted in run_estimator, so the we can test it</span></div><div class="line"><a name="l01239"></a><span class="lineno"> 1239</span>&#160;<span class="stringliteral">    and, if it is None, we return &quot;None&quot; and the original LDNe</span></div><div class="line"><a name="l01240"></a><span class="lineno"> 1240</span>&#160;<span class="stringliteral">    estimate.  Otherwise, we </span></div><div class="line"><a name="l01241"></a><span class="lineno"> 1241</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l01242"></a><span class="lineno"> 1242</span>&#160;</div><div class="line"><a name="l01243"></a><span class="lineno"> 1243</span>&#160;    s_bias_adjusted_ne_value=<span class="stringliteral">&quot;None&quot;</span></div><div class="line"><a name="l01244"></a><span class="lineno"> 1244</span>&#160;</div><div class="line"><a name="l01245"></a><span class="lineno"> 1245</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l01246"></a><span class="lineno"> 1246</span>&#160;<span class="stringliteral">    2018_04_04.  We add adjustments to CIs:</span></div><div class="line"><a name="l01247"></a><span class="lineno"> 1247</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l01248"></a><span class="lineno"> 1248</span>&#160;</div><div class="line"><a name="l01249"></a><span class="lineno"> 1249</span>&#160;    s_bias_adjusted_ne_value=<span class="stringliteral">&quot;None&quot;</span></div><div class="line"><a name="l01250"></a><span class="lineno"> 1250</span>&#160;    s_bias_adjusted_ci_hi_value=<span class="stringliteral">&quot;None&quot;</span></div><div class="line"><a name="l01251"></a><span class="lineno"> 1251</span>&#160;    s_bias_adjusted_ci_low_value=<span class="stringliteral">&quot;None&quot;</span></div><div class="line"><a name="l01252"></a><span class="lineno"> 1252</span>&#160;</div><div class="line"><a name="l01253"></a><span class="lineno"> 1253</span>&#160;</div><div class="line"><a name="l01254"></a><span class="lineno"> 1254</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l01255"></a><span class="lineno"> 1255</span>&#160;<span class="stringliteral">    We need the original Ne (Nb)  estimate</span></div><div class="line"><a name="l01256"></a><span class="lineno"> 1256</span>&#160;<span class="stringliteral">    in order to do the bias adjustment, or,</span></div><div class="line"><a name="l01257"></a><span class="lineno"> 1257</span>&#160;<span class="stringliteral">    if we cannot do the bias adjustment, </span></div><div class="line"><a name="l01258"></a><span class="lineno"> 1258</span>&#160;<span class="stringliteral">    (that is, we don&#39;t have an Nb/Ne ratio),</span></div><div class="line"><a name="l01259"></a><span class="lineno"> 1259</span>&#160;<span class="stringliteral">    then to write the original to the ne_est_adj</span></div><div class="line"><a name="l01260"></a><span class="lineno"> 1260</span>&#160;<span class="stringliteral">    column.  This allows the Viz plotter to</span></div><div class="line"><a name="l01261"></a><span class="lineno"> 1261</span>&#160;<span class="stringliteral">    use just the one column to plot the estimate,</span></div><div class="line"><a name="l01262"></a><span class="lineno"> 1262</span>&#160;<span class="stringliteral">    whether or not a bias adjustment was performed.</span></div><div class="line"><a name="l01263"></a><span class="lineno"> 1263</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l01264"></a><span class="lineno"> 1264</span>&#160;    i_col_ne_est=\</div><div class="line"><a name="l01265"></a><span class="lineno"> 1265</span>&#160;                o_ne_estimator.getOutputColumnNumberForFieldName( \</div><div class="line"><a name="l01266"></a><span class="lineno"> 1266</span>&#160;                                            COL_NAME_NEESTIMATOR_NE_ESTIMATE )</div><div class="line"><a name="l01267"></a><span class="lineno"> 1267</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l01268"></a><span class="lineno"> 1268</span>&#160;<span class="stringliteral">    2018_04_04.  We are adding adjusted values for the</span></div><div class="line"><a name="l01269"></a><span class="lineno"> 1269</span>&#160;<span class="stringliteral">    high and low jackknifed CI values.</span></div><div class="line"><a name="l01270"></a><span class="lineno"> 1270</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l01271"></a><span class="lineno"> 1271</span>&#160;    i_col_ci_hi=o_ne_estimator.getOutputColumnNumberForFieldName( \</div><div class="line"><a name="l01272"></a><span class="lineno"> 1272</span>&#160;                                            COL_NAME_NEESTIMATOR_CI_HI )</div><div class="line"><a name="l01273"></a><span class="lineno"> 1273</span>&#160;</div><div class="line"><a name="l01274"></a><span class="lineno"> 1274</span>&#160;    i_col_ci_low=o_ne_estimator.getOutputColumnNumberForFieldName( \</div><div class="line"><a name="l01275"></a><span class="lineno"> 1275</span>&#160;                                            COL_NAME_NEESTIMATOR_CI_LOW )</div><div class="line"><a name="l01276"></a><span class="lineno"> 1276</span>&#160;</div><div class="line"><a name="l01277"></a><span class="lineno"> 1277</span>&#160;    f_unadjusted_estimate=float( lv_results_list[ i_col_ne_est ] )</div><div class="line"><a name="l01278"></a><span class="lineno"> 1278</span>&#160;    f_ci_hi=float( lv_results_list[ i_col_ci_hi ] )</div><div class="line"><a name="l01279"></a><span class="lineno"> 1279</span>&#160;    f_ci_low=float( lv_results_list[ i_col_ci_low ] )</div><div class="line"><a name="l01280"></a><span class="lineno"> 1280</span>&#160;</div><div class="line"><a name="l01281"></a><span class="lineno"> 1281</span>&#160;    <span class="keywordflow">if</span> f_nbne_ratio <span class="keywordflow">is</span> <span class="keywordflow">not</span> <span class="keywordtype">None</span>:</div><div class="line"><a name="l01282"></a><span class="lineno"> 1282</span>&#160;</div><div class="line"><a name="l01283"></a><span class="lineno"> 1283</span>&#160;        f_adj_estimate=<a class="code" href="namespaceagestrucne_1_1pgdriveneestimator.html#aaf2e5f67e91fb3bff9ad32c7dd0ccf53">do_ldne_bias_adjustment</a>( f_unadjusted_estimate,</div><div class="line"><a name="l01284"></a><span class="lineno"> 1284</span>&#160;                                                            f_nbne_ratio )</div><div class="line"><a name="l01285"></a><span class="lineno"> 1285</span>&#160;        f_ci_hi_adjusted=<a class="code" href="namespaceagestrucne_1_1pgdriveneestimator.html#aaf2e5f67e91fb3bff9ad32c7dd0ccf53">do_ldne_bias_adjustment</a>( f_ci_hi,</div><div class="line"><a name="l01286"></a><span class="lineno"> 1286</span>&#160;                                                        f_nbne_ratio )</div><div class="line"><a name="l01287"></a><span class="lineno"> 1287</span>&#160;        f_ci_low_adjusted=<a class="code" href="namespaceagestrucne_1_1pgdriveneestimator.html#aaf2e5f67e91fb3bff9ad32c7dd0ccf53">do_ldne_bias_adjustment</a>( f_ci_low,</div><div class="line"><a name="l01288"></a><span class="lineno"> 1288</span>&#160;                                                        f_nbne_ratio )</div><div class="line"><a name="l01289"></a><span class="lineno"> 1289</span>&#160;</div><div class="line"><a name="l01290"></a><span class="lineno"> 1290</span>&#160;        s_bias_adjusted_ne_value=str( round( f_adj_estimate, 3 ) ) </div><div class="line"><a name="l01291"></a><span class="lineno"> 1291</span>&#160;        s_bias_adjusted_ci_hi_value=str( round( f_ci_hi_adjusted, 3 ) ) </div><div class="line"><a name="l01292"></a><span class="lineno"> 1292</span>&#160;        s_bias_adjusted_ci_low_value=str( round( f_ci_low_adjusted, 3 ) ) </div><div class="line"><a name="l01293"></a><span class="lineno"> 1293</span>&#160;    <span class="keywordflow">else</span>:</div><div class="line"><a name="l01294"></a><span class="lineno"> 1294</span>&#160;        <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l01295"></a><span class="lineno"> 1295</span>&#160;<span class="stringliteral">        2017_02_15.  Instead of writing &quot;None&quot; as the bias-adjusted estimate, </span></div><div class="line"><a name="l01296"></a><span class="lineno"> 1296</span>&#160;<span class="stringliteral">        when we have no bias adjustment, we&#39;ll enter the original estimate.  </span></div><div class="line"><a name="l01297"></a><span class="lineno"> 1297</span>&#160;<span class="stringliteral">        This allows Brian T to find the Nb (Ne) values to plot in the new </span></div><div class="line"><a name="l01298"></a><span class="lineno"> 1298</span>&#160;<span class="stringliteral">        ne_est_adj column instead of the original est_ne column.  </span></div><div class="line"><a name="l01299"></a><span class="lineno"> 1299</span>&#160;<span class="stringliteral">        By using original estimates when no bias adjustment was done,</span></div><div class="line"><a name="l01300"></a><span class="lineno"> 1300</span>&#160;<span class="stringliteral">        he does not have to use conditional language to find </span></div><div class="line"><a name="l01301"></a><span class="lineno"> 1301</span>&#160;<span class="stringliteral">        his Ne estimate.</span></div><div class="line"><a name="l01302"></a><span class="lineno"> 1302</span>&#160;<span class="stringliteral">        &#39;&#39;&#39;</span></div><div class="line"><a name="l01303"></a><span class="lineno"> 1303</span>&#160;        s_bias_adjusted_ne_value=str( f_unadjusted_estimate )</div><div class="line"><a name="l01304"></a><span class="lineno"> 1304</span>&#160;        s_bias_adjusted_ci_hi_value=str( f_ci_hi )</div><div class="line"><a name="l01305"></a><span class="lineno"> 1305</span>&#160;        s_bias_adjusted_ci_low_value=str( f_ci_low  )</div><div class="line"><a name="l01306"></a><span class="lineno"> 1306</span>&#160;</div><div class="line"><a name="l01307"></a><span class="lineno"> 1307</span>&#160;    <span class="comment">#end if we have a nb/ne ratio, do bias adjustment </span></div><div class="line"><a name="l01308"></a><span class="lineno"> 1308</span>&#160;    <span class="keywordflow">return</span> str( f_nbne_ratio ), s_bias_adjusted_ne_value, s_bias_adjusted_ci_hi_value, s_bias_adjusted_ci_low_value</div><div class="line"><a name="l01309"></a><span class="lineno"> 1309</span>&#160;<span class="comment">#end get_string_values_ldne_ratio_and_bias_adjustment</span></div><div class="line"><a name="l01310"></a><span class="lineno"> 1310</span>&#160;</div><div class="ttc" id="namespaceagestrucne_1_1pgdriveneestimator_html_aaf2e5f67e91fb3bff9ad32c7dd0ccf53"><div class="ttname"><a href="namespaceagestrucne_1_1pgdriveneestimator.html#aaf2e5f67e91fb3bff9ad32c7dd0ccf53">agestrucne.pgdriveneestimator.do_ldne_bias_adjustment</a></div><div class="ttdeci">def do_ldne_bias_adjustment(f_ldne_estimate, f_nbne_ratio)</div><div class="ttdef"><b>Definition:</b> <a href="pgdriveneestimator_8py_source.html#l01349">pgdriveneestimator.py:1349</a></div></div>
</div><!-- fragment -->
</div>
</div>
<a class="anchor" id="adebe964cadd26d7aca365586ef17e3b1"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def agestrucne.pgdriveneestimator.get_subsample_genepop_file_name </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>s_original_genepop_file_name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>s_loci_subsample_tag</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>s_this_pop_sample_name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>s_temporary_directory</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">Was originally simply a matter of replacing, in the original genepop file,
dots with underscores, but Windows revealed a case that needed attention 
(see "replace" below).
</pre> 
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l02331">2331</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>
<div class="fragment"><div class="line"><a name="l02331"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html#adebe964cadd26d7aca365586ef17e3b1"> 2331</a></span>&#160;                                                    s_temporary_directory ):</div><div class="line"><a name="l02332"></a><span class="lineno"> 2332</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l02333"></a><span class="lineno"> 2333</span>&#160;<span class="stringliteral">    Was originally simply a matter of replacing, in the original genepop file,</span></div><div class="line"><a name="l02334"></a><span class="lineno"> 2334</span>&#160;<span class="stringliteral">    dots with underscores, but Windows revealed a case that needed attention </span></div><div class="line"><a name="l02335"></a><span class="lineno"> 2335</span>&#160;<span class="stringliteral">    (see &quot;replace&quot; below).</span></div><div class="line"><a name="l02336"></a><span class="lineno"> 2336</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l02337"></a><span class="lineno"> 2337</span>&#160;</div><div class="line"><a name="l02338"></a><span class="lineno"> 2338</span>&#160;    s_genepop_file_subsample=s_original_genepop_file_name \</div><div class="line"><a name="l02339"></a><span class="lineno"> 2339</span>&#160;                                + INPUT_FILE_DOT_CHAR_REPLACEMENT \</div><div class="line"><a name="l02340"></a><span class="lineno"> 2340</span>&#160;                                +  s_loci_subsample_tag \</div><div class="line"><a name="l02341"></a><span class="lineno"> 2341</span>&#160;                                + INPUT_FILE_DOT_CHAR_REPLACEMENT \</div><div class="line"><a name="l02342"></a><span class="lineno"> 2342</span>&#160;                                + <span class="stringliteral">&quot;g&quot;</span> + INPUT_FILE_DOT_CHAR_REPLACEMENT \</div><div class="line"><a name="l02343"></a><span class="lineno"> 2343</span>&#160;                                + s_this_pop_sample_name</div><div class="line"><a name="l02344"></a><span class="lineno"> 2344</span>&#160;    </div><div class="line"><a name="l02345"></a><span class="lineno"> 2345</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l02346"></a><span class="lineno"> 2346</span>&#160;<span class="stringliteral">    Note: In the original file name, if a period is left to the left of that</span></div><div class="line"><a name="l02347"></a><span class="lineno"> 2347</span>&#160;<span class="stringliteral">    used do delimit the file extension, then NeEstimator, when naming</span></div><div class="line"><a name="l02348"></a><span class="lineno"> 2348</span>&#160;<span class="stringliteral">    &quot;*NoDat.txt&quot; file, will replace the left-most period and all text that</span></div><div class="line"><a name="l02349"></a><span class="lineno"> 2349</span>&#160;<span class="stringliteral">    follows it,  with &quot;NoDat.txt&quot; -- this then removes any percent and replicate</span></div><div class="line"><a name="l02350"></a><span class="lineno"> 2350</span>&#160;<span class="stringliteral">    number text I&#39;m using to make files unique, and so creates a non-uniq</span></div><div class="line"><a name="l02351"></a><span class="lineno"> 2351</span>&#160;<span class="stringliteral">    *NoDat.txt file name, which, when one process removes it, and another goes</span></div><div class="line"><a name="l02352"></a><span class="lineno"> 2352</span>&#160;<span class="stringliteral">    looking for it , throws an exception -- (an exception from the</span></div><div class="line"><a name="l02353"></a><span class="lineno"> 2353</span>&#160;<span class="stringliteral">    multiprocessing.Pools async mapping -- file-not-found, but no trace since it</span></div><div class="line"><a name="l02354"></a><span class="lineno"> 2354</span>&#160;<span class="stringliteral">    occurs in a different process from the main).  This peculiar mangling of a</span></div><div class="line"><a name="l02355"></a><span class="lineno"> 2355</span>&#160;<span class="stringliteral">    users input file -- was preventing me in previous versions, from deleting</span></div><div class="line"><a name="l02356"></a><span class="lineno"> 2356</span>&#160;<span class="stringliteral">    NoDat files, and also I think meant that only the last process in a given</span></div><div class="line"><a name="l02357"></a><span class="lineno"> 2357</span>&#160;<span class="stringliteral">    group of replicates (share same genepop file and proportion sampled) to use</span></div><div class="line"><a name="l02358"></a><span class="lineno"> 2358</span>&#160;<span class="stringliteral">    the NoDAt file got to write it&#39;s NoDat info.  Removing all periods from the</span></div><div class="line"><a name="l02359"></a><span class="lineno"> 2359</span>&#160;<span class="stringliteral">    input file to the NeEstimator program, I belive, completely solves the</span></div><div class="line"><a name="l02360"></a><span class="lineno"> 2360</span>&#160;<span class="stringliteral">    problems.</span></div><div class="line"><a name="l02361"></a><span class="lineno"> 2361</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l02362"></a><span class="lineno"> 2362</span>&#160;</div><div class="line"><a name="l02363"></a><span class="lineno"> 2363</span>&#160;    s_basename=os.path.basename( s_genepop_file_subsample )</div><div class="line"><a name="l02364"></a><span class="lineno"> 2364</span>&#160;</div><div class="line"><a name="l02365"></a><span class="lineno"> 2365</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l02366"></a><span class="lineno"> 2366</span>&#160;<span class="stringliteral">    2017_03_27.  We replace the path given by the original</span></div><div class="line"><a name="l02367"></a><span class="lineno"> 2367</span>&#160;<span class="stringliteral">    genepop file with a temporary directory, if available.</span></div><div class="line"><a name="l02368"></a><span class="lineno"> 2368</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l02369"></a><span class="lineno"> 2369</span>&#160;    s_dirs=<span class="keywordtype">None</span></div><div class="line"><a name="l02370"></a><span class="lineno"> 2370</span>&#160;</div><div class="line"><a name="l02371"></a><span class="lineno"> 2371</span>&#160;    <span class="keywordflow">if</span> s_temporary_directory <span class="keywordflow">is</span> <span class="keywordtype">None</span>:</div><div class="line"><a name="l02372"></a><span class="lineno"> 2372</span>&#160;        s_dirs=os.path.dirname( s_genepop_file_subsample )</div><div class="line"><a name="l02373"></a><span class="lineno"> 2373</span>&#160;    <span class="keywordflow">else</span>:</div><div class="line"><a name="l02374"></a><span class="lineno"> 2374</span>&#160;        s_dirs=s_temporary_directory</div><div class="line"><a name="l02375"></a><span class="lineno"> 2375</span>&#160;    <span class="comment">#end if we have no temp directory, else use it instead of genepop file dir</span></div><div class="line"><a name="l02376"></a><span class="lineno"> 2376</span>&#160;</div><div class="line"><a name="l02377"></a><span class="lineno"> 2377</span>&#160;    s_subsample_file_basename=s_basename.replace( <span class="stringliteral">&quot;.&quot;</span>, INPUT_FILE_DOT_CHAR_REPLACEMENT )</div><div class="line"><a name="l02378"></a><span class="lineno"> 2378</span>&#160;</div><div class="line"><a name="l02379"></a><span class="lineno"> 2379</span>&#160;    s_genepop_file_subsample=os.path.join( s_dirs, s_subsample_file_basename )</div><div class="line"><a name="l02380"></a><span class="lineno"> 2380</span>&#160;    </div><div class="line"><a name="l02381"></a><span class="lineno"> 2381</span>&#160;    <span class="comment">#remmed out -- problems when the orig file has path that includes </span></div><div class="line"><a name="l02382"></a><span class="lineno"> 2382</span>&#160;    <span class="comment">#directories with dots in their names, so the wholesale replace</span></div><div class="line"><a name="l02383"></a><span class="lineno"> 2383</span>&#160;    <span class="comment">#then includes directories that are renamed and (no doubt) non-existent</span></div><div class="line"><a name="l02384"></a><span class="lineno"> 2384</span>&#160;</div><div class="line"><a name="l02385"></a><span class="lineno"> 2385</span>&#160;    <span class="comment">#s_genepop_file_subsample=s_genepop_file_subsample.replace( &quot;.&quot;, INPUT_FILE_DOT_CHAR_REPLACEMENT )</span></div><div class="line"><a name="l02386"></a><span class="lineno"> 2386</span>&#160;    </div><div class="line"><a name="l02387"></a><span class="lineno"> 2387</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l02388"></a><span class="lineno"> 2388</span>&#160;<span class="stringliteral">    problem seen in Windows, in which when this mod is invoked as __main__ from a shell</span></div><div class="line"><a name="l02389"></a><span class="lineno"> 2389</span>&#160;<span class="stringliteral">    then &quot;.\\&quot; is appended to (at least) one genepop file name.  This then is mangled by</span></div><div class="line"><a name="l02390"></a><span class="lineno"> 2390</span>&#160;<span class="stringliteral">    the above replace, so we restore the leading dot char:</span></div><div class="line"><a name="l02391"></a><span class="lineno"> 2391</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l02392"></a><span class="lineno"> 2392</span>&#160;</div><div class="line"><a name="l02393"></a><span class="lineno"> 2393</span>&#160;    <span class="keywordflow">if</span> s_genepop_file_subsample.startswith(  <span class="stringliteral">&quot;_&quot;</span> ):</div><div class="line"><a name="l02394"></a><span class="lineno"> 2394</span>&#160;        s_genepop_file_subsample=<span class="stringliteral">&quot;.&quot;</span> + s_genepop_file_subsample[1:] </div><div class="line"><a name="l02395"></a><span class="lineno"> 2395</span>&#160;    <span class="comment">#end if path was mangled by replace</span></div><div class="line"><a name="l02396"></a><span class="lineno"> 2396</span>&#160;</div><div class="line"><a name="l02397"></a><span class="lineno"> 2397</span>&#160;    <span class="keywordflow">return</span> s_genepop_file_subsample</div><div class="line"><a name="l02398"></a><span class="lineno"> 2398</span>&#160;<span class="comment">#end get_subsample_genepop_file_name</span></div><div class="line"><a name="l02399"></a><span class="lineno"> 2399</span>&#160;</div></div><!-- fragment -->
</div>
</div>
<a class="anchor" id="a2ea5983385f4ef857bcf935d904c59b1"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def agestrucne.pgdriveneestimator.make_subsample_genepop_file_name </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>s_original_genepop_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>s_temporary_directory</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_genepop_file_count</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_indiv_sample_count</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i_loci_subsample_count</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">2017_04_03.  This def replaces def get_subsample_genepop_file_name,
in order to shorten the intermediate genepop file names, to address
the path length limitations of Windows.
</pre> 
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l02404">2404</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>
<div class="fragment"><div class="line"><a name="l02404"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html#a2ea5983385f4ef857bcf935d904c59b1"> 2404</a></span>&#160;                                                        i_loci_subsample_count ):</div><div class="line"><a name="l02405"></a><span class="lineno"> 2405</span>&#160;        <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l02406"></a><span class="lineno"> 2406</span>&#160;<span class="stringliteral">        2017_04_03.  This def replaces def get_subsample_genepop_file_name,</span></div><div class="line"><a name="l02407"></a><span class="lineno"> 2407</span>&#160;<span class="stringliteral">        in order to shorten the intermediate genepop file names, to address</span></div><div class="line"><a name="l02408"></a><span class="lineno"> 2408</span>&#160;<span class="stringliteral">        the path length limitations of Windows.</span></div><div class="line"><a name="l02409"></a><span class="lineno"> 2409</span>&#160;<span class="stringliteral">        &#39;&#39;&#39;</span></div><div class="line"><a name="l02410"></a><span class="lineno"> 2410</span>&#160;        s_parent_dir=<span class="keywordtype">None</span></div><div class="line"><a name="l02411"></a><span class="lineno"> 2411</span>&#160;</div><div class="line"><a name="l02412"></a><span class="lineno"> 2412</span>&#160;        <span class="keywordflow">if</span> s_temporary_directory <span class="keywordflow">is</span> <span class="keywordtype">None</span>:</div><div class="line"><a name="l02413"></a><span class="lineno"> 2413</span>&#160;            s_parent_dir=os.path.dirname( s_original_genepop_file )</div><div class="line"><a name="l02414"></a><span class="lineno"> 2414</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l02415"></a><span class="lineno"> 2415</span>&#160;</div><div class="line"><a name="l02416"></a><span class="lineno"> 2416</span>&#160;            s_parent_dir=s_temporary_directory  </div><div class="line"><a name="l02417"></a><span class="lineno"> 2417</span>&#160;        <span class="comment">#end if no temp dir use orig genepop dir, else use temp</span></div><div class="line"><a name="l02418"></a><span class="lineno"> 2418</span>&#160;</div><div class="line"><a name="l02419"></a><span class="lineno"> 2419</span>&#160;        s_file_name=<span class="stringliteral">&quot;f&quot;</span> + str( i_genepop_file_count ) \</div><div class="line"><a name="l02420"></a><span class="lineno"> 2420</span>&#160;                    + <span class="stringliteral">&quot;i&quot;</span> + str( i_indiv_sample_count ) \</div><div class="line"><a name="l02421"></a><span class="lineno"> 2421</span>&#160;                    + <span class="stringliteral">&quot;l&quot;</span> + str( i_loci_subsample_count )</div><div class="line"><a name="l02422"></a><span class="lineno"> 2422</span>&#160;</div><div class="line"><a name="l02423"></a><span class="lineno"> 2423</span>&#160;        s_subsample_genepop_file=os.path.join( s_parent_dir, s_file_name )</div><div class="line"><a name="l02424"></a><span class="lineno"> 2424</span>&#160;        </div><div class="line"><a name="l02425"></a><span class="lineno"> 2425</span>&#160;        <span class="keywordflow">return</span> s_subsample_genepop_file</div><div class="line"><a name="l02426"></a><span class="lineno"> 2426</span>&#160;</div><div class="line"><a name="l02427"></a><span class="lineno"> 2427</span>&#160;<span class="comment">#end make_subsample_genepop_file_name</span></div><div class="line"><a name="l02428"></a><span class="lineno"> 2428</span>&#160;</div><div class="line"><a name="l02429"></a><span class="lineno"> 2429</span>&#160;<span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l02430"></a><span class="lineno"> 2430</span>&#160;<span class="stringliteral">2017_07_07.  I&#39;m adding the min and max loci numbers</span></div><div class="line"><a name="l02431"></a><span class="lineno"> 2431</span>&#160;<span class="stringliteral">to the arg list for def add_to_set_of_calls_to_do_estimate,</span></div><div class="line"><a name="l02432"></a><span class="lineno"> 2432</span>&#160;<span class="stringliteral">so that these will be passed to def do_estimate, which can </span></div><div class="line"><a name="l02433"></a><span class="lineno"> 2433</span>&#160;<span class="stringliteral">compute the heterozygosity across the range of loci the caller</span></div><div class="line"><a name="l02434"></a><span class="lineno"> 2434</span>&#160;<span class="stringliteral">specified.  This was motivated by genepop file input for which</span></div><div class="line"><a name="l02435"></a><span class="lineno"> 2435</span>&#160;<span class="stringliteral">loci 1-100 are microsats, and 101-500 are SNPs, and we had</span></div><div class="line"><a name="l02436"></a><span class="lineno"> 2436</span>&#160;<span class="stringliteral">het values targeted for each type, and nb values computed</span></div><div class="line"><a name="l02437"></a><span class="lineno"> 2437</span>&#160;<span class="stringliteral">for each type by usig loci ranges.  Thus, we wanted to see</span></div><div class="line"><a name="l02438"></a><span class="lineno"> 2438</span>&#160;<span class="stringliteral">the het values as calculated across only the microsats or</span></div><div class="line"><a name="l02439"></a><span class="lineno"> 2439</span>&#160;<span class="stringliteral">SNPs, whichever was specified using loci range.</span></div><div class="line"><a name="l02440"></a><span class="lineno"> 2440</span>&#160;<span class="stringliteral">&#39;&#39;&#39;</span></div></div><!-- fragment -->
</div>
</div>
<a class="anchor" id="a8373a4ba256fa5cd5d0bc88da91815b4"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def agestrucne.pgdriveneestimator.mymain </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>q_args</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">This code was at mod level under if __name__=="__main__",
but moved inside this def so this mod can be imported
and run inside python with a set of args identical to
those required on the command line.

This was an adaptation to enable PGGuiNeEstimator objects
to import this mod, then call this def after they load 
params (args for this driver) from their interface

2017_05_31. We have added a new final string argument
"True" or "False" to indicate whether to use GUI
messaging. Note that parse_args does not parse this
final arg, but this def evaluates it immediately
in order to propery use error messaging.
</pre><pre class="fragment">The multiprocessing event arg is not yet added to 
the list, so that we'll find the gui messaging flag
at its designated index minus one.
</pre> 
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l03800">3800</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>
<div class="fragment"><div class="line"><a name="l03800"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html#a8373a4ba256fa5cd5d0bc88da91815b4"> 3800</a></span>&#160;<span class="keyword">def </span><a class="code" href="namespaceagestrucne_1_1pgdriveneestimator.html#a8373a4ba256fa5cd5d0bc88da91815b4">mymain</a>( *q_args ):</div><div class="line"><a name="l03801"></a><span class="lineno"> 3801</span>&#160;</div><div class="line"><a name="l03802"></a><span class="lineno"> 3802</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l03803"></a><span class="lineno"> 3803</span>&#160;<span class="stringliteral">    This code was at mod level under if __name__==&quot;__main__&quot;,</span></div><div class="line"><a name="l03804"></a><span class="lineno"> 3804</span>&#160;<span class="stringliteral">    but moved inside this def so this mod can be imported</span></div><div class="line"><a name="l03805"></a><span class="lineno"> 3805</span>&#160;<span class="stringliteral">    and run inside python with a set of args identical to</span></div><div class="line"><a name="l03806"></a><span class="lineno"> 3806</span>&#160;<span class="stringliteral">    those required on the command line.</span></div><div class="line"><a name="l03807"></a><span class="lineno"> 3807</span>&#160;<span class="stringliteral"></span></div><div class="line"><a name="l03808"></a><span class="lineno"> 3808</span>&#160;<span class="stringliteral">    This was an adaptation to enable PGGuiNeEstimator objects</span></div><div class="line"><a name="l03809"></a><span class="lineno"> 3809</span>&#160;<span class="stringliteral">    to import this mod, then call this def after they load </span></div><div class="line"><a name="l03810"></a><span class="lineno"> 3810</span>&#160;<span class="stringliteral">    params (args for this driver) from their interface</span></div><div class="line"><a name="l03811"></a><span class="lineno"> 3811</span>&#160;<span class="stringliteral">    </span></div><div class="line"><a name="l03812"></a><span class="lineno"> 3812</span>&#160;<span class="stringliteral">    2017_05_31. We have added a new final string argument</span></div><div class="line"><a name="l03813"></a><span class="lineno"> 3813</span>&#160;<span class="stringliteral">    &quot;True&quot; or &quot;False&quot; to indicate whether to use GUI</span></div><div class="line"><a name="l03814"></a><span class="lineno"> 3814</span>&#160;<span class="stringliteral">    messaging. Note that parse_args does not parse this</span></div><div class="line"><a name="l03815"></a><span class="lineno"> 3815</span>&#160;<span class="stringliteral">    final arg, but this def evaluates it immediately</span></div><div class="line"><a name="l03816"></a><span class="lineno"> 3816</span>&#160;<span class="stringliteral">    in order to propery use error messaging.</span></div><div class="line"><a name="l03817"></a><span class="lineno"> 3817</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l03818"></a><span class="lineno"> 3818</span>&#160;</div><div class="line"><a name="l03819"></a><span class="lineno"> 3819</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l03820"></a><span class="lineno"> 3820</span>&#160;<span class="stringliteral">    The multiprocessing event arg is not yet added to </span></div><div class="line"><a name="l03821"></a><span class="lineno"> 3821</span>&#160;<span class="stringliteral">    the list, so that we&#39;ll find the gui messaging flag</span></div><div class="line"><a name="l03822"></a><span class="lineno"> 3822</span>&#160;<span class="stringliteral">    at its designated index minus one.</span></div><div class="line"><a name="l03823"></a><span class="lineno"> 3823</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l03824"></a><span class="lineno"> 3824</span>&#160;    s_gui_flag=q_args[ IDX_USE_GUI_MESSAGING - 1 ]</div><div class="line"><a name="l03825"></a><span class="lineno"> 3825</span>&#160;</div><div class="line"><a name="l03826"></a><span class="lineno"> 3826</span>&#160;    <span class="keywordflow">if</span> s_gui_flag <span class="keywordflow">not</span> <span class="keywordflow">in</span> [ <span class="stringliteral">&quot;True&quot;</span>, <span class="stringliteral">&quot;False&quot;</span> ]:</div><div class="line"><a name="l03827"></a><span class="lineno"> 3827</span>&#160;</div><div class="line"><a name="l03828"></a><span class="lineno"> 3828</span>&#160;        s_msg=<span class="stringliteral">&quot;In pgdriveneestimator.py, def mymain, &quot;</span> \</div><div class="line"><a name="l03829"></a><span class="lineno"> 3829</span>&#160;                        + <span class="stringliteral">&quot;the program did not find the &quot;</span> \</div><div class="line"><a name="l03830"></a><span class="lineno"> 3830</span>&#160;                        + <span class="stringliteral">&quot;\&quot;True\&quot; or \&quot;False\&quot; value &quot;</span> \</div><div class="line"><a name="l03831"></a><span class="lineno"> 3831</span>&#160;                        + <span class="stringliteral">&quot;for the parameter that indicates &quot;</span> \</div><div class="line"><a name="l03832"></a><span class="lineno"> 3832</span>&#160;                        + <span class="stringliteral">&quot;whether to activate GUI messaging.  &quot;</span> \</div><div class="line"><a name="l03833"></a><span class="lineno"> 3833</span>&#160;                        + <span class="stringliteral">&quot;The value found is: &quot;</span> + s_gui_flag + <span class="stringliteral">&quot;.&quot;</span></div><div class="line"><a name="l03834"></a><span class="lineno"> 3834</span>&#160;        <span class="keywordflow">raise</span> Exception( s_msg )</div><div class="line"><a name="l03835"></a><span class="lineno"> 3835</span>&#160;    <span class="comment">#end if no flag value</span></div><div class="line"><a name="l03836"></a><span class="lineno"> 3836</span>&#160;</div><div class="line"><a name="l03837"></a><span class="lineno"> 3837</span>&#160;    <a class="code" href="namespaceagestrucne_1_1pgdriveneestimator.html#ac504fdce8e8069dc8342ab9bcb40c49e">set_messaging_procedures</a>( s_gui_flag )</div><div class="line"><a name="l03838"></a><span class="lineno"> 3838</span>&#160;</div><div class="line"><a name="l03839"></a><span class="lineno"> 3839</span>&#160;    <span class="keywordflow">try</span>:</div><div class="line"><a name="l03840"></a><span class="lineno"> 3840</span>&#160;</div><div class="line"><a name="l03841"></a><span class="lineno"> 3841</span>&#160;        <span class="keywordflow">if</span> VERY_VERBOSE:</div><div class="line"><a name="l03842"></a><span class="lineno"> 3842</span>&#160;            print( <span class="stringliteral">&quot;In pgdriveneestimator, def mymain&quot;</span> )</div><div class="line"><a name="l03843"></a><span class="lineno"> 3843</span>&#160;        <span class="comment">#end if VERY_VERBOSE</span></div><div class="line"><a name="l03844"></a><span class="lineno"> 3844</span>&#160;</div><div class="line"><a name="l03845"></a><span class="lineno"> 3845</span>&#160;        <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l03846"></a><span class="lineno"> 3846</span>&#160;<span class="stringliteral">        To allow the pygenomics genomics.ne2.controller modules</span></div><div class="line"><a name="l03847"></a><span class="lineno"> 3847</span>&#160;<span class="stringliteral">        Ne2 estimator class object to loop over the paths in the</span></div><div class="line"><a name="l03848"></a><span class="lineno"> 3848</span>&#160;<span class="stringliteral">        PATH environmental variable, and be ensured that each</span></div><div class="line"><a name="l03849"></a><span class="lineno"> 3849</span>&#160;<span class="stringliteral">        exists.</span></div><div class="line"><a name="l03850"></a><span class="lineno"> 3850</span>&#160;<span class="stringliteral">        &#39;&#39;&#39;</span></div><div class="line"><a name="l03851"></a><span class="lineno"> 3851</span>&#160;        pgut.remove_non_existent_paths_from_path_variable()</div><div class="line"><a name="l03852"></a><span class="lineno"> 3852</span>&#160;</div><div class="line"><a name="l03853"></a><span class="lineno"> 3853</span>&#160;        <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l03854"></a><span class="lineno"> 3854</span>&#160;<span class="stringliteral">        New code to confirm the selected executable,</span></div><div class="line"><a name="l03855"></a><span class="lineno"> 3855</span>&#160;<span class="stringliteral">        Now that we are using LDNe2 by default,</span></div><div class="line"><a name="l03856"></a><span class="lineno"> 3856</span>&#160;<span class="stringliteral">        with Ne2 as the alternative, as indicated by the</span></div><div class="line"><a name="l03857"></a><span class="lineno"> 3857</span>&#160;<span class="stringliteral">        flag constant.</span></div><div class="line"><a name="l03858"></a><span class="lineno"> 3858</span>&#160;<span class="stringliteral">        &#39;&#39;&#39;</span></div><div class="line"><a name="l03859"></a><span class="lineno"> 3859</span>&#160;        b_found_executable=<span class="keyword">False</span></div><div class="line"><a name="l03860"></a><span class="lineno"> 3860</span>&#160;</div><div class="line"><a name="l03861"></a><span class="lineno"> 3861</span>&#160;        <span class="keywordflow">if</span> ESTIMATOR_TO_USE==LDNE2:</div><div class="line"><a name="l03862"></a><span class="lineno"> 3862</span>&#160;            b_found_executable = set_mod_level_var_to_ldne2_executable_path()</div><div class="line"><a name="l03863"></a><span class="lineno"> 3863</span>&#160;        <span class="keywordflow">elif</span> ESTIMATOR_TO_USE==NEESTIMATOR:</div><div class="line"><a name="l03864"></a><span class="lineno"> 3864</span>&#160;            b_found_executable = did_find_ne_estimator_executable()</div><div class="line"><a name="l03865"></a><span class="lineno"> 3865</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l03866"></a><span class="lineno"> 3866</span>&#160;            s_msg = <span class="stringliteral">&quot;In pgdriveneestimator.py, def mymain(), &quot;</span> \</div><div class="line"><a name="l03867"></a><span class="lineno"> 3867</span>&#160;                            + <span class="stringliteral">&quot;checking for selected estimator executable. &quot;</span> \</div><div class="line"><a name="l03868"></a><span class="lineno"> 3868</span>&#160;                            + <span class="stringliteral">&quot;The selected estimator program is unknown: &quot;</span> \</div><div class="line"><a name="l03869"></a><span class="lineno"> 3869</span>&#160;                            + str( ESTIMATOR_TO_USE ) + <span class="stringliteral">&quot;.&quot;</span></div><div class="line"><a name="l03870"></a><span class="lineno"> 3870</span>&#160;            <span class="keywordflow">raise</span> Exception( s_msg )</div><div class="line"><a name="l03871"></a><span class="lineno"> 3871</span>&#160;        <span class="comment">#dnd if using ldne, else neestimator else unknown</span></div><div class="line"><a name="l03872"></a><span class="lineno"> 3872</span>&#160;        </div><div class="line"><a name="l03873"></a><span class="lineno"> 3873</span>&#160;        <span class="keywordflow">if</span> <span class="keywordflow">not</span> b_found_executable:</div><div class="line"><a name="l03874"></a><span class="lineno"> 3874</span>&#160;            s_msg = <span class="stringliteral">&quot;In pgdriveneestimator.py, def mymain(), &quot;</span> \</div><div class="line"><a name="l03875"></a><span class="lineno"> 3875</span>&#160;                                + <span class="stringliteral">&quot;did not find NeEstimator executable.&quot;</span> </div><div class="line"><a name="l03876"></a><span class="lineno"> 3876</span>&#160;            <span class="keywordflow">raise</span> Exception(s_msg )</div><div class="line"><a name="l03877"></a><span class="lineno"> 3877</span>&#160;        <span class="comment">#end if can&#39;t find executable</span></div><div class="line"><a name="l03878"></a><span class="lineno"> 3878</span>&#160;        </div><div class="line"><a name="l03879"></a><span class="lineno"> 3879</span>&#160;        <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l03880"></a><span class="lineno"> 3880</span>&#160;<span class="stringliteral">        We adapted this def to open and close the main and secondary </span></div><div class="line"><a name="l03881"></a><span class="lineno"> 3881</span>&#160;<span class="stringliteral">        output files. Originally, it simply passed open file objects to def,</span></div><div class="line"><a name="l03882"></a><span class="lineno"> 3882</span>&#160;<span class="stringliteral">        parse args.  Now we we check for string vs file for these</span></div><div class="line"><a name="l03883"></a><span class="lineno"> 3883</span>&#160;<span class="stringliteral">        args, as the console version will still pass stdout and sterr</span></div><div class="line"><a name="l03884"></a><span class="lineno"> 3884</span>&#160;<span class="stringliteral">        file objects, but the GUI will call with string filenames.  </span></div><div class="line"><a name="l03885"></a><span class="lineno"> 3885</span>&#160;<span class="stringliteral">        We strip off the last 3 items, which now need to be checked </span></div><div class="line"><a name="l03886"></a><span class="lineno"> 3886</span>&#160;<span class="stringliteral">        for type string file names and assume, if not string, then </span></div><div class="line"><a name="l03887"></a><span class="lineno"> 3887</span>&#160;<span class="stringliteral">        the objects passed are assumed to be (open-for-writing) </span></div><div class="line"><a name="l03888"></a><span class="lineno"> 3888</span>&#160;<span class="stringliteral">        file objects.  We also pass along the last arg, the multiprocessing</span></div><div class="line"><a name="l03889"></a><span class="lineno"> 3889</span>&#160;<span class="stringliteral">        event.  </span></div><div class="line"><a name="l03890"></a><span class="lineno"> 3890</span>&#160;<span class="stringliteral">        &#39;&#39;&#39;</span></div><div class="line"><a name="l03891"></a><span class="lineno"> 3891</span>&#160;        seq_unaltered_args=q_args[0:IDX_LAST_CONSOLE_ARG + 1 ]</div><div class="line"><a name="l03892"></a><span class="lineno"> 3892</span>&#160;</div><div class="line"><a name="l03893"></a><span class="lineno"> 3893</span>&#160;        v_main_outfile_arg= q_args[ IDX_MAIN_OUTFILE ]</div><div class="line"><a name="l03894"></a><span class="lineno"> 3894</span>&#160;        v_secondary_outfile_arg=q_args[ IDX_SECONDARY_OUTFILE ]</div><div class="line"><a name="l03895"></a><span class="lineno"> 3895</span>&#160;</div><div class="line"><a name="l03896"></a><span class="lineno"> 3896</span>&#160;        <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l03897"></a><span class="lineno"> 3897</span>&#160;<span class="stringliteral">        New argument, added 2017_03_27, to allow def do_estimate</span></div><div class="line"><a name="l03898"></a><span class="lineno"> 3898</span>&#160;<span class="stringliteral">        to write its intermediate files inside a single temp</span></div><div class="line"><a name="l03899"></a><span class="lineno"> 3899</span>&#160;<span class="stringliteral">        dir, whose name is known to the caller, so caller can</span></div><div class="line"><a name="l03900"></a><span class="lineno"> 3900</span>&#160;<span class="stringliteral">        remove it.  </span></div><div class="line"><a name="l03901"></a><span class="lineno"> 3901</span>&#160;<span class="stringliteral">        </span></div><div class="line"><a name="l03902"></a><span class="lineno"> 3902</span>&#160;<span class="stringliteral">        Note that we subtract one to get its postiion in the</span></div><div class="line"><a name="l03903"></a><span class="lineno"> 3903</span>&#160;<span class="stringliteral">        passed args.  This argument is passed as the last,</span></div><div class="line"><a name="l03904"></a><span class="lineno"> 3904</span>&#160;<span class="stringliteral">        but the caller will not have passed the multiprocessing</span></div><div class="line"><a name="l03905"></a><span class="lineno"> 3905</span>&#160;<span class="stringliteral">        event argument, so the index for our temp directory is</span></div><div class="line"><a name="l03906"></a><span class="lineno"> 3906</span>&#160;<span class="stringliteral">        one less than the final position as passed to </span></div><div class="line"><a name="l03907"></a><span class="lineno"> 3907</span>&#160;<span class="stringliteral">        drive_estimator (and subsequenctly to parse_args):</span></div><div class="line"><a name="l03908"></a><span class="lineno"> 3908</span>&#160;<span class="stringliteral">        &#39;&#39;&#39;</span></div><div class="line"><a name="l03909"></a><span class="lineno"> 3909</span>&#160;        v_temporary_directory_arg=q_args[ IDX_TEMPORARY_DIRECTORY - 1 ]</div><div class="line"><a name="l03910"></a><span class="lineno"> 3910</span>&#160;</div><div class="line"><a name="l03911"></a><span class="lineno"> 3911</span>&#160;        <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l03912"></a><span class="lineno"> 3912</span>&#160;<span class="stringliteral">        Note that we check for string type, and otherwise</span></div><div class="line"><a name="l03913"></a><span class="lineno"> 3913</span>&#160;<span class="stringliteral">        assume open file objects. (In Linux sys.std{out,err} return</span></div><div class="line"><a name="l03914"></a><span class="lineno"> 3914</span>&#160;<span class="stringliteral">        &quot;file&quot; as type, but in Windows they return iostream or</span></div><div class="line"><a name="l03915"></a><span class="lineno"> 3915</span>&#160;<span class="stringliteral">        something similar, but not file object.</span></div><div class="line"><a name="l03916"></a><span class="lineno"> 3916</span>&#160;<span class="stringliteral">        &#39;&#39;&#39;</span></div><div class="line"><a name="l03917"></a><span class="lineno"> 3917</span>&#160;        <span class="keywordflow">if</span> type( v_main_outfile_arg  ) == str:</div><div class="line"><a name="l03918"></a><span class="lineno"> 3918</span>&#160;            o_main_outfile=open( v_main_outfile_arg, <span class="stringliteral">&#39;w&#39;</span> )</div><div class="line"><a name="l03919"></a><span class="lineno"> 3919</span>&#160;            o_secondary_outfile=open( v_secondary_outfile_arg, <span class="stringliteral">&#39;w&#39;</span> )</div><div class="line"><a name="l03920"></a><span class="lineno"> 3920</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l03921"></a><span class="lineno"> 3921</span>&#160;            o_main_outfile=v_main_outfile_arg</div><div class="line"><a name="l03922"></a><span class="lineno"> 3922</span>&#160;            o_secondary_outfile=v_secondary_outfile_arg</div><div class="line"><a name="l03923"></a><span class="lineno"> 3923</span>&#160;        <span class="comment">#end if passed non-file object, open file, else pass along file object</span></div><div class="line"><a name="l03924"></a><span class="lineno"> 3924</span>&#160;</div><div class="line"><a name="l03925"></a><span class="lineno"> 3925</span>&#160;        <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l03926"></a><span class="lineno"> 3926</span>&#160;<span class="stringliteral">        Reconstruct our arg list after creating the outfile objects.</span></div><div class="line"><a name="l03927"></a><span class="lineno"> 3927</span>&#160;<span class="stringliteral"></span></div><div class="line"><a name="l03928"></a><span class="lineno"> 3928</span>&#160;<span class="stringliteral">        2017_02_12.  Having replaced the use of the python multiprocessing.Process</span></div><div class="line"><a name="l03929"></a><span class="lineno"> 3929</span>&#160;<span class="stringliteral">        object with a Popen call when calling this def from the pgutilities.py, we</span></div><div class="line"><a name="l03930"></a><span class="lineno"> 3930</span>&#160;<span class="stringliteral">        no longer have the multiprocessing event available.  For now, rather than</span></div><div class="line"><a name="l03931"></a><span class="lineno"> 3931</span>&#160;<span class="stringliteral">        removing this arg, the former multiprocessing event to allow communication</span></div><div class="line"><a name="l03932"></a><span class="lineno"> 3932</span>&#160;<span class="stringliteral">        between pgutilities.py and this code, we simply default this arg to None:</span></div><div class="line"><a name="l03933"></a><span class="lineno"> 3933</span>&#160;<span class="stringliteral">        &#39;&#39;&#39;</span></div><div class="line"><a name="l03934"></a><span class="lineno"> 3934</span>&#160;        o_event=<span class="keywordtype">None</span></div><div class="line"><a name="l03935"></a><span class="lineno"> 3935</span>&#160;</div><div class="line"><a name="l03936"></a><span class="lineno"> 3936</span>&#160;        seq_args_to_parse=seq_unaltered_args + ( o_main_outfile, </div><div class="line"><a name="l03937"></a><span class="lineno"> 3937</span>&#160;                                                    o_secondary_outfile, </div><div class="line"><a name="l03938"></a><span class="lineno"> 3938</span>&#160;                                                                    o_event, </div><div class="line"><a name="l03939"></a><span class="lineno"> 3939</span>&#160;                                                    v_temporary_directory_arg )</div><div class="line"><a name="l03940"></a><span class="lineno"> 3940</span>&#160;</div><div class="line"><a name="l03941"></a><span class="lineno"> 3941</span>&#160;        drive_estimator( *seq_args_to_parse )</div><div class="line"><a name="l03942"></a><span class="lineno"> 3942</span>&#160;</div><div class="line"><a name="l03943"></a><span class="lineno"> 3943</span>&#160;        <span class="keywordflow">if</span> o_main_outfile != sys.stdout:</div><div class="line"><a name="l03944"></a><span class="lineno"> 3944</span>&#160;            o_main_outfile.close()</div><div class="line"><a name="l03945"></a><span class="lineno"> 3945</span>&#160;        <span class="comment">#end if not stdout, close</span></div><div class="line"><a name="l03946"></a><span class="lineno"> 3946</span>&#160;</div><div class="line"><a name="l03947"></a><span class="lineno"> 3947</span>&#160;        <span class="keywordflow">if</span> o_secondary_outfile != sys.stderr:</div><div class="line"><a name="l03948"></a><span class="lineno"> 3948</span>&#160;            o_secondary_outfile.close()</div><div class="line"><a name="l03949"></a><span class="lineno"> 3949</span>&#160;        <span class="comment">#end if not stderr, close file</span></div><div class="line"><a name="l03950"></a><span class="lineno"> 3950</span>&#160;</div><div class="line"><a name="l03951"></a><span class="lineno"> 3951</span>&#160;    <span class="keywordflow">except</span> Exception <span class="keyword">as</span> oex:</div><div class="line"><a name="l03952"></a><span class="lineno"> 3952</span>&#160;        o_traceback=sys.exc_info()[ 2 ]</div><div class="line"><a name="l03953"></a><span class="lineno"> 3953</span>&#160;        s_trace=pgut.get_traceback_info_about_offending_code( \</div><div class="line"><a name="l03954"></a><span class="lineno"> 3954</span>&#160;                                                    o_traceback )</div><div class="line"><a name="l03955"></a><span class="lineno"> 3955</span>&#160;        <span class="keywordflow">if</span> ACTIVATE_GUI_MESSAGING:</div><div class="line"><a name="l03956"></a><span class="lineno"> 3956</span>&#160;            s_errormsg=str( oex )</div><div class="line"><a name="l03957"></a><span class="lineno"> 3957</span>&#160;            GUIERR( <span class="keywordtype">None</span> , s_errormsg + <span class="stringliteral">&quot;\n&quot;</span> + s_trace )</div><div class="line"><a name="l03958"></a><span class="lineno"> 3958</span>&#160;        <span class="comment">#end if using gui messaging</span></div><div class="line"><a name="l03959"></a><span class="lineno"> 3959</span>&#160;        </div><div class="line"><a name="l03960"></a><span class="lineno"> 3960</span>&#160;        <span class="keywordflow">raise</span> Exception( <span class="stringliteral">&quot;In pgdriveneestimator.py, def mymain, &quot;</span>  \</div><div class="line"><a name="l03961"></a><span class="lineno"> 3961</span>&#160;                            + <span class="stringliteral">&quot;an exception was caught with message: &quot;</span> \</div><div class="line"><a name="l03962"></a><span class="lineno"> 3962</span>&#160;                            + str(oex) + <span class="stringliteral">&quot;, and trace back: &quot;</span> \</div><div class="line"><a name="l03963"></a><span class="lineno"> 3963</span>&#160;                            + s_trace )</div><div class="line"><a name="l03964"></a><span class="lineno"> 3964</span>&#160;    <span class="comment">#end try ... except</span></div><div class="line"><a name="l03965"></a><span class="lineno"> 3965</span>&#160;</div><div class="line"><a name="l03966"></a><span class="lineno"> 3966</span>&#160;    <span class="keywordflow">return</span></div><div class="line"><a name="l03967"></a><span class="lineno"> 3967</span>&#160;<span class="comment">#end mymain</span></div><div class="line"><a name="l03968"></a><span class="lineno"> 3968</span>&#160;</div><div class="ttc" id="namespaceagestrucne_1_1pgdriveneestimator_html_a8373a4ba256fa5cd5d0bc88da91815b4"><div class="ttname"><a href="namespaceagestrucne_1_1pgdriveneestimator.html#a8373a4ba256fa5cd5d0bc88da91815b4">agestrucne.pgdriveneestimator.mymain</a></div><div class="ttdeci">def mymain(q_args)</div><div class="ttdef"><b>Definition:</b> <a href="pgdriveneestimator_8py_source.html#l03800">pgdriveneestimator.py:3800</a></div></div>
<div class="ttc" id="namespaceagestrucne_1_1pgdriveneestimator_html_ac504fdce8e8069dc8342ab9bcb40c49e"><div class="ttname"><a href="namespaceagestrucne_1_1pgdriveneestimator.html#ac504fdce8e8069dc8342ab9bcb40c49e">agestrucne.pgdriveneestimator.set_messaging_procedures</a></div><div class="ttdeci">def set_messaging_procedures(s_gui_flag)</div><div class="ttdef"><b>Definition:</b> <a href="pgdriveneestimator_8py_source.html#l03767">pgdriveneestimator.py:3767</a></div></div>
</div><!-- fragment -->
</div>
</div>
<a class="anchor" id="af82482f5cfed335c23bde8de169b44be"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def agestrucne.pgdriveneestimator.mymainlongfilelist </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>q_args</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">2017_02_21.  This def is added as an alternative call
for pgutilitites run_driveneestimator_in_new_process
and its call to call_driveneestimator_using_subprocess,
to solve a Windows problem.  When calling with a long
list of (many hundreds) of genepop files, we saw a limit
of Windows ability to correctly make the call via pOpen, 
due to a limit on the number of characters it can handle
in a command.  Now the above pgutilities defs will have
tested the command length.  If it exceeds a threshold,
the comma-delimited list of genepop files will be
written to a temporary file.  This def will read in
the list from that file, delete the file, and then
call mymain with the now normalized arg 1.

We assume the list is formatted as it is when mymain
is called directly, and, further, that it is in a single,
first line in the file.
</pre> 
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l03728">3728</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>
<div class="fragment"><div class="line"><a name="l03728"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html#af82482f5cfed335c23bde8de169b44be"> 3728</a></span>&#160;<span class="keyword">def </span><a class="code" href="namespaceagestrucne_1_1pgdriveneestimator.html#af82482f5cfed335c23bde8de169b44be">mymainlongfilelist</a>( *q_args ):</div><div class="line"><a name="l03729"></a><span class="lineno"> 3729</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l03730"></a><span class="lineno"> 3730</span>&#160;<span class="stringliteral">    2017_02_21.  This def is added as an alternative call</span></div><div class="line"><a name="l03731"></a><span class="lineno"> 3731</span>&#160;<span class="stringliteral">    for pgutilitites run_driveneestimator_in_new_process</span></div><div class="line"><a name="l03732"></a><span class="lineno"> 3732</span>&#160;<span class="stringliteral">    and its call to call_driveneestimator_using_subprocess,</span></div><div class="line"><a name="l03733"></a><span class="lineno"> 3733</span>&#160;<span class="stringliteral">    to solve a Windows problem.  When calling with a long</span></div><div class="line"><a name="l03734"></a><span class="lineno"> 3734</span>&#160;<span class="stringliteral">    list of (many hundreds) of genepop files, we saw a limit</span></div><div class="line"><a name="l03735"></a><span class="lineno"> 3735</span>&#160;<span class="stringliteral">    of Windows ability to correctly make the call via pOpen, </span></div><div class="line"><a name="l03736"></a><span class="lineno"> 3736</span>&#160;<span class="stringliteral">    due to a limit on the number of characters it can handle</span></div><div class="line"><a name="l03737"></a><span class="lineno"> 3737</span>&#160;<span class="stringliteral">    in a command.  Now the above pgutilities defs will have</span></div><div class="line"><a name="l03738"></a><span class="lineno"> 3738</span>&#160;<span class="stringliteral">    tested the command length.  If it exceeds a threshold,</span></div><div class="line"><a name="l03739"></a><span class="lineno"> 3739</span>&#160;<span class="stringliteral">    the comma-delimited list of genepop files will be</span></div><div class="line"><a name="l03740"></a><span class="lineno"> 3740</span>&#160;<span class="stringliteral">    written to a temporary file.  This def will read in</span></div><div class="line"><a name="l03741"></a><span class="lineno"> 3741</span>&#160;<span class="stringliteral">    the list from that file, delete the file, and then</span></div><div class="line"><a name="l03742"></a><span class="lineno"> 3742</span>&#160;<span class="stringliteral">    call mymain with the now normalized arg 1.</span></div><div class="line"><a name="l03743"></a><span class="lineno"> 3743</span>&#160;<span class="stringliteral"></span></div><div class="line"><a name="l03744"></a><span class="lineno"> 3744</span>&#160;<span class="stringliteral">    We assume the list is formatted as it is when mymain</span></div><div class="line"><a name="l03745"></a><span class="lineno"> 3745</span>&#160;<span class="stringliteral">    is called directly, and, further, that it is in a single,</span></div><div class="line"><a name="l03746"></a><span class="lineno"> 3746</span>&#160;<span class="stringliteral">    first line in the file.</span></div><div class="line"><a name="l03747"></a><span class="lineno"> 3747</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l03748"></a><span class="lineno"> 3748</span>&#160;</div><div class="line"><a name="l03749"></a><span class="lineno"> 3749</span>&#160;    s_temp_file_with_genepop_file_list=q_args[ IDX_GENEPOP_FILES ]</div><div class="line"><a name="l03750"></a><span class="lineno"> 3750</span>&#160;</div><div class="line"><a name="l03751"></a><span class="lineno"> 3751</span>&#160;    o_temp_file_with_genepop_file=open( s_temp_file_with_genepop_file_list )</div><div class="line"><a name="l03752"></a><span class="lineno"> 3752</span>&#160;</div><div class="line"><a name="l03753"></a><span class="lineno"> 3753</span>&#160;    <span class="comment">#Thie file&#39;s single line should give the correct first argument of</span></div><div class="line"><a name="l03754"></a><span class="lineno"> 3754</span>&#160;    <span class="comment">#a list of genepop files, so that mymain can pass it to parse_args </span></div><div class="line"><a name="l03755"></a><span class="lineno"> 3755</span>&#160;    <span class="comment">#so that it is parsed into a list of file names.</span></div><div class="line"><a name="l03756"></a><span class="lineno"> 3756</span>&#160;    s_genepop_file_list=o_temp_file_with_genepop_file.readline().strip()</div><div class="line"><a name="l03757"></a><span class="lineno"> 3757</span>&#160;    q_corrected_args=( s_genepop_file_list, ) + q_args[ 1 : ]</div><div class="line"><a name="l03758"></a><span class="lineno"> 3758</span>&#160;</div><div class="line"><a name="l03759"></a><span class="lineno"> 3759</span>&#160;    o_temp_file_with_genepop_file.close()</div><div class="line"><a name="l03760"></a><span class="lineno"> 3760</span>&#160;</div><div class="line"><a name="l03761"></a><span class="lineno"> 3761</span>&#160;    <a class="code" href="namespaceagestrucne_1_1pgdriveneestimator.html#a8373a4ba256fa5cd5d0bc88da91815b4">mymain</a>( *q_corrected_args )</div><div class="line"><a name="l03762"></a><span class="lineno"> 3762</span>&#160;</div><div class="line"><a name="l03763"></a><span class="lineno"> 3763</span>&#160;    <span class="keywordflow">return</span></div><div class="line"><a name="l03764"></a><span class="lineno"> 3764</span>&#160;</div><div class="line"><a name="l03765"></a><span class="lineno"> 3765</span>&#160;<span class="comment">#end mymainlongfilelist</span></div><div class="line"><a name="l03766"></a><span class="lineno"> 3766</span>&#160;</div><div class="ttc" id="namespaceagestrucne_1_1pgdriveneestimator_html_a8373a4ba256fa5cd5d0bc88da91815b4"><div class="ttname"><a href="namespaceagestrucne_1_1pgdriveneestimator.html#a8373a4ba256fa5cd5d0bc88da91815b4">agestrucne.pgdriveneestimator.mymain</a></div><div class="ttdeci">def mymain(q_args)</div><div class="ttdef"><b>Definition:</b> <a href="pgdriveneestimator_8py_source.html#l03800">pgdriveneestimator.py:3800</a></div></div>
<div class="ttc" id="namespaceagestrucne_1_1pgdriveneestimator_html_af82482f5cfed335c23bde8de169b44be"><div class="ttname"><a href="namespaceagestrucne_1_1pgdriveneestimator.html#af82482f5cfed335c23bde8de169b44be">agestrucne.pgdriveneestimator.mymainlongfilelist</a></div><div class="ttdeci">def mymainlongfilelist(q_args)</div><div class="ttdef"><b>Definition:</b> <a href="pgdriveneestimator_8py_source.html#l03728">pgdriveneestimator.py:3728</a></div></div>
</div><!-- fragment -->
</div>
</div>
<a class="anchor" id="aa1487bffaa5efe1e3ce66928866a6e8f"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def agestrucne.pgdriveneestimator.parse_sampling_list_that_uses_indiv_id_fields </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>ls_sampling_list_items</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">We require that we have at least 3 items
to set up  sampling using individual id field
criteria:
--list of field names (space delimited)
    ex: "age parent1 parent2"
--list of field types (python types, space delimited)
    ex: "int str str"
--at least one parameter (not validated here
    -&gt; example in criteria expression  "age &gt; 3"
    -&gt;example, in cohort we expect several params like max. age     
2016_11_10
    -&gt;example, in cohorts by percentages, we'll have the max age
      followed by a hyphented list of percentage numerics.  This
      will not be parsed here, but will be passsed as part of the
      list returned as  ls_params</pre> 
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l01715">1715</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>
<div class="fragment"><div class="line"><a name="l01715"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html#aa1487bffaa5efe1e3ce66928866a6e8f"> 1715</a></span>&#160;<span class="keyword">def </span><a class="code" href="namespaceagestrucne_1_1pgdriveneestimator.html#aa1487bffaa5efe1e3ce66928866a6e8f">parse_sampling_list_that_uses_indiv_id_fields</a>( ls_sampling_list_items ):</div><div class="line"><a name="l01716"></a><span class="lineno"> 1716</span>&#160;</div><div class="line"><a name="l01717"></a><span class="lineno"> 1717</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l01718"></a><span class="lineno"> 1718</span>&#160;<span class="stringliteral">    We require that we have at least 3 items</span></div><div class="line"><a name="l01719"></a><span class="lineno"> 1719</span>&#160;<span class="stringliteral">    to set up  sampling using individual id field</span></div><div class="line"><a name="l01720"></a><span class="lineno"> 1720</span>&#160;<span class="stringliteral">    criteria:</span></div><div class="line"><a name="l01721"></a><span class="lineno"> 1721</span>&#160;<span class="stringliteral">    --list of field names (space delimited)</span></div><div class="line"><a name="l01722"></a><span class="lineno"> 1722</span>&#160;<span class="stringliteral">        ex: &quot;age parent1 parent2&quot;</span></div><div class="line"><a name="l01723"></a><span class="lineno"> 1723</span>&#160;<span class="stringliteral">    --list of field types (python types, space delimited)</span></div><div class="line"><a name="l01724"></a><span class="lineno"> 1724</span>&#160;<span class="stringliteral">        ex: &quot;int str str&quot;</span></div><div class="line"><a name="l01725"></a><span class="lineno"> 1725</span>&#160;<span class="stringliteral">    --at least one parameter (not validated here</span></div><div class="line"><a name="l01726"></a><span class="lineno"> 1726</span>&#160;<span class="stringliteral">        -&gt; example in criteria expression  &quot;age &gt; 3&quot;</span></div><div class="line"><a name="l01727"></a><span class="lineno"> 1727</span>&#160;<span class="stringliteral">        -&gt;example, in cohort we expect several params like max. age     </span></div><div class="line"><a name="l01728"></a><span class="lineno"> 1728</span>&#160;<span class="stringliteral">    2016_11_10</span></div><div class="line"><a name="l01729"></a><span class="lineno"> 1729</span>&#160;<span class="stringliteral">        -&gt;example, in cohorts by percentages, we&#39;ll have the max age</span></div><div class="line"><a name="l01730"></a><span class="lineno"> 1730</span>&#160;<span class="stringliteral">          followed by a hyphented list of percentage numerics.  This</span></div><div class="line"><a name="l01731"></a><span class="lineno"> 1731</span>&#160;<span class="stringliteral">          will not be parsed here, but will be passsed as part of the</span></div><div class="line"><a name="l01732"></a><span class="lineno"> 1732</span>&#160;<span class="stringliteral">          list returned as  ls_params</span></div><div class="line"><a name="l01733"></a><span class="lineno"> 1733</span>&#160;<span class="stringliteral"></span></div><div class="line"><a name="l01734"></a><span class="lineno"> 1734</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l01735"></a><span class="lineno"> 1735</span>&#160;</div><div class="line"><a name="l01736"></a><span class="lineno"> 1736</span>&#160;    MIN_NUM_SAMPLING_ITEMS=3</div><div class="line"><a name="l01737"></a><span class="lineno"> 1737</span>&#160;    IDX_FIELD_NAMES=0</div><div class="line"><a name="l01738"></a><span class="lineno"> 1738</span>&#160;    IDX_FIELD_TYPES=1</div><div class="line"><a name="l01739"></a><span class="lineno"> 1739</span>&#160;    IDX_FIRST_PARAM=2</div><div class="line"><a name="l01740"></a><span class="lineno"> 1740</span>&#160;    FIELD_DELIMITER=<span class="stringliteral">&quot;;&quot;</span></div><div class="line"><a name="l01741"></a><span class="lineno"> 1741</span>&#160;</div><div class="line"><a name="l01742"></a><span class="lineno"> 1742</span>&#160;    i_tot_sampling_items=len( ls_sampling_list_items )</div><div class="line"><a name="l01743"></a><span class="lineno"> 1743</span>&#160;</div><div class="line"><a name="l01744"></a><span class="lineno"> 1744</span>&#160;    <span class="keywordflow">if</span> i_tot_sampling_items &lt; MIN_NUM_SAMPLING_ITEMS:</div><div class="line"><a name="l01745"></a><span class="lineno"> 1745</span>&#160;        s_msg = <span class="stringliteral">&quot;In pgdriveneestimator.py, def parse_sampling_list_that_uses_indiv_id_fields, &quot;</span> \</div><div class="line"><a name="l01746"></a><span class="lineno"> 1746</span>&#160;                    + <span class="stringliteral">&quot;for the sampling criteria list, expecting at 3 items &quot;</span> \</div><div class="line"><a name="l01747"></a><span class="lineno"> 1747</span>&#160;                    + <span class="stringliteral">&quot;(field names, field types, and one or more &quot;</span> \</div><div class="line"><a name="l01748"></a><span class="lineno"> 1748</span>&#160;                    + <span class="stringliteral">&quot;parameters.  Found list: &quot;</span> \</div><div class="line"><a name="l01749"></a><span class="lineno"> 1749</span>&#160;                    + str( ls_sampling_list_items ) + <span class="stringliteral">&quot;.&quot;</span></div><div class="line"><a name="l01750"></a><span class="lineno"> 1750</span>&#160;        <span class="keywordflow">raise</span> Exception( s_msg )</div><div class="line"><a name="l01751"></a><span class="lineno"> 1751</span>&#160;</div><div class="line"><a name="l01752"></a><span class="lineno"> 1752</span>&#160;    <span class="comment">#end if too few criteria items</span></div><div class="line"><a name="l01753"></a><span class="lineno"> 1753</span>&#160;    ls_field_names=ls_sampling_list_items[ IDX_FIELD_NAMES ].split( FIELD_DELIMITER )</div><div class="line"><a name="l01754"></a><span class="lineno"> 1754</span>&#160;    ls_types=ls_sampling_list_items[ IDX_FIELD_TYPES ].split( FIELD_DELIMITER )</div><div class="line"><a name="l01755"></a><span class="lineno"> 1755</span>&#160;    <span class="keywordflow">try</span>:</div><div class="line"><a name="l01756"></a><span class="lineno"> 1756</span>&#160;</div><div class="line"><a name="l01757"></a><span class="lineno"> 1757</span>&#160;        lo_field_types=[ eval( s_type ) <span class="keywordflow">for</span> s_type <span class="keywordflow">in</span> ls_types ]</div><div class="line"><a name="l01758"></a><span class="lineno"> 1758</span>&#160;    </div><div class="line"><a name="l01759"></a><span class="lineno"> 1759</span>&#160;    <span class="keywordflow">except</span> Exception <span class="keyword">as</span> oex:</div><div class="line"><a name="l01760"></a><span class="lineno"> 1760</span>&#160;        s_msg=<span class="stringliteral">&quot;In pgdriveneestimator.py, def parse_sampling_list_that_uses_indiv_id_fields, &quot;</span> \</div><div class="line"><a name="l01761"></a><span class="lineno"> 1761</span>&#160;                    + <span class="stringliteral">&quot;Exception trying to evaluate the field type list. &quot;</span>\</div><div class="line"><a name="l01762"></a><span class="lineno"> 1762</span>&#160;                    + <span class="stringliteral">&quot;Type list: &quot;</span> + str( ls_types ) + <span class="stringliteral">&quot;.  Exception raised: &quot;</span> \</div><div class="line"><a name="l01763"></a><span class="lineno"> 1763</span>&#160;                    + str( oex ) + <span class="stringliteral">&quot;.&quot;</span></div><div class="line"><a name="l01764"></a><span class="lineno"> 1764</span>&#160;        <span class="keywordflow">raise</span> Exception ( s_msg )</div><div class="line"><a name="l01765"></a><span class="lineno"> 1765</span>&#160;    <span class="comment">#end try ... except</span></div><div class="line"><a name="l01766"></a><span class="lineno"> 1766</span>&#160;</div><div class="line"><a name="l01767"></a><span class="lineno"> 1767</span>&#160;    ls_params=[]</div><div class="line"><a name="l01768"></a><span class="lineno"> 1768</span>&#160;</div><div class="line"><a name="l01769"></a><span class="lineno"> 1769</span>&#160;    <span class="keywordflow">for</span> idx <span class="keywordflow">in</span> range( IDX_FIRST_PARAM, i_tot_sampling_items ):</div><div class="line"><a name="l01770"></a><span class="lineno"> 1770</span>&#160;        ls_params.append( ls_sampling_list_items[ idx ] )</div><div class="line"><a name="l01771"></a><span class="lineno"> 1771</span>&#160;    <span class="comment">#end for each item with idx&gt;1</span></div><div class="line"><a name="l01772"></a><span class="lineno"> 1772</span>&#160;</div><div class="line"><a name="l01773"></a><span class="lineno"> 1773</span>&#160;    <span class="keywordflow">return</span> ls_field_names, lo_field_types, ls_params</div><div class="line"><a name="l01774"></a><span class="lineno"> 1774</span>&#160;</div><div class="line"><a name="l01775"></a><span class="lineno"> 1775</span>&#160;<span class="comment">#end parse_sampling_list_that_uses_indiv_id_fields</span></div><div class="line"><a name="l01776"></a><span class="lineno"> 1776</span>&#160;</div><div class="ttc" id="namespaceagestrucne_1_1pgdriveneestimator_html_aa1487bffaa5efe1e3ce66928866a6e8f"><div class="ttname"><a href="namespaceagestrucne_1_1pgdriveneestimator.html#aa1487bffaa5efe1e3ce66928866a6e8f">agestrucne.pgdriveneestimator.parse_sampling_list_that_uses_indiv_id_fields</a></div><div class="ttdeci">def parse_sampling_list_that_uses_indiv_id_fields(ls_sampling_list_items)</div><div class="ttdef"><b>Definition:</b> <a href="pgdriveneestimator_8py_source.html#l01715">pgdriveneestimator.py:1715</a></div></div>
</div><!-- fragment -->
</div>
</div>
<a class="anchor" id="a447133ba0efcd50042521407336db81f"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def agestrucne.pgdriveneestimator.raise_exception_if_non_single_pop_file </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>o_genepopfile</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">Not currently in use ( 2016_08_10 ) -- was used when
genepop files with &gt;1 population were not accepted.
</pre> 
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l01183">1183</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>
<div class="fragment"><div class="line"><a name="l01183"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html#a447133ba0efcd50042521407336db81f"> 1183</a></span>&#160;<span class="keyword">def </span><a class="code" href="namespaceagestrucne_1_1pgdriveneestimator.html#a447133ba0efcd50042521407336db81f">raise_exception_if_non_single_pop_file</a>( o_genepopfile ):</div><div class="line"><a name="l01184"></a><span class="lineno"> 1184</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l01185"></a><span class="lineno"> 1185</span>&#160;<span class="stringliteral">    Not currently in use ( 2016_08_10 ) -- was used when</span></div><div class="line"><a name="l01186"></a><span class="lineno"> 1186</span>&#160;<span class="stringliteral">    genepop files with &gt;1 population were not accepted.</span></div><div class="line"><a name="l01187"></a><span class="lineno"> 1187</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l01188"></a><span class="lineno"> 1188</span>&#160;    i_total_pops=o_genepopfile.pop_total</div><div class="line"><a name="l01189"></a><span class="lineno"> 1189</span>&#160;</div><div class="line"><a name="l01190"></a><span class="lineno"> 1190</span>&#160;    <span class="keywordflow">if</span> ( i_total_pops != 1 ):</div><div class="line"><a name="l01191"></a><span class="lineno"> 1191</span>&#160;        s_msg=<span class="stringliteral">&quot;In pgdriveneestimator.py:&quot;</span> + OUTPUT_ENDLINE</div><div class="line"><a name="l01192"></a><span class="lineno"> 1192</span>&#160;        <span class="keywordflow">if</span> i_total_pops &lt; 1:</div><div class="line"><a name="l01193"></a><span class="lineno"> 1193</span>&#160;            s_msg+=<span class="stringliteral">&quot;No \&quot;pop\&quot; entry found in file, &quot;</span> \</div><div class="line"><a name="l01194"></a><span class="lineno"> 1194</span>&#160;                    + o_genepopfile.original_file_name</div><div class="line"><a name="l01195"></a><span class="lineno"> 1195</span>&#160;        <span class="keywordflow">else</span>:</div><div class="line"><a name="l01196"></a><span class="lineno"> 1196</span>&#160;            s_msg+=<span class="stringliteral">&quot;Current implementation of drive accepts genepop &quot;</span> \</div><div class="line"><a name="l01197"></a><span class="lineno"> 1197</span>&#160;                + <span class="stringliteral">&quot;files with only a single pop.&quot;</span></div><div class="line"><a name="l01198"></a><span class="lineno"> 1198</span>&#160;        <span class="comment">#end if lt 1 else more</span></div><div class="line"><a name="l01199"></a><span class="lineno"> 1199</span>&#160;        <span class="keywordflow">raise</span> Exception( s_msg )</div><div class="line"><a name="l01200"></a><span class="lineno"> 1200</span>&#160;    <span class="comment">#end if non-single pop file</span></div><div class="line"><a name="l01201"></a><span class="lineno"> 1201</span>&#160;</div><div class="line"><a name="l01202"></a><span class="lineno"> 1202</span>&#160;    <span class="keywordflow">return</span></div><div class="line"><a name="l01203"></a><span class="lineno"> 1203</span>&#160;<span class="comment">#end raise_exception_non_single_pop_file</span></div><div class="line"><a name="l01204"></a><span class="lineno"> 1204</span>&#160;</div><div class="ttc" id="namespaceagestrucne_1_1pgdriveneestimator_html_a447133ba0efcd50042521407336db81f"><div class="ttname"><a href="namespaceagestrucne_1_1pgdriveneestimator.html#a447133ba0efcd50042521407336db81f">agestrucne.pgdriveneestimator.raise_exception_if_non_single_pop_file</a></div><div class="ttdeci">def raise_exception_if_non_single_pop_file(o_genepopfile)</div><div class="ttdef"><b>Definition:</b> <a href="pgdriveneestimator_8py_source.html#l01183">pgdriveneestimator.py:1183</a></div></div>
</div><!-- fragment -->
</div>
</div>
<a class="anchor" id="ac504fdce8e8069dc8342ab9bcb40c49e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def agestrucne.pgdriveneestimator.set_messaging_procedures </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>s_gui_flag</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<pre class="fragment">2017_05_31. This def was added in order to
control the GUI messaging, and limit it to
the processes that call def mymain from
the GUI-based program.  Console based calls
will have the flag set to "False."
</pre> 
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l03767">3767</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>
<div class="fragment"><div class="line"><a name="l03767"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html#ac504fdce8e8069dc8342ab9bcb40c49e"> 3767</a></span>&#160;<span class="keyword">def </span><a class="code" href="namespaceagestrucne_1_1pgdriveneestimator.html#ac504fdce8e8069dc8342ab9bcb40c49e">set_messaging_procedures</a>( s_gui_flag ):</div><div class="line"><a name="l03768"></a><span class="lineno"> 3768</span>&#160;    <span class="stringliteral">&#39;&#39;&#39;</span></div><div class="line"><a name="l03769"></a><span class="lineno"> 3769</span>&#160;<span class="stringliteral">    2017_05_31. This def was added in order to</span></div><div class="line"><a name="l03770"></a><span class="lineno"> 3770</span>&#160;<span class="stringliteral">    control the GUI messaging, and limit it to</span></div><div class="line"><a name="l03771"></a><span class="lineno"> 3771</span>&#160;<span class="stringliteral">    the processes that call def mymain from</span></div><div class="line"><a name="l03772"></a><span class="lineno"> 3772</span>&#160;<span class="stringliteral">    the GUI-based program.  Console based calls</span></div><div class="line"><a name="l03773"></a><span class="lineno"> 3773</span>&#160;<span class="stringliteral">    will have the flag set to &quot;False.&quot;</span></div><div class="line"><a name="l03774"></a><span class="lineno"> 3774</span>&#160;<span class="stringliteral">    &#39;&#39;&#39;</span></div><div class="line"><a name="l03775"></a><span class="lineno"> 3775</span>&#160;</div><div class="line"><a name="l03776"></a><span class="lineno"> 3776</span>&#160;    <span class="keyword">global</span> ACTIVATE_GUI_MESSAGING</div><div class="line"><a name="l03777"></a><span class="lineno"> 3777</span>&#160;</div><div class="line"><a name="l03778"></a><span class="lineno"> 3778</span>&#160;    <span class="keywordflow">if</span> s_gui_flag == <span class="stringliteral">&quot;True&quot;</span>:</div><div class="line"><a name="l03779"></a><span class="lineno"> 3779</span>&#160;</div><div class="line"><a name="l03780"></a><span class="lineno"> 3780</span>&#160;        <span class="keyword">global</span> GUIERR</div><div class="line"><a name="l03781"></a><span class="lineno"> 3781</span>&#160;        <span class="keyword">global</span> GUIINFO</div><div class="line"><a name="l03782"></a><span class="lineno"> 3782</span>&#160;</div><div class="line"><a name="l03783"></a><span class="lineno"> 3783</span>&#160;        ACTIVATE_GUI_MESSAGING=<span class="keyword">True</span></div><div class="line"><a name="l03784"></a><span class="lineno"> 3784</span>&#160;        <span class="keyword">from</span> <a class="code" href="namespaceagestrucne_1_1pgguiutilities.html">agestrucne.pgguiutilities</a> <span class="keyword">import</span> PGGUIErrorMessage</div><div class="line"><a name="l03785"></a><span class="lineno"> 3785</span>&#160;        <span class="keyword">from</span> <a class="code" href="namespaceagestrucne_1_1pgguiutilities.html">agestrucne.pgguiutilities</a> <span class="keyword">import</span> PGGUIInfoMessage</div><div class="line"><a name="l03786"></a><span class="lineno"> 3786</span>&#160;        GUIERR=PGGUIErrorMessage</div><div class="line"><a name="l03787"></a><span class="lineno"> 3787</span>&#160;        GUIINFO=PGGUIInfoMessage</div><div class="line"><a name="l03788"></a><span class="lineno"> 3788</span>&#160;    <span class="keywordflow">elif</span> s_gui_flag == <span class="stringliteral">&quot;False&quot;</span>:</div><div class="line"><a name="l03789"></a><span class="lineno"> 3789</span>&#160;        ACTIVATE_GUI_MESSAGING=<span class="keyword">False</span></div><div class="line"><a name="l03790"></a><span class="lineno"> 3790</span>&#160;    <span class="keywordflow">else</span>:</div><div class="line"><a name="l03791"></a><span class="lineno"> 3791</span>&#160;        s_msg=<span class="stringliteral">&quot;In pgdriveneestimator, def set_messaging_procedures, &quot;</span> \</div><div class="line"><a name="l03792"></a><span class="lineno"> 3792</span>&#160;                    + <span class="stringliteral">&quot;the program found an unknown value for the &quot;</span> \</div><div class="line"><a name="l03793"></a><span class="lineno"> 3793</span>&#160;                    + <span class="stringliteral">&quot;flag that indicates whether to use GUI &quot;</span> \</div><div class="line"><a name="l03794"></a><span class="lineno"> 3794</span>&#160;                    + <span class="stringliteral">&quot;messaging: &quot;</span> + s_gui_flag + <span class="stringliteral">&quot;.&quot;</span></div><div class="line"><a name="l03795"></a><span class="lineno"> 3795</span>&#160;        <span class="keywordflow">raise</span> Exception( s_msg )</div><div class="line"><a name="l03796"></a><span class="lineno"> 3796</span>&#160;    <span class="comment">#end if we are using gui messaging</span></div><div class="line"><a name="l03797"></a><span class="lineno"> 3797</span>&#160;    <span class="keywordflow">return</span></div><div class="line"><a name="l03798"></a><span class="lineno"> 3798</span>&#160;<span class="comment">#end set_messaging_procedures</span></div><div class="line"><a name="l03799"></a><span class="lineno"> 3799</span>&#160;</div><div class="ttc" id="namespaceagestrucne_1_1pgguiutilities_html"><div class="ttname"><a href="namespaceagestrucne_1_1pgguiutilities.html">agestrucne.pgguiutilities</a></div><div class="ttdef"><b>Definition:</b> <a href="pgguiutilities_8py_source.html#l00001">pgguiutilities.py:1</a></div></div>
<div class="ttc" id="namespaceagestrucne_1_1pgdriveneestimator_html_ac504fdce8e8069dc8342ab9bcb40c49e"><div class="ttname"><a href="namespaceagestrucne_1_1pgdriveneestimator.html#ac504fdce8e8069dc8342ab9bcb40c49e">agestrucne.pgdriveneestimator.set_messaging_procedures</a></div><div class="ttdeci">def set_messaging_procedures(s_gui_flag)</div><div class="ttdef"><b>Definition:</b> <a href="pgdriveneestimator_8py_source.html#l03767">pgdriveneestimator.py:3767</a></div></div>
</div><!-- fragment -->
</div>
</div>
<h2 class="groupheader">Variable Documentation</h2>
<a class="anchor" id="a8cc4108d0e8f1f28708a86f2d97c788b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">list agestrucne.pgdriveneestimator.HIDDEN_ARGS</td>
        </tr>
      </table>
</div><div class="memdoc">
<b>Initial value:</b><div class="fragment"><div class="line"><a name="l00001"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html">    1</a></span>&#160;= [ <span class="stringliteral">&quot;file_object_main&quot;</span>, <span class="stringliteral">&quot;file_object_secondary&quot;</span>, </div><div class="line"><a name="l00002"></a><span class="lineno">    2</span>&#160;                    <span class="stringliteral">&quot;multiprocessing_event&quot;</span>, <span class="stringliteral">&quot;temporary_directory&quot;</span>  ]</div></div><!-- fragment -->
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l00343">343</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>

</div>
</div>
<a class="anchor" id="a632aa31326a0a89aca1ad86c5be25e7e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">list agestrucne.pgdriveneestimator.LS_ARGS_HELP_OPTIONAL</td>
        </tr>
      </table>
</div><div class="memdoc">
<b>Initial value:</b><div class="fragment"><div class="line"><a name="l00001"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html">    1</a></span>&#160;= [  <span class="stringliteral">&quot;total processes to use (single integer) Default is 1 process.&quot;</span>,</div><div class="line"><a name="l00002"></a><span class="lineno">    2</span>&#160;                <span class="stringliteral">&quot;\&quot;no_debug\&quot;, \&quot;debug1\&quot;, \&quot;debug2\&quot;, \&quot;debug3\&quot;, \&quot;testserial\&quot;, \&quot;testmulti\&quot;&quot;</span> \</div><div class="line"><a name="l00003"></a><span class="lineno">    3</span>&#160;                + <span class="stringliteral">&quot;.  Indicates a run mode. The default is \&quot;no_debug\&quot;, which runs multiplexed &quot;</span> \</div><div class="line"><a name="l00004"></a><span class="lineno">    4</span>&#160;                + <span class="stringliteral">&quot;with standard output.  &quot;</span> \</div><div class="line"><a name="l00005"></a><span class="lineno">    5</span>&#160;                + <span class="stringliteral">&quot;Other modes except \&quot;testmulti\&quot;, run non-parallelized, with increasing output.  &quot;</span> \</div><div class="line"><a name="l00006"></a><span class="lineno">    6</span>&#160;                + <span class="stringliteral">&quot;Debug 3, for example, adds to the output a table listing, for each indiv. &quot;</span> \</div><div class="line"><a name="l00007"></a><span class="lineno">    7</span>&#160;                + <span class="stringliteral">&quot;in each file, which replicate Ne estimates include the individual.  It also preserves &quot;</span> \</div><div class="line"><a name="l00008"></a><span class="lineno">    8</span>&#160;                + <span class="stringliteral">&quot;the intermediate genepop and LDNe2&#39;s output files&quot;</span>,</div><div class="line"><a name="l00009"></a><span class="lineno">    9</span>&#160;                <span class="stringliteral">&quot;An Nb/Ne ratio, if supplied, will be used in a bias adjustment to the LDNE &quot;</span> \</div><div class="line"><a name="l00010"></a><span class="lineno">   10</span>&#160;                + <span class="stringliteral">&quot;estimates (after Waples, 2014).  Ignored if --donbbiasadjust is \&quot;False.\&quot;  &quot;</span> \</div><div class="line"><a name="l00011"></a><span class="lineno">   11</span>&#160;                    + <span class="stringliteral">&quot;If you use 0.0 here, the program will look in the genepop file &quot;</span> \</div><div class="line"><a name="l00012"></a><span class="lineno">   12</span>&#160;                    + <span class="stringliteral">&quot;header for an entry giving the Nb/Ne ratio as \&quot;nbne=&lt;value&gt;\&quot;.&quot;</span>, </div><div class="line"><a name="l00013"></a><span class="lineno">   13</span>&#160;                <span class="stringliteral">&quot;True|False, whether to do the Nb bias adjustment. &quot;</span> \</div><div class="line"><a name="l00014"></a><span class="lineno">   14</span>&#160;                + <span class="stringliteral">&quot;If False, the Nb/Ne ratio parameter is ignored.&quot;</span> ]</div></div><!-- fragment -->
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l00263">263</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>

</div>
</div>
<a class="anchor" id="a14642564b63c70a5e681137460d5badc"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">list agestrucne.pgdriveneestimator.LS_FLAGS_LONG_REQUIRED</td>
        </tr>
      </table>
</div><div class="memdoc">
<b>Initial value:</b><div class="fragment"><div class="line"><a name="l00001"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html">    1</a></span>&#160;= [ <span class="stringliteral">&quot;--gpfiles&quot;</span>, <span class="stringliteral">&quot;--scheme&quot;</span>, <span class="stringliteral">&quot;--params&quot;</span>, <span class="stringliteral">&quot;--minpopsize&quot;</span>, <span class="stringliteral">&quot;--maxpopsize&quot;</span>, <span class="stringliteral">&quot;--poprange&quot;</span>, </div><div class="line"><a name="l00002"></a><span class="lineno">    2</span>&#160;                        <span class="stringliteral">&quot;--minallelefreq&quot;</span>,  <span class="stringliteral">&quot;--monogamy&quot;</span>,  <span class="stringliteral">&quot;--chromlocifile&quot;</span>, <span class="stringliteral">&quot;--allelepairingscheme&quot;</span>, </div><div class="line"><a name="l00003"></a><span class="lineno">    3</span>&#160;                        <span class="stringliteral">&quot;--replicates&quot;</span>, <span class="stringliteral">&quot;--locischeme&quot;</span>, <span class="stringliteral">&quot;--locischemeparams&quot;</span>,  </div><div class="line"><a name="l00004"></a><span class="lineno">    4</span>&#160;                        <span class="stringliteral">&quot;--mintotalloci&quot;</span>, <span class="stringliteral">&quot;--maxtotalloci&quot;</span>, <span class="stringliteral">&quot;--locirange&quot;</span>, <span class="stringliteral">&quot;--locireplicates&quot;</span> ]</div></div><!-- fragment -->
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l00185">185</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>

</div>
</div>
<a class="anchor" id="a7c2814699a352b30869020a5235e308c"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">list agestrucne.pgdriveneestimator.MAIN_TABLE_RUN_INFO_COLS</td>
        </tr>
      </table>
</div><div class="memdoc">
<b>Initial value:</b><div class="fragment"><div class="line"><a name="l00001"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html">    1</a></span>&#160;= [ <span class="stringliteral">&quot;original_file&quot;</span>, <span class="stringliteral">&quot;pop&quot;</span>, <span class="stringliteral">&quot;census&quot;</span>,   <span class="stringliteral">&quot;indiv_count&quot;</span>, </div><div class="line"><a name="l00002"></a><span class="lineno">    2</span>&#160;                            <span class="stringliteral">&quot;sample_value&quot;</span>, <span class="stringliteral">&quot;replicate_number&quot;</span>, <span class="stringliteral">&quot;loci_sample_value&quot;</span>, </div><div class="line"><a name="l00003"></a><span class="lineno">    3</span>&#160;                            <span class="stringliteral">&quot;loci_replicate_number&quot;</span>,<span class="stringliteral">&quot;min_allele_freq&quot;</span> ]</div></div><!-- fragment -->
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l00479">479</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>

</div>
</div>
<a class="anchor" id="a12890ea18077e7bdd6904a5907a4df8f"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">list agestrucne.pgdriveneestimator.SAMPLE_SCHEMES_NON_CRITERIA</td>
        </tr>
      </table>
</div><div class="memdoc">
<b>Initial value:</b><div class="fragment"><div class="line"><a name="l00001"></a><span class="lineno"><a class="line" href="namespaceagestrucne_1_1pgdriveneestimator.html">    1</a></span>&#160;= [ SAMPLE_BY_PERCENTAGES, </div><div class="line"><a name="l00002"></a><span class="lineno">    2</span>&#160;                                    SAMPLE_BY_REMOVAL ]</div></div><!-- fragment -->
<p>Definition at line <a class="el" href="pgdriveneestimator_8py_source.html#l00420">420</a> of file <a class="el" href="pgdriveneestimator_8py_source.html">pgdriveneestimator.py</a>.</p>

</div>
</div>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.11
</small></address>
</body>
</html>
